
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />
<meta property="og:title" content="T027 · Kinase similarity: Ligand profile" />
<meta property="og:type" content="website" />
<meta property="og:url" content="https://projects.volkamerlab.org/teachopencadd/talktorials/T027_kinase_similarity_ligand_profile.html" />
<meta property="og:site_name" content="TeachOpenCADD" />
<meta property="og:description" content="Note: This talktorial is a part of TeachOpenCADD, a platform that aims to teach domain-specific skills and to provide pipeline templates as starting points for research projects. Authors: Talia B. Kimber, 2021, Volkamer lab, Charité, Dominique Sydow, 2021, Volkamer lab, Charité, Andrea Volkamer, ..." />
<meta property="og:image" content="https://raw.githubusercontent.com/volkamerlab/teachopencadd/master/docs/_static/images/TeachOpenCADD_topics.png" />
<meta property="og:image:alt" content="TeachOpenCADD" />
<meta name="description" content="Note: This talktorial is a part of TeachOpenCADD, a platform that aims to teach domain-specific skills and to provide pipeline templates as starting points for research projects. Authors: Talia B. Kimber, 2021, Volkamer lab, Charité, Dominique Sydow, 2021, Volkamer lab, Charité, Andrea Volkamer, ..." />

  <meta name="viewport" content="width=device-width,initial-scale=1">
  <meta http-equiv="x-ua-compatible" content="ie=edge">
  <meta name="lang:clipboard.copy" content="Copy to clipboard">
  <meta name="lang:clipboard.copied" content="Copied to clipboard">
  <meta name="lang:search.language" content="en">
  <meta name="lang:search.pipeline.stopwords" content="True">
  <meta name="lang:search.pipeline.trimmer" content="True">
  <meta name="lang:search.result.none" content="No matching documents">
  <meta name="lang:search.result.one" content="1 matching document">
  <meta name="lang:search.result.other" content="# matching documents">
  <meta name="lang:search.tokenizer" content="[\s\-]+">

  
    <link href="https://fonts.gstatic.com/" rel="preconnect" crossorigin>
    <link href="https://fonts.googleapis.com/css?family=Roboto+Mono:400,500,700|Roboto:300,400,400i,700&display=fallback" rel="stylesheet">

    <style>
      body,
      input {
        font-family: "Roboto", "Helvetica Neue", Helvetica, Arial, sans-serif
      }

      code,
      kbd,
      pre {
        font-family: "Roboto Mono", "Courier New", Courier, monospace
      }
    </style>
  

  <link rel="stylesheet" href="../_static/stylesheets/application.css"/>
  <link rel="stylesheet" href="../_static/stylesheets/application-palette.css"/>
  <link rel="stylesheet" href="../_static/stylesheets/application-fixes.css"/>
  
  <link rel="stylesheet" href="../_static/fonts/material-icons.css"/>
  
  <meta name="theme-color" content="#2196f3">
  <script src="../_static/javascripts/modernizr.js"></script>
  
    <script async src="../_static/cookieconsent.min.js"></script>
    <script>
        window.addEventListener("load", function(){
        window.cookieconsent.initialise({
            "palette": {
            "popup": {
                "background": "#f0f0f0",
                "text": "#999"
            },
            "button": {
                "text": "#fff",
                "background": "#009688"
            }
            },
            "theme": "classic"
        })});
    </script>
    <!-- Global site tag (gtag.js) - Google Analytics -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=G-Q6ZE82CNZB"></script>
    <script>
        window.dataLayer = window.dataLayer || [];
        function gtag(){dataLayer.push(arguments);}
        gtag('js', new Date());

        gtag('config', 'G-Q6ZE82CNZB');
    </script>
  
    <link rel="apple-touch-icon" href="../_static/images/apple-icon-152x152.png"/>
  
  
    <title>T027 · Kinase similarity: Ligand profile &#8212; TeachOpenCADD 0 documentation</title>
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../_static/material.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/custom.css" />
    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/sphinx_highlight.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="https://unpkg.com/@jupyter-widgets/html-manager@^0.20.1/dist/embed-amd.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="T028 · Kinase similarity: Compare different perspectives" href="T028_kinase_similarity_compare_perspectives.html" />
    <link rel="prev" title="T026 · Kinase similarity: Interaction fingerprints" href="T026_kinase_similarity_ifp.html" />
  
    <link rel="apple-touch-icon" href="../_static/images/apple-icon-152x152.png"/>
  
   

  </head>
  <body dir=ltr
        data-md-color-primary=teal data-md-color-accent=cyan>
  
  <svg class="md-svg">
    <defs data-children-count="0">
      
      <svg xmlns="http://www.w3.org/2000/svg" width="416" height="448" viewBox="0 0 416 448" id="__github"><path fill="currentColor" d="M160 304q0 10-3.125 20.5t-10.75 19T128 352t-18.125-8.5-10.75-19T96 304t3.125-20.5 10.75-19T128 256t18.125 8.5 10.75 19T160 304zm160 0q0 10-3.125 20.5t-10.75 19T288 352t-18.125-8.5-10.75-19T256 304t3.125-20.5 10.75-19T288 256t18.125 8.5 10.75 19T320 304zm40 0q0-30-17.25-51T296 232q-10.25 0-48.75 5.25Q229.5 240 208 240t-39.25-2.75Q130.75 232 120 232q-29.5 0-46.75 21T56 304q0 22 8 38.375t20.25 25.75 30.5 15 35 7.375 37.25 1.75h42q20.5 0 37.25-1.75t35-7.375 30.5-15 20.25-25.75T360 304zm56-44q0 51.75-15.25 82.75-9.5 19.25-26.375 33.25t-35.25 21.5-42.5 11.875-42.875 5.5T212 416q-19.5 0-35.5-.75t-36.875-3.125-38.125-7.5-34.25-12.875T37 371.5t-21.5-28.75Q0 312 0 260q0-59.25 34-99-6.75-20.5-6.75-42.5 0-29 12.75-54.5 27 0 47.5 9.875t47.25 30.875Q171.5 96 212 96q37 0 70 8 26.25-20.5 46.75-30.25T376 64q12.75 25.5 12.75 54.5 0 21.75-6.75 42 34 40 34 99.5z"/></svg>
      
    </defs>
  </svg>
  
  <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer">
  <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search">
  <label class="md-overlay" data-md-component="overlay" for="__drawer"></label>
  <a href="#talktorials/T027_kinase_similarity_ligand_profile" tabindex="1" class="md-skip"> Skip to content </a>
  <header class="md-header" data-md-component="header">
  <nav class="md-header-nav md-grid">
    <div class="md-flex navheader">
      <div class="md-flex__cell md-flex__cell--shrink">
        <a href="../index.html" title="TeachOpenCADD 0 documentation"
           class="md-header-nav__button md-logo">
          
            <i class="md-icon">school</i>
          
        </a>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        <label class="md-icon md-icon--menu md-header-nav__button" for="__drawer"></label>
      </div>
      <div class="md-flex__cell md-flex__cell--stretch">
        <div class="md-flex__ellipsis md-header-nav__title" data-md-component="title">
          <span class="md-header-nav__topic">TeachOpenCADD</span>
          <span class="md-header-nav__topic"> T027 · Kinase similarity: Ligand profile </span>
        </div>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        <label class="md-icon md-icon--search md-header-nav__button" for="__search"></label>
        
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" action="../search.html" method="get" name="search">
      <input type="text" class="md-search__input" name="q" placeholder="Search"
             autocapitalize="off" autocomplete="off" spellcheck="false"
             data-md-component="query" data-md-state="active">
      <label class="md-icon md-search__icon" for="__search"></label>
      <button type="reset" class="md-icon md-search__icon" data-md-component="reset" tabindex="-1">
        &#xE5CD;
      </button>
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="result">
          <div class="md-search-result__meta">
            Type to start searching
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>

      </div>
      
        <div class="md-flex__cell md-flex__cell--shrink">
          <div class="md-header-nav__source">
            <a href="https://github.com/volkamerlab/teachopencadd/" title="Go to repository" class="md-source" data-md-source="github">

    <div class="md-source__icon">
      <svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 24 24" width="28" height="28">
        <use xlink:href="#__github" width="24" height="24"></use>
      </svg>
    </div>
  
  <div class="md-source__repository">
    TeachOpenCADD
  </div>
</a>
          </div>
        </div>
      
      
    </div>
  </nav>
</header>

  
  <div class="md-container">
    
    
    
  <nav class="md-tabs" data-md-component="tabs">
    <div class="md-tabs__inner md-grid">
      <ul class="md-tabs__list">
            
            <li class="md-tabs__item"><a href="../talktorials.html" class="md-tabs__link">Our talktorials</a></li>
            
            <li class="md-tabs__item"><a href="../installing.html" class="md-tabs__link">Run locally</a></li>
            
            <li class="md-tabs__item"><a href="../contribute.html" class="md-tabs__link">Development</a></li>
            
            <li class="md-tabs__item"><a href="../contact.html" class="md-tabs__link">Contact</a></li>
            
            <li class="md-tabs__item"><a href="../citation.html" class="md-tabs__link">Citation</a></li>
          <li class="md-tabs__item"><a href="../all_talktorials.html" class="md-tabs__link">Complete list of talktorials</a></li>
      </ul>
    </div>
  </nav>
    <main class="md-main">
      <div class="md-main__inner md-grid" data-md-component="container">
        
          <div class="md-sidebar md-sidebar--primary" data-md-component="navigation">
            <div class="md-sidebar__scrollwrap">
              <div class="md-sidebar__inner">
                <nav class="md-nav md-nav--primary" data-md-level="0">
  <label class="md-nav__title md-nav__title--site" for="__drawer">
    <a href="../index.html" title="TeachOpenCADD 0 documentation" class="md-nav__button md-logo">
      
        <i class="md-icon">school</i>
      
    </a>
    <a href="../index.html"
       title="TeachOpenCADD 0 documentation">TeachOpenCADD</a>
  </label>
    <div class="md-nav__source">
      <a href="https://github.com/volkamerlab/teachopencadd/" title="Go to repository" class="md-source" data-md-source="github">

    <div class="md-source__icon">
      <svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 24 24" width="28" height="28">
        <use xlink:href="#__github" width="24" height="24"></use>
      </svg>
    </div>
  
  <div class="md-source__repository">
    TeachOpenCADD
  </div>
</a>
    </div>
  
  

  
  <ul class="md-nav__list">
    <li class="md-nav__item">
    
      <span class="md-nav__link caption"><span class="caption-text">Our talktorials</span></span>
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="../all_talktorials.html" class="md-nav__link">Complete list of talktorials</a>
      <ul class="md-nav__list"> 
    <li class="md-nav__item">
    
    
      <a href="T001_query_chembl.html" class="md-nav__link">T001 · Compound data acquisition (ChEMBL)</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T002_compound_adme.html" class="md-nav__link">T002 · Molecular filtering: ADME and lead-likeness criteria</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T003_compound_unwanted_substructures.html" class="md-nav__link">T003 · Molecular filtering: unwanted substructures</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T004_compound_similarity.html" class="md-nav__link">T004 · Ligand-based screening: compound similarity</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T005_compound_clustering.html" class="md-nav__link">T005 · Compound clustering</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T006_compound_maximum_common_substructures.html" class="md-nav__link">T006 · Maximum common substructure</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T007_compound_activity_machine_learning.html" class="md-nav__link">T007 · Ligand-based screening: machine learning</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T008_query_pdb.html" class="md-nav__link">T008 · Protein data acquisition: Protein Data Bank (PDB)</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T009_compound_ensemble_pharmacophores.html" class="md-nav__link">T009 · Ligand-based pharmacophores</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T010_binding_site_comparison.html" class="md-nav__link">T010 · Binding site similarity and off-target prediction</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T011_query_online_api_webservices.html" class="md-nav__link">T011 · Querying online API webservices</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T012_query_klifs.html" class="md-nav__link">T012 · Data acquisition from KLIFS</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T013_query_pubchem.html" class="md-nav__link">T013 · Data acquisition from PubChem</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T014_binding_site_detection.html" class="md-nav__link">T014 · Binding site detection</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T015_protein_ligand_docking.html" class="md-nav__link">T015 · Protein ligand docking</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T016_protein_ligand_interactions.html" class="md-nav__link">T016 · Protein-ligand interactions</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T017_advanced_nglview_usage.html" class="md-nav__link">T017 · Advanced NGLview usage</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T018_automated_cadd_pipeline.html" class="md-nav__link">T018 · Automated pipeline for lead optimization</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T019_md_simulation.html" class="md-nav__link">T019 · Molecular dynamics simulation</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T020_md_analysis.html" class="md-nav__link">T020 · Analyzing molecular dynamics simulations</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T021_one_hot_encoding.html" class="md-nav__link">T021 · One-Hot Encoding</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T022_ligand_based_screening_neural_network.html" class="md-nav__link">T022 · Ligand-based screening: neural networks</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T023_what_is_a_kinase.html" class="md-nav__link">T023 · What is a kinase?</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T024_kinase_similarity_sequence.html" class="md-nav__link">T024 · Kinase similarity: Sequence</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T025_kinase_similarity_kissim.html" class="md-nav__link">T025 · Kinase similarity: Kinase pocket (KiSSim fingerprint)</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T026_kinase_similarity_ifp.html" class="md-nav__link">T026 · Kinase similarity: Interaction fingerprints</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
    <input class="md-toggle md-nav__toggle" data-md-toggle="toc" type="checkbox" id="__toc">
    <label class="md-nav__link md-nav__link--active" for="__toc"> T027 · Kinase similarity: Ligand profile </label>
    
      <a href="#" class="md-nav__link md-nav__link--active">T027 · Kinase similarity: Ligand profile</a>
      
        
<nav class="md-nav md-nav--secondary">
    <label class="md-nav__title" for="__toc">Contents</label>
  <ul class="md-nav__list" data-md-scrollfix="">
        <li class="md-nav__item"><a href="#talktorials-t027-kinase-similarity-ligand-profile--page-root" class="md-nav__link">T027 · Kinase similarity: Ligand profile</a><nav class="md-nav">
              <ul class="md-nav__list">
        <li class="md-nav__item"><a href="#Aim-of-this-talktorial" class="md-nav__link">Aim of this talktorial</a><nav class="md-nav">
              <ul class="md-nav__list">
        <li class="md-nav__item"><a href="#Contents-in-Theory" class="md-nav__link">Contents in <em>Theory</em></a>
        </li>
        <li class="md-nav__item"><a href="#Contents-in-Practical" class="md-nav__link">Contents in <em>Practical</em></a>
        </li>
        <li class="md-nav__item"><a href="#References" class="md-nav__link">References</a>
        </li></ul>
            </nav>
        </li>
        <li class="md-nav__item"><a href="#Theory" class="md-nav__link">Theory</a><nav class="md-nav">
              <ul class="md-nav__list">
        <li class="md-nav__item"><a href="#Kinase-dataset" class="md-nav__link">Kinase dataset</a>
        </li>
        <li class="md-nav__item"><a href="#Bioactivity-data" class="md-nav__link">Bioactivity data</a>
        </li>
        <li class="md-nav__item"><a href="#Kinase-similarity-descriptor:-Ligand-profile" class="md-nav__link">Kinase similarity descriptor: Ligand profile</a><nav class="md-nav">
              <ul class="md-nav__list">
        <li class="md-nav__item"><a href="#Kinase-similarity" class="md-nav__link">Kinase similarity</a>
        </li>
        <li class="md-nav__item"><a href="#Kinase-promiscuity" class="md-nav__link">Kinase promiscuity</a>
        </li></ul>
            </nav>
        </li>
        <li class="md-nav__item"><a href="#From-similarity-matrix-to-distance-matrix" class="md-nav__link">From similarity matrix to distance matrix</a>
        </li></ul>
            </nav>
        </li>
        <li class="md-nav__item"><a href="#Practical" class="md-nav__link">Practical</a><nav class="md-nav">
              <ul class="md-nav__list">
        <li class="md-nav__item"><a href="#Define-the-kinases-of-interest" class="md-nav__link">Define the kinases of interest</a>
        </li>
        <li class="md-nav__item"><a href="#Retrieve-the-data" class="md-nav__link">Retrieve the data</a>
        </li>
        <li class="md-nav__item"><a href="#Preprocess-the-data" class="md-nav__link">Preprocess the data</a><nav class="md-nav">
              <ul class="md-nav__list">
        <li class="md-nav__item"><a href="#Hit-or-non-hit" class="md-nav__link">Hit or non-hit</a>
        </li></ul>
            </nav>
        </li>
        <li class="md-nav__item"><a href="#id1" class="md-nav__link">Kinase promiscuity</a>
        </li>
        <li class="md-nav__item"><a href="#id2" class="md-nav__link">Kinase similarity</a><nav class="md-nav">
              <ul class="md-nav__list">
        <li class="md-nav__item"><a href="#Visualize-similarity-as-kinase-matrix" class="md-nav__link">Visualize similarity as kinase matrix</a>
        </li>
        <li class="md-nav__item"><a href="#Save-kinase-similarity-matrix" class="md-nav__link">Save kinase similarity matrix</a>
        </li></ul>
            </nav>
        </li>
        <li class="md-nav__item"><a href="#Kinase-distance-matrix" class="md-nav__link">Kinase distance matrix</a><nav class="md-nav">
              <ul class="md-nav__list">
        <li class="md-nav__item"><a href="#Save-kinase-distance-matrix" class="md-nav__link">Save kinase distance matrix</a>
        </li></ul>
            </nav>
        </li></ul>
            </nav>
        </li>
        <li class="md-nav__item"><a href="#Discussion" class="md-nav__link">Discussion</a>
        </li>
        <li class="md-nav__item"><a href="#Quiz" class="md-nav__link">Quiz</a>
        </li></ul>
            </nav>
        </li>
    
<li class="md-nav__item"><a class="md-nav__extra_link" href="../_sources/talktorials/T027_kinase_similarity_ligand_profile.nblink.txt">Show Source</a> </li>

  </ul>
</nav>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T028_kinase_similarity_compare_perspectives.html" class="md-nav__link">T028 · Kinase similarity: Compare different perspectives</a>
      
    
    </li></ul>
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="../talktorials.html" class="md-nav__link">Talktorials by collection</a>
      
    
    </li>
    <li class="md-nav__item">
    
      <span class="md-nav__link caption"><span class="caption-text">Run locally</span></span>
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="../installing.html" class="md-nav__link">Installing</a>
      
    
    </li>
    <li class="md-nav__item">
    
      <span class="md-nav__link caption"><span class="caption-text">Development</span></span>
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="../contribute.html" class="md-nav__link">For contributors</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="../api.html" class="md-nav__link">API Documentation</a>
      
    
    </li>
    <li class="md-nav__item">
    
      <span class="md-nav__link caption"><span class="caption-text">About TeachOpenCADD</span></span>
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="../contact.html" class="md-nav__link">Contact</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="../acknowledgments.html" class="md-nav__link">Acknowledgments</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="../citation.html" class="md-nav__link">Citation</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="../license.html" class="md-nav__link">License</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="../funding.html" class="md-nav__link">Funding</a>
      
    
    </li>
    <li class="md-nav__item">
    
      <span class="md-nav__link caption"><span class="caption-text">External resources</span></span>
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="../external_dependencies.html" class="md-nav__link">Packages and webservers used in TeachOpenCADD</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="../external_tutorials_collections.html" class="md-nav__link">External tutorials and collections</a>
      
    
    </li>
  </ul>
  

</nav>
              </div>
            </div>
          </div>
          <div class="md-sidebar md-sidebar--secondary" data-md-component="toc">
            <div class="md-sidebar__scrollwrap">
              <div class="md-sidebar__inner">
                
<nav class="md-nav md-nav--secondary">
    <label class="md-nav__title" for="__toc">Contents</label>
  <ul class="md-nav__list" data-md-scrollfix="">
        <li class="md-nav__item"><a href="#talktorials-t027-kinase-similarity-ligand-profile--page-root" class="md-nav__link">T027 · Kinase similarity: Ligand profile</a><nav class="md-nav">
              <ul class="md-nav__list">
        <li class="md-nav__item"><a href="#Aim-of-this-talktorial" class="md-nav__link">Aim of this talktorial</a><nav class="md-nav">
              <ul class="md-nav__list">
        <li class="md-nav__item"><a href="#Contents-in-Theory" class="md-nav__link">Contents in <em>Theory</em></a>
        </li>
        <li class="md-nav__item"><a href="#Contents-in-Practical" class="md-nav__link">Contents in <em>Practical</em></a>
        </li>
        <li class="md-nav__item"><a href="#References" class="md-nav__link">References</a>
        </li></ul>
            </nav>
        </li>
        <li class="md-nav__item"><a href="#Theory" class="md-nav__link">Theory</a><nav class="md-nav">
              <ul class="md-nav__list">
        <li class="md-nav__item"><a href="#Kinase-dataset" class="md-nav__link">Kinase dataset</a>
        </li>
        <li class="md-nav__item"><a href="#Bioactivity-data" class="md-nav__link">Bioactivity data</a>
        </li>
        <li class="md-nav__item"><a href="#Kinase-similarity-descriptor:-Ligand-profile" class="md-nav__link">Kinase similarity descriptor: Ligand profile</a><nav class="md-nav">
              <ul class="md-nav__list">
        <li class="md-nav__item"><a href="#Kinase-similarity" class="md-nav__link">Kinase similarity</a>
        </li>
        <li class="md-nav__item"><a href="#Kinase-promiscuity" class="md-nav__link">Kinase promiscuity</a>
        </li></ul>
            </nav>
        </li>
        <li class="md-nav__item"><a href="#From-similarity-matrix-to-distance-matrix" class="md-nav__link">From similarity matrix to distance matrix</a>
        </li></ul>
            </nav>
        </li>
        <li class="md-nav__item"><a href="#Practical" class="md-nav__link">Practical</a><nav class="md-nav">
              <ul class="md-nav__list">
        <li class="md-nav__item"><a href="#Define-the-kinases-of-interest" class="md-nav__link">Define the kinases of interest</a>
        </li>
        <li class="md-nav__item"><a href="#Retrieve-the-data" class="md-nav__link">Retrieve the data</a>
        </li>
        <li class="md-nav__item"><a href="#Preprocess-the-data" class="md-nav__link">Preprocess the data</a><nav class="md-nav">
              <ul class="md-nav__list">
        <li class="md-nav__item"><a href="#Hit-or-non-hit" class="md-nav__link">Hit or non-hit</a>
        </li></ul>
            </nav>
        </li>
        <li class="md-nav__item"><a href="#id1" class="md-nav__link">Kinase promiscuity</a>
        </li>
        <li class="md-nav__item"><a href="#id2" class="md-nav__link">Kinase similarity</a><nav class="md-nav">
              <ul class="md-nav__list">
        <li class="md-nav__item"><a href="#Visualize-similarity-as-kinase-matrix" class="md-nav__link">Visualize similarity as kinase matrix</a>
        </li>
        <li class="md-nav__item"><a href="#Save-kinase-similarity-matrix" class="md-nav__link">Save kinase similarity matrix</a>
        </li></ul>
            </nav>
        </li>
        <li class="md-nav__item"><a href="#Kinase-distance-matrix" class="md-nav__link">Kinase distance matrix</a><nav class="md-nav">
              <ul class="md-nav__list">
        <li class="md-nav__item"><a href="#Save-kinase-distance-matrix" class="md-nav__link">Save kinase distance matrix</a>
        </li></ul>
            </nav>
        </li></ul>
            </nav>
        </li>
        <li class="md-nav__item"><a href="#Discussion" class="md-nav__link">Discussion</a>
        </li>
        <li class="md-nav__item"><a href="#Quiz" class="md-nav__link">Quiz</a>
        </li></ul>
            </nav>
        </li>
    
<li class="md-nav__item"><a class="md-nav__extra_link" href="../_sources/talktorials/T027_kinase_similarity_ligand_profile.nblink.txt">Show Source</a> </li>

<li id="searchbox" class="md-nav__item"></li>

  </ul>
</nav>
              </div>
            </div>
          </div>
        
        <div class="md-content">
          <article class="md-content__inner md-typeset" role="main">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars and line breaks on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
    white-space: pre;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt .copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
.jp-RenderedHTMLCommon table,
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
.jp-RenderedHTMLCommon thead,
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
.jp-RenderedHTMLCommon tr,
.jp-RenderedHTMLCommon th,
.jp-RenderedHTMLCommon td,
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
.jp-RenderedHTMLCommon th,
div.rendered_html th {
  font-weight: bold;
}
.jp-RenderedHTMLCommon tbody tr:nth-child(odd),
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
.jp-RenderedHTMLCommon tbody tr:hover,
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}
</style>
<section id="T027-·-Kinase-similarity:-Ligand-profile">
<h1 id="talktorials-t027-kinase-similarity-ligand-profile--page-root">T027 · Kinase similarity: Ligand profile<a class="headerlink" href="#talktorials-t027-kinase-similarity-ligand-profile--page-root" title="Permalink to this heading">¶</a></h1>
<p><strong>Note:</strong> This talktorial is a part of TeachOpenCADD, a platform that aims to teach domain-specific skills and to provide pipeline templates as starting points for research projects.</p>
<p>Authors:</p>
<ul class="simple">
<li><p>Talia B. Kimber, 2021, <a class="reference external" href="https://volkamerlab.org/">Volkamer lab, Charité</a></p></li>
<li><p>Dominique Sydow, 2021, <a class="reference external" href="https://volkamerlab.org/">Volkamer lab, Charité</a></p></li>
<li><p>Andrea Volkamer, 2021, <a class="reference external" href="https://volkamerlab.org/">Volkamer lab, Charité</a></p></li>
</ul>
<section id="Aim-of-this-talktorial">
<h2 id="Aim-of-this-talktorial">Aim of this talktorial<a class="headerlink" href="#Aim-of-this-talktorial" title="Permalink to this heading">¶</a></h2>
<p>The aim of this talktorial is to investigate kinase similarity through ligand profiling data (ChEMBL29). In the context of drug design, the following assumption is often made: if a compound was tested active on two different kinases, it is suspected that these two kinases may have some degree of similarity. We will use this assumption in this talktorial. The concept of kinase promiscuity is also covered.</p>
<section id="Contents-in-Theory">
<h3 id="Contents-in-Theory">Contents in <em>Theory</em><a class="headerlink" href="#Contents-in-Theory" title="Permalink to this heading">¶</a></h3>
<ul class="simple">
<li><p>Kinase dataset</p></li>
<li><p>Bioactivity data</p></li>
<li><p>Kinase similarity descriptor: Ligand profile</p>
<ul>
<li><p>Kinase similarity</p></li>
<li><p>Kinase promiscuity</p></li>
</ul>
</li>
<li><p>From similarity matrix to distance matrix</p></li>
</ul>
</section>
<section id="Contents-in-Practical">
<h3 id="Contents-in-Practical">Contents in <em>Practical</em><a class="headerlink" href="#Contents-in-Practical" title="Permalink to this heading">¶</a></h3>
<ul class="simple">
<li><p>Define the kinases of interest</p></li>
<li><p>Retrieve the data</p></li>
<li><p>Preprocess the data</p>
<ul>
<li><p>Hit or non-hit</p></li>
</ul>
</li>
<li><p>Kinase promiscuity</p></li>
<li><p>Kinase similarity</p>
<ul>
<li><p>Visualize similarity as kinase matrix</p></li>
<li><p>Save kinase similarity matrix</p></li>
</ul>
</li>
<li><p>Kinase distance matrix</p>
<ul>
<li><p>Save kinase distance matrix</p></li>
</ul>
</li>
</ul>
</section>
<section id="References">
<h3 id="References">References<a class="headerlink" href="#References" title="Permalink to this heading">¶</a></h3>
<ul class="simple">
<li><p>Kinase dataset: <a class="reference external" href="https://www.mdpi.com/1420-3049/26/3/629">Molecules (2021), 26(3), 629</a></p></li>
<li><p>ChEMBL database</p>
<ul>
<li><p>Website: <a class="reference external" href="https://www.ebi.ac.uk/chembl/">https://www.ebi.ac.uk/chembl/</a></p></li>
<li><p>Paper: <a class="reference external" href="https://doi.org/10.1093/nar/gkw1074">Nucleic Acid Research (2017), 45(D1), D945-D954</a></p></li>
</ul>
</li>
<li><p>KinMap</p>
<ul>
<li><p>Website: <a class="reference external" href="http://www.kinhub.org/kinmap/">http://www.kinhub.org/kinmap/</a></p></li>
<li><p>Paper: <a class="reference external" href="https://doi.org/10.1186/s12859-016-1433-7">BMC Bioinformatics (2017), 18(1), 16</a></p></li>
</ul>
</li>
</ul>
</section>
</section>
<section id="Theory">
<h2 id="Theory">Theory<a class="headerlink" href="#Theory" title="Permalink to this heading">¶</a></h2>
<section id="Kinase-dataset">
<h3 id="Kinase-dataset">Kinase dataset<a class="headerlink" href="#Kinase-dataset" title="Permalink to this heading">¶</a></h3>
<p>We use the kinase selection as defined in <strong>Talktorial T023</strong>.</p>
</section>
<section id="Bioactivity-data">
<h3 id="Bioactivity-data">Bioactivity data<a class="headerlink" href="#Bioactivity-data" title="Permalink to this heading">¶</a></h3>
<p>In order to measure kinase similarity through ligand profiling data, bioactivity data is retrieved from the well-known <a class="reference external" href="https://www.ebi.ac.uk/chembl/">ChEMBL</a> database and the query focuses on human kinases. Luckily, a curated version of ChEMBL29 is already freely available through the Openkinome organization, see <a class="reference external" href="https://github.com/openkinome/kinodata">https://github.com/openkinome/kinodata</a>. For more details on querying the ChEMBL database, please refer to <strong>Talktorial T001</strong>.</p>
<p>In drug design, it is common to binarize the activity of a compound against a target of interest as a “hit” or “non-hit”. Practically speaking, this is done using a cutoff value for measured activity. If the activity is greater than the cutoff, the compound is labeled as active (hit), and inactive (non-hit) otherwise.</p>
<img alt="Manning tree with number of ChEMBL activities per kinase (KinMap)" src="../_images/kinmap_n_activities_per_kinase.png"/>
<p><em>Figure 1:</em> Number of ChEMBL29 bioactivities per kinase - as collected in <a class="reference external" href="https://github.com/openkinome/kinodata">kinodata</a> - mapped onto the Manning kinome tree using <a class="reference external" href="http://www.kinhub.org/kinmap/">KinMap</a>. The figure has been generated in <strong>Talktorial T023</strong>.</p>
</section>
<section id="Kinase-similarity-descriptor:-Ligand-profile">
<h3 id="Kinase-similarity-descriptor:-Ligand-profile">Kinase similarity descriptor: Ligand profile<a class="headerlink" href="#Kinase-similarity-descriptor:-Ligand-profile" title="Permalink to this heading">¶</a></h3>
<p>As a measure of similarity, we use ligand profiling data in this talktorial.</p>
<section id="Kinase-similarity">
<h4 id="Kinase-similarity">Kinase similarity<a class="headerlink" href="#Kinase-similarity" title="Permalink to this heading">¶</a></h4>
<p>We use the following metric as similarity between two kinases <span class="math notranslate nohighlight">\(K_i\)</span> and <span class="math notranslate nohighlight">\(K_j\)</span> :</p>
<div class="math notranslate nohighlight">
\[\text{similarity}(K_i, K_j) = \frac{\#\text{ of compounds that were tested active on both } K_i \text{ and } K_j}
{\#\text{ of compounds that were tested on both } K_i \text{ and } K_j}.\]</div>
<p>Assuming that only one compound was tested on two kinases, and that the compound was tested as active for one and inactive for the other, then the similarity between these two kinases would be zero.</p>
</section>
<section id="Kinase-promiscuity">
<h4 id="Kinase-promiscuity">Kinase promiscuity<a class="headerlink" href="#Kinase-promiscuity" title="Permalink to this heading">¶</a></h4>
<p>Computing the similarity between a kinase and itself may be interpreted as kinase promiscuity, where the similarity described above would therefore represent the fraction of active compounds over all tested compounds.</p>
</section>
</section>
<section id="From-similarity-matrix-to-distance-matrix">
<h3 id="From-similarity-matrix-to-distance-matrix">From similarity matrix to distance matrix<a class="headerlink" href="#From-similarity-matrix-to-distance-matrix" title="Permalink to this heading">¶</a></h3>
<p>As discussed in <strong>Talktorial T024</strong>, we convert the similarity matrix to a distance matrix.</p>
</section>
</section>
<section id="Practical">
<h2 id="Practical">Practical<a class="headerlink" href="#Practical" title="Permalink to this heading">¶</a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">pathlib</span> <span class="kn">import</span> <span class="n">Path</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="kn">import</span> <span class="n">Counter</span>

<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>

<span class="kn">from</span> <span class="nn">rdkit</span> <span class="kn">import</span> <span class="n">Chem</span>
<span class="kn">from</span> <span class="nn">rdkit.Chem</span> <span class="kn">import</span> <span class="n">Draw</span>
<span class="kn">from</span> <span class="nn">rdkit.Chem.Draw</span> <span class="kn">import</span> <span class="n">IPythonConsole</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">HERE</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">_dh</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
<span class="n">DATA</span> <span class="o">=</span> <span class="n">HERE</span> <span class="o">/</span> <span class="s2">"data"</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">configs</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">HERE</span> <span class="o">/</span> <span class="s2">"../T023_what_is_a_kinase/data/pipeline_configs.csv"</span><span class="p">)</span>
<span class="n">configs</span> <span class="o">=</span> <span class="n">configs</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">"variable"</span><span class="p">)[</span><span class="s2">"default_value"</span><span class="p">]</span>
<span class="n">DEMO</span> <span class="o">=</span> <span class="nb">bool</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">configs</span><span class="p">[</span><span class="s2">"DEMO"</span><span class="p">]))</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">"Run in demo mode: </span><span class="si">{</span><span class="n">DEMO</span><span class="si">}</span><span class="s2">"</span><span class="p">)</span>
<span class="c1"># NBVAL_CHECK_OUTPUT</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Run in demo mode: True
</pre></div></div>
</div>
<section id="Define-the-kinases-of-interest">
<h3 id="Define-the-kinases-of-interest">Define the kinases of interest<a class="headerlink" href="#Define-the-kinases-of-interest" title="Permalink to this heading">¶</a></h3>
<p>Let’s load the kinase selection as defined in <strong>Talktorial T023</strong>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">kinase_selection_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
    <span class="n">HERE</span>
    <span class="o">/</span> <span class="s2">"../T023_what_is_a_kinase/</span><span class="se">\</span>
<span class="s2">data/kinase_selection.csv"</span>
<span class="p">)</span>
<span class="n">kinase_selection_df</span>
<span class="c1"># NBVAL_CHECK_OUTPUT</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
<thead>
<tr style="text-align: right;">
<th></th>
<th>kinase</th>
<th>kinase_klifs</th>
<th>uniprot_id</th>
<th>group</th>
<th>full_kinase_name</th>
</tr>
</thead>
<tbody>
<tr>
<th>0</th>
<td>EGFR</td>
<td>EGFR</td>
<td>P00533</td>
<td>TK</td>
<td>Epidermal growth factor receptor</td>
</tr>
<tr>
<th>1</th>
<td>ErbB2</td>
<td>ErbB2</td>
<td>P04626</td>
<td>TK</td>
<td>Erythroblastic leukemia viral oncogene homolog 2</td>
</tr>
<tr>
<th>2</th>
<td>PI3K</td>
<td>p110a</td>
<td>P42336</td>
<td>Atypical</td>
<td>Phosphatidylinositol-3-kinase</td>
</tr>
<tr>
<th>3</th>
<td>VEGFR2</td>
<td>KDR</td>
<td>P35968</td>
<td>TK</td>
<td>Vascular endothelial growth factor receptor 2</td>
</tr>
<tr>
<th>4</th>
<td>BRAF</td>
<td>BRAF</td>
<td>P15056</td>
<td>TKL</td>
<td>Rapidly accelerated fibrosarcoma isoform B</td>
</tr>
<tr>
<th>5</th>
<td>CDK2</td>
<td>CDK2</td>
<td>P24941</td>
<td>CMGC</td>
<td>Cyclic-dependent kinase 2</td>
</tr>
<tr>
<th>6</th>
<td>LCK</td>
<td>LCK</td>
<td>P06239</td>
<td>TK</td>
<td>Lymphocyte-specific protein tyrosine kinase</td>
</tr>
<tr>
<th>7</th>
<td>MET</td>
<td>MET</td>
<td>P08581</td>
<td>TK</td>
<td>Mesenchymal-epithelial transition factor</td>
</tr>
<tr>
<th>8</th>
<td>p38a</td>
<td>p38a</td>
<td>Q16539</td>
<td>CMGC</td>
<td>p38 mitogen activated protein kinase alpha</td>
</tr>
</tbody>
</table>
</div></div>
</div>
</section>
<section id="Retrieve-the-data">
<h3 id="Retrieve-the-data">Retrieve the data<a class="headerlink" href="#Retrieve-the-data" title="Permalink to this heading">¶</a></h3>
<p>We retrieve a pre-curated version of a kinase subset of ChEMBL29 freely available at Openkinome, see <a class="reference external" href="https://github.com/openkinome/kinodata/releases/tag/v0.3">https://github.com/openkinome/kinodata/releases/tag/v0.3</a>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">path</span> <span class="o">=</span> <span class="s2">"https://github.com/openkinome/kinodata/releases/download/</span><span class="se">\</span>
<span class="s2">v0.3/activities-chembl29_v0.3.zip"</span>
<span class="c1"># Load data and reset index so that it starts from 0</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">"Current shape of data: </span><span class="si">{</span><span class="n">data</span><span class="o">.</span><span class="n">shape</span><span class="si">}</span><span class="s2">"</span><span class="p">)</span>
<span class="n">data</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
<span class="c1"># NBVAL_CHECK_OUTPUT</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Current shape of data: (190634, 16)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
<thead>
<tr style="text-align: right;">
<th></th>
<th>activities.activity_id</th>
<th>assays.chembl_id</th>
<th>target_dictionary.chembl_id</th>
<th>molecule_dictionary.chembl_id</th>
<th>molecule_dictionary.max_phase</th>
<th>activities.standard_type</th>
<th>activities.standard_value</th>
<th>activities.standard_units</th>
<th>compound_structures.canonical_smiles</th>
<th>compound_structures.standard_inchi</th>
<th>component_sequences.sequence</th>
<th>assays.confidence_score</th>
<th>docs.chembl_id</th>
<th>docs.year</th>
<th>docs.authors</th>
<th>UniprotID</th>
</tr>
</thead>
<tbody>
<tr>
<th>0</th>
<td>16291323</td>
<td>CHEMBL3705523</td>
<td>CHEMBL2973</td>
<td>CHEMBL3666724</td>
<td>0</td>
<td>pIC50</td>
<td>14.096910</td>
<td>nM</td>
<td>CCCC(=O)Nc1cccc(-c2nc(Nc3ccc4[nH]ncc4c3)c3cc(O...</td>
<td>InChI=1S/C31H33N7O3/c1-2-4-29(40)33-22-6-3-5-2...</td>
<td>MSRPPPTGKMPGAPETAPGDGAGASRQRKLEALIRDPRSPINVESL...</td>
<td>9</td>
<td>CHEMBL3639077</td>
<td>2014.0</td>
<td>NaN</td>
<td>O75116</td>
</tr>
<tr>
<th>1</th>
<td>16264754</td>
<td>CHEMBL3705523</td>
<td>CHEMBL2973</td>
<td>CHEMBL3666728</td>
<td>0</td>
<td>pIC50</td>
<td>14.000000</td>
<td>nM</td>
<td>CCCC(=O)Nc1cccc(-c2nc(Nc3ccc4[nH]ncc4c3)c3cc(O...</td>
<td>InChI=1S/C34H40N8O3/c1-5-7-32(43)36-24-9-6-8-2...</td>
<td>MSRPPPTGKMPGAPETAPGDGAGASRQRKLEALIRDPRSPINVESL...</td>
<td>9</td>
<td>CHEMBL3639077</td>
<td>2014.0</td>
<td>NaN</td>
<td>O75116</td>
</tr>
<tr>
<th>2</th>
<td>16306943</td>
<td>CHEMBL3705523</td>
<td>CHEMBL2973</td>
<td>CHEMBL1968705</td>
<td>0</td>
<td>pIC50</td>
<td>14.000000</td>
<td>nM</td>
<td>CCCC(=O)Nc1cccc(-c2nc(Nc3ccc4[nH]ncc4c3)c3cc(O...</td>
<td>InChI=1S/C31H33N7O2/c1-2-6-29(39)33-23-8-5-7-2...</td>
<td>MSRPPPTGKMPGAPETAPGDGAGASRQRKLEALIRDPRSPINVESL...</td>
<td>9</td>
<td>CHEMBL3639077</td>
<td>2014.0</td>
<td>NaN</td>
<td>O75116</td>
</tr>
<tr>
<th>3</th>
<td>16340050</td>
<td>CHEMBL3705523</td>
<td>CHEMBL2973</td>
<td>CHEMBL1997433</td>
<td>0</td>
<td>pIC50</td>
<td>13.958607</td>
<td>nM</td>
<td>CCCC(=O)Nc1cccc(-c2nc(Nc3ccc4[nH]ncc4c3)c3cc(O...</td>
<td>InChI=1S/C28H28N6O3/c1-3-5-26(35)30-20-7-4-6-1...</td>
<td>MSRPPPTGKMPGAPETAPGDGAGASRQRKLEALIRDPRSPINVESL...</td>
<td>9</td>
<td>CHEMBL3639077</td>
<td>2014.0</td>
<td>NaN</td>
<td>O75116</td>
</tr>
<tr>
<th>4</th>
<td>16287186</td>
<td>CHEMBL3705523</td>
<td>CHEMBL2973</td>
<td>CHEMBL3666721</td>
<td>0</td>
<td>pIC50</td>
<td>13.920819</td>
<td>nM</td>
<td>CCCC(=O)Nc1cccc(-c2nc(Nc3ccc4[nH]ncc4c3)c3cc(O...</td>
<td>InChI=1S/C32H35N7O2/c1-2-7-30(40)34-24-9-6-8-2...</td>
<td>MSRPPPTGKMPGAPETAPGDGAGASRQRKLEALIRDPRSPINVESL...</td>
<td>9</td>
<td>CHEMBL3639077</td>
<td>2014.0</td>
<td>NaN</td>
<td>O75116</td>
</tr>
</tbody>
</table>
</div></div>
</div>
</section>
<section id="Preprocess-the-data">
<h3 id="Preprocess-the-data">Preprocess the data<a class="headerlink" href="#Preprocess-the-data" title="Permalink to this heading">¶</a></h3>
<p>We look at the type of activity and the associated units.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span>
    <span class="sa">f</span><span class="s2">"Activities: </span><span class="si">{</span><span class="nb">sorted</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="s1">'activities.standard_type'</span><span class="p">]))</span><span class="si">}</span><span class="se">\n</span><span class="s2">"</span>
    <span class="sa">f</span><span class="s2">"Units: </span><span class="si">{</span><span class="nb">set</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="s1">'activities.standard_units'</span><span class="p">])</span><span class="si">}</span><span class="s2">"</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Activities: ['pIC50', 'pKd', 'pKi']
Units: {'nM'}
</pre></div></div>
</div>
<p>Let’s keep the entries which have pIC50 values only.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="n">data</span><span class="p">[</span><span class="s2">"activities.standard_type"</span><span class="p">]</span> <span class="o">==</span> <span class="s2">"pIC50"</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data</span><span class="o">.</span><span class="n">columns</span>
<span class="c1"># NBVAL_CHECK_OUTPUT</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Index(['activities.activity_id', 'assays.chembl_id',
       'target_dictionary.chembl_id', 'molecule_dictionary.chembl_id',
       'molecule_dictionary.max_phase', 'activities.standard_type',
       'activities.standard_value', 'activities.standard_units',
       'compound_structures.canonical_smiles',
       'compound_structures.standard_inchi', 'component_sequences.sequence',
       'assays.confidence_score', 'docs.chembl_id', 'docs.year',
       'docs.authors', 'UniprotID'],
      dtype='object')
</pre></div></div>
</div>
<p>The DataFrame contains many columns that won’t be necessary for the rest of the notebook which are therefore removed. Only relevant information is kept, namely the canonical SMILES of the compound (<code class="docutils literal notranslate"><span class="pre">compound_structures.canonical_smiles</span></code>), the measured activity (<code class="docutils literal notranslate"><span class="pre">activities.standard_value</span></code>) and the UniProt ID of the kinase (<code class="docutils literal notranslate"><span class="pre">UniprotID</span></code>). These columns are renamed for readability.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="p">[[</span><span class="s2">"compound_structures.canonical_smiles"</span><span class="p">,</span> <span class="s2">"activities.standard_value"</span><span class="p">,</span> <span class="s2">"UniprotID"</span><span class="p">]]</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span>
    <span class="n">columns</span><span class="o">=</span><span class="p">{</span>
        <span class="s2">"compound_structures.canonical_smiles"</span><span class="p">:</span> <span class="s2">"smiles"</span><span class="p">,</span>
        <span class="s2">"activities.standard_value"</span><span class="p">:</span> <span class="s2">"pIC50"</span><span class="p">,</span>
    <span class="p">}</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">"Current shape of data: </span><span class="si">{</span><span class="n">data</span><span class="o">.</span><span class="n">shape</span><span class="si">}</span><span class="s2">"</span><span class="p">)</span>
<span class="n">data</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
<span class="c1"># NBVAL_CHECK_OUTPUT</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Current shape of data: (160857, 3)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
<thead>
<tr style="text-align: right;">
<th></th>
<th>smiles</th>
<th>pIC50</th>
<th>UniprotID</th>
</tr>
</thead>
<tbody>
<tr>
<th>0</th>
<td>CCCC(=O)Nc1cccc(-c2nc(Nc3ccc4[nH]ncc4c3)c3cc(O...</td>
<td>14.096910</td>
<td>O75116</td>
</tr>
<tr>
<th>1</th>
<td>CCCC(=O)Nc1cccc(-c2nc(Nc3ccc4[nH]ncc4c3)c3cc(O...</td>
<td>14.000000</td>
<td>O75116</td>
</tr>
<tr>
<th>2</th>
<td>CCCC(=O)Nc1cccc(-c2nc(Nc3ccc4[nH]ncc4c3)c3cc(O...</td>
<td>14.000000</td>
<td>O75116</td>
</tr>
<tr>
<th>3</th>
<td>CCCC(=O)Nc1cccc(-c2nc(Nc3ccc4[nH]ncc4c3)c3cc(O...</td>
<td>13.958607</td>
<td>O75116</td>
</tr>
<tr>
<th>4</th>
<td>CCCC(=O)Nc1cccc(-c2nc(Nc3ccc4[nH]ncc4c3)c3cc(O...</td>
<td>13.920819</td>
<td>O75116</td>
</tr>
</tbody>
</table>
</div></div>
</div>
<p>NA values are dropped.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">"Current shape of data: </span><span class="si">{</span><span class="n">data</span><span class="o">.</span><span class="n">shape</span><span class="si">}</span><span class="s2">"</span><span class="p">)</span>
<span class="c1"># NBVAL_CHECK_OUTPUT</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Current shape of data: (160703, 3)
</pre></div></div>
</div>
<p>We only keep the data for the query kinases:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="n">data</span><span class="p">[</span><span class="s2">"UniprotID"</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">kinase_selection_df</span><span class="p">[</span><span class="s2">"uniprot_id"</span><span class="p">])]</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">"Current shape of data: </span><span class="si">{</span><span class="n">data</span><span class="o">.</span><span class="n">shape</span><span class="si">}</span><span class="s2">"</span><span class="p">)</span>
<span class="n">data</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
<span class="c1"># NBVAL_CHECK_OUTPUT</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Current shape of data: (33427, 3)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
<thead>
<tr style="text-align: right;">
<th></th>
<th>smiles</th>
<th>pIC50</th>
<th>UniprotID</th>
</tr>
</thead>
<tbody>
<tr>
<th>58</th>
<td>Brc1cccc(Nc2ncnc3cc4ccccc4cc23)c1</td>
<td>11.522879</td>
<td>P00533</td>
</tr>
<tr>
<th>98</th>
<td>CN(C)c1cc2c(Nc3cccc(Br)c3)ncnc2cn1</td>
<td>11.221849</td>
<td>P00533</td>
</tr>
<tr>
<th>99</th>
<td>CCOc1cc2ncnc(Nc3cccc(Br)c3)c2cc1OCC</td>
<td>11.221849</td>
<td>P00533</td>
</tr>
<tr>
<th>140</th>
<td>CNc1cc2c(Nc3cccc(Br)c3)ncnc2cn1</td>
<td>11.096910</td>
<td>P00533</td>
</tr>
<tr>
<th>141</th>
<td>Brc1cccc(Nc2ncnc3cc4[nH]cnc4cc23)c1</td>
<td>11.096910</td>
<td>P00533</td>
</tr>
</tbody>
</table>
</div></div>
</div>
<p>Let’s look at example data (which corresponds to the first row in the kinase selection DataFrame):</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">example_kinase</span> <span class="o">=</span> <span class="n">kinase_selection_df</span><span class="p">[</span><span class="s2">"kinase_klifs"</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
<span class="n">example_uniprot</span> <span class="o">=</span> <span class="n">kinase_selection_df</span><span class="p">[</span><span class="s2">"uniprot_id"</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>

<span class="n">example_data</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="n">data</span><span class="p">[</span><span class="s2">"UniprotID"</span><span class="p">]</span> <span class="o">==</span> <span class="n">example_uniprot</span><span class="p">]</span>

<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">"Example kinase: </span><span class="si">{</span><span class="n">example_kinase</span><span class="si">}</span><span class="s2">"</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Example kinase: EGFR
</pre></div></div>
</div>
<p>Some compounds have been tested several times against a target, as shown below.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">measured_compounds</span> <span class="o">=</span> <span class="n">Counter</span><span class="p">(</span><span class="n">example_data</span><span class="p">[</span><span class="s2">"smiles"</span><span class="p">])</span>
<span class="k">try</span><span class="p">:</span>
    <span class="n">top_measured_compounds</span> <span class="o">=</span> <span class="n">measured_compounds</span><span class="o">.</span><span class="n">most_common</span><span class="p">()[</span><span class="mi">0</span><span class="p">:</span><span class="mi">5</span><span class="p">]</span>
<span class="k">except</span> <span class="ne">IndexError</span><span class="p">:</span>
    <span class="n">top_measured_compounds</span> <span class="o">=</span> <span class="n">measured_compounds</span><span class="o">.</span><span class="n">most_common</span><span class="p">()</span>
<span class="n">top_measured_compounds</span>
<span class="c1"># NBVAL_CHECK_OUTPUT</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[('COc1cc2ncnc(Nc3ccc(F)c(Cl)c3)c2cc1OCCCN1CCOCC1', 39),
 ('C#Cc1cccc(Nc2ncnc3cc(OCCOC)c(OCCOC)cc23)c1', 27),
 ('C=CC(=O)Nc1cc(Nc2nccc(-c3cn(C)c4ccccc34)n2)c(OC)cc1N(C)CCN(C)C', 15),
 ('C=CC(=O)Nc1cccc(Oc2nc(Nc3ccc(N4CCN(C)CC4)cc3OC)ncc2Cl)c1', 11),
 ('CS(=O)(=O)CCNCc1ccc(-c2ccc3ncnc(Nc4ccc(OCc5cccc(F)c5)c(Cl)c4)c3c2)o1', 8)]
</pre></div></div>
</div>
<p>We have a look at those compounds.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mols</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">entry</span> <span class="ow">in</span> <span class="n">top_measured_compounds</span><span class="p">:</span>
    <span class="n">mols</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Chem</span><span class="o">.</span><span class="n">MolFromSmiles</span><span class="p">(</span><span class="n">entry</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span>
<span class="n">Draw</span><span class="o">.</span><span class="n">MolsToGridImage</span><span class="p">(</span><span class="n">mols</span><span class="p">,</span> <span class="n">molsPerRow</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="output_area docutils container">
<img alt="../_images/talktorials_T027_kinase_similarity_ligand_profile_44_0.png" src="../_images/talktorials_T027_kinase_similarity_ligand_profile_44_0.png"/>
</div>
</div>
<p>In this example (demo mode), the first molecule is gefitinib, a known FDA-approved drug against EGFR.</p>
<p>As a simple workaround — since we prefer to have one activity value per compound-kinase pair — we keep the value for which the compound has the best activity value, i.e., the highest pIC50 value.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s2">"UniprotID"</span><span class="p">,</span> <span class="s2">"smiles"</span><span class="p">])[</span><span class="s2">"pIC50"</span><span class="p">]</span><span class="o">.</span><span class="n">max</span><span class="p">()</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
<span class="n">data</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
<span class="c1"># NBVAL_CHECK_OUTPUT</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
<thead>
<tr style="text-align: right;">
<th></th>
<th>UniprotID</th>
<th>smiles</th>
<th>pIC50</th>
</tr>
</thead>
<tbody>
<tr>
<th>0</th>
<td>P00533</td>
<td>Br.CC(Nc1ncnc2[nH]c(-c3ccc(O)cc3)cc12)c1ccc(C(...</td>
<td>5.336488</td>
</tr>
<tr>
<th>1</th>
<td>P00533</td>
<td>Br.CC(Nc1ncnc2[nH]c(-c3ccc(O)cc3)cc12)c1cccc2c...</td>
<td>5.996539</td>
</tr>
<tr>
<th>2</th>
<td>P00533</td>
<td>Br.CC[C@@H](Nc1ncnc2[nH]c(-c3ccc(O)cc3)cc12)c1...</td>
<td>8.397940</td>
</tr>
<tr>
<th>3</th>
<td>P00533</td>
<td>Br.C[C@@H](Nc1ncnc2[nH]c(-c3ccc(O)cc3)cc12)c1c...</td>
<td>7.207608</td>
</tr>
<tr>
<th>4</th>
<td>P00533</td>
<td>Br.C[C@@H](Nc1ncnc2[nH]c(-c3ccc(O)cc3)cc12)c1c...</td>
<td>8.420216</td>
</tr>
</tbody>
</table>
</div></div>
</div>
<section id="Hit-or-non-hit">
<h4 id="Hit-or-non-hit">Hit or non-hit<a class="headerlink" href="#Hit-or-non-hit" title="Permalink to this heading">¶</a></h4>
<p>Finally, we binarize the pIC50 values to obtain hit or non-hit using a cutoff. We use a pIC50 cutoff of <span class="math notranslate nohighlight">\(6.3\)</span>, similarly to the cutoff used in <a class="reference external" href="https://www.mdpi.com/1420-3049/26/3/629">Molecules (2021), 26(3), 629</a>.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cutoff</span> <span class="o">=</span> <span class="mf">6.3</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">binarize_pic50</span><span class="p">(</span><span class="n">pic50_value</span><span class="p">,</span> <span class="n">threshold</span><span class="p">):</span>
    <span class="sd">"""</span>
<span class="sd">    Binarizes a scalar value given a threshold.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    pic50_value : float</span>
<span class="sd">        The measurement pIC50 value of a kinase-ligand pair.</span>
<span class="sd">    threshold : float</span>
<span class="sd">        The cutoff to determine activity.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    int</span>
<span class="sd">        1 if the pIC50 value is above the threshold, which indicates activity.</span>
<span class="sd">        0 otherwise.</span>
<span class="sd">    """</span>
    <span class="k">if</span> <span class="n">pic50_value</span> <span class="o">&gt;=</span> <span class="n">threshold</span><span class="p">:</span>
        <span class="k">return</span> <span class="mi">1</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="mi">0</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data</span><span class="p">[</span><span class="s2">"activity_binary"</span><span class="p">]</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">"pIC50"</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">binarize_pic50</span><span class="p">,</span> <span class="n">args</span><span class="o">=</span><span class="p">(</span><span class="n">cutoff</span><span class="p">,))</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">"Current shape of data: </span><span class="si">{</span><span class="n">data</span><span class="o">.</span><span class="n">shape</span><span class="si">}</span><span class="s2">"</span><span class="p">)</span>
<span class="n">data</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
<span class="c1"># NBVAL_CHECK_OUTPUT</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Current shape of data: (32916, 4)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
<thead>
<tr style="text-align: right;">
<th></th>
<th>UniprotID</th>
<th>smiles</th>
<th>pIC50</th>
<th>activity_binary</th>
</tr>
</thead>
<tbody>
<tr>
<th>0</th>
<td>P00533</td>
<td>Br.CC(Nc1ncnc2[nH]c(-c3ccc(O)cc3)cc12)c1ccc(C(...</td>
<td>5.336488</td>
<td>0</td>
</tr>
<tr>
<th>1</th>
<td>P00533</td>
<td>Br.CC(Nc1ncnc2[nH]c(-c3ccc(O)cc3)cc12)c1cccc2c...</td>
<td>5.996539</td>
<td>0</td>
</tr>
<tr>
<th>2</th>
<td>P00533</td>
<td>Br.CC[C@@H](Nc1ncnc2[nH]c(-c3ccc(O)cc3)cc12)c1...</td>
<td>8.397940</td>
<td>1</td>
</tr>
<tr>
<th>3</th>
<td>P00533</td>
<td>Br.C[C@@H](Nc1ncnc2[nH]c(-c3ccc(O)cc3)cc12)c1c...</td>
<td>7.207608</td>
<td>1</td>
</tr>
<tr>
<th>4</th>
<td>P00533</td>
<td>Br.C[C@@H](Nc1ncnc2[nH]c(-c3ccc(O)cc3)cc12)c1c...</td>
<td>8.420216</td>
<td>1</td>
</tr>
</tbody>
</table>
</div></div>
</div>
</section>
</section>
<section id="id1">
<h3 id="id1">Kinase promiscuity<a class="headerlink" href="#id1" title="Permalink to this heading">¶</a></h3>
<p>We now look at the kinase promiscuity.</p>
<p>For a given kinase, three values are computed:</p>
<ol class="arabic simple">
<li><p>the total number of measured compounds against the given kinase,</p></li>
<li><p>the number of active compounds against the kinase, and</p></li>
<li><p>the fraction of active compounds, i.e., the ratio of active compounds over the total number of measured compounds per kinase.</p></li>
</ol>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">kinase_to_activity_numbers</span><span class="p">(</span><span class="n">uniprot_id</span><span class="p">,</span> <span class="n">activity_df</span><span class="p">):</span>
    <span class="sd">"""</span>
<span class="sd">    Retrieve the three values for a given kinase.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    uniprot_id : str</span>
<span class="sd">        The UniProt ID of the kinase of interest, e.g. "P00533" for "EGFR".</span>
<span class="sd">    activity_df : pd.DataFrame</span>
<span class="sd">        The dataframe with activity values for kinases.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    tuple : (int, int, float)</span>
<span class="sd">        The three metrics:</span>
<span class="sd">        1. The total number of measured compounds against the kinase.</span>
<span class="sd">        2. The number of active compounds against the kinase.</span>
<span class="sd">        3. The fraction of active compounds against the kinase.</span>
<span class="sd">    """</span>
    <span class="n">kinase_data</span> <span class="o">=</span> <span class="n">activity_df</span><span class="p">[</span><span class="n">activity_df</span><span class="p">[</span><span class="s2">"UniprotID"</span><span class="p">]</span> <span class="o">==</span> <span class="n">uniprot_id</span><span class="p">]</span>
    <span class="n">total_measured_compounds</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">kinase_data</span><span class="p">)</span>
    <span class="n">active_compounds</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">kinase_data</span><span class="p">[</span><span class="n">kinase_data</span><span class="p">[</span><span class="s2">"activity_binary"</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span><span class="p">])</span>
    <span class="k">if</span> <span class="n">total_measured_compounds</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">fraction</span> <span class="o">=</span> <span class="n">active_compounds</span> <span class="o">/</span> <span class="n">total_measured_compounds</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">"No compounds were measured for this kinase."</span><span class="p">)</span>
        <span class="n">fraction</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span>
    <span class="k">return</span> <span class="p">(</span><span class="n">total_measured_compounds</span><span class="p">,</span> <span class="n">active_compounds</span><span class="p">,</span> <span class="n">fraction</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Let’s see what information we get for the first kinase in our dataset:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">example_kinase</span> <span class="o">=</span> <span class="n">kinase_selection_df</span><span class="p">[</span><span class="s2">"kinase_klifs"</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
<span class="n">example_uniprot</span> <span class="o">=</span> <span class="n">kinase_selection_df</span><span class="p">[</span><span class="s2">"uniprot_id"</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>

<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">"</span><span class="si">{</span><span class="n">example_kinase</span><span class="si">}</span><span class="s2"> (</span><span class="si">{</span><span class="n">example_uniprot</span><span class="si">}</span><span class="s2">):"</span><span class="p">)</span>
<span class="n">example_metrics</span> <span class="o">=</span> <span class="n">kinase_to_activity_numbers</span><span class="p">(</span><span class="n">example_uniprot</span><span class="p">,</span> <span class="n">data</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span>
    <span class="sa">f</span><span class="s2">"</span><span class="si">{</span><span class="s1">'Total number of measured compounds:'</span> <span class="si">:</span><span class="s2"> &lt;40</span><span class="si">}</span><span class="s2">"</span>
    <span class="sa">f</span><span class="s2">"</span><span class="si">{</span><span class="n">example_metrics</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2"> </span><span class="se">\n</span><span class="s2">"</span>
    <span class="sa">f</span><span class="s2">"</span><span class="si">{</span><span class="s1">'Number of active compounds:'</span> <span class="si">:</span><span class="s2"> &lt;40</span><span class="si">}</span><span class="s2">"</span>
    <span class="sa">f</span><span class="s2">"</span><span class="si">{</span><span class="n">example_metrics</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s2"> </span><span class="se">\n</span><span class="s2">"</span>
    <span class="sa">f</span><span class="s2">"</span><span class="si">{</span><span class="s1">'Fraction of active compounds:'</span> <span class="si">:</span><span class="s2"> &lt;40</span><span class="si">}</span><span class="s2">"</span>
    <span class="sa">f</span><span class="s2">"</span><span class="si">{</span><span class="n">example_metrics</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2"> </span><span class="se">\n</span><span class="s2">"</span>
<span class="p">)</span>
<span class="c1"># NBVAL_CHECK_OUTPUT</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
EGFR (P00533):
Total number of measured compounds:     5965
Number of active compounds:             3635
Fraction of active compounds:           0.61

</pre></div></div>
</div>
<p>Let’s create a table from these values for all kinases:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">promiscuity_table</span><span class="p">(</span><span class="n">kinase_selection</span><span class="p">,</span> <span class="n">activity_df</span><span class="p">):</span>
    <span class="sd">"""</span>
<span class="sd">    Create a table with all three values for all kinases.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    kinase_selection : pd.DataFrame</span>
<span class="sd">        The DataFrame for the chosen kinases.</span>
<span class="sd">    activity_df : pd.DataFrame</span>
<span class="sd">        The DataFrame with activity values for kinases.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    promiscuity_table : pd.DataFrame</span>
<span class="sd">        A DataFrame with the kinases as rows and values as columns.</span>
<span class="sd">    """</span>
    <span class="n">promiscuity_table</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
        <span class="n">index</span><span class="o">=</span><span class="n">kinase_selection</span><span class="p">[</span><span class="s2">"kinase_klifs"</span><span class="p">],</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">"total"</span><span class="p">,</span> <span class="s2">"actives"</span><span class="p">,</span> <span class="s2">"fraction"</span><span class="p">]</span>
    <span class="p">)</span>
    <span class="n">promiscuity_table</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">promiscuity_table</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">uniprot_id</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">kinase_selection</span><span class="p">[</span><span class="s2">"kinase_klifs"</span><span class="p">],</span> <span class="n">kinase_selection</span><span class="p">[</span><span class="s2">"uniprot_id"</span><span class="p">]):</span>
        <span class="n">values</span> <span class="o">=</span> <span class="n">kinase_to_activity_numbers</span><span class="p">(</span><span class="n">uniprot_id</span><span class="p">,</span> <span class="n">activity_df</span><span class="p">)</span>
        <span class="n">promiscuity_table</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">values</span>
    <span class="k">return</span> <span class="n">promiscuity_table</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">kinase_promiscuity_df</span> <span class="o">=</span> <span class="n">promiscuity_table</span><span class="p">(</span><span class="n">kinase_selection_df</span><span class="p">,</span> <span class="n">data</span><span class="p">)</span>
<span class="n">kinase_promiscuity_df</span>
<span class="c1"># NBVAL_CHECK_OUTPUT</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
<thead>
<tr style="text-align: right;">
<th></th>
<th>total</th>
<th>actives</th>
<th>fraction</th>
</tr>
</thead>
<tbody>
<tr>
<th>EGFR</th>
<td>5965</td>
<td>3635</td>
<td>0.609388</td>
</tr>
<tr>
<th>ErbB2</th>
<td>1700</td>
<td>1031</td>
<td>0.606471</td>
</tr>
<tr>
<th>p110a</th>
<td>4393</td>
<td>2827</td>
<td>0.643524</td>
</tr>
<tr>
<th>KDR</th>
<td>7641</td>
<td>5328</td>
<td>0.697291</td>
</tr>
<tr>
<th>BRAF</th>
<td>3688</td>
<td>2992</td>
<td>0.81128</td>
</tr>
<tr>
<th>CDK2</th>
<td>1500</td>
<td>815</td>
<td>0.543333</td>
</tr>
<tr>
<th>LCK</th>
<td>1560</td>
<td>935</td>
<td>0.599359</td>
</tr>
<tr>
<th>MET</th>
<td>2832</td>
<td>2248</td>
<td>0.793785</td>
</tr>
<tr>
<th>p38a</th>
<td>3637</td>
<td>2778</td>
<td>0.763816</td>
</tr>
</tbody>
</table>
</div></div>
</div>
<p>Let’s beautify the table:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">kinase_promiscuity_df</span><span class="o">.</span><span class="n">style</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s2">"</span><span class="si">{:.3f}</span><span class="s2">"</span><span class="p">,</span> <span class="n">subset</span><span class="o">=</span><span class="p">[</span><span class="s2">"fraction"</span><span class="p">])</span><span class="o">.</span><span class="n">background_gradient</span><span class="p">(</span>
    <span class="n">cmap</span><span class="o">=</span><span class="s2">"Purples"</span><span class="p">,</span> <span class="n">subset</span><span class="o">=</span><span class="p">[</span><span class="s2">"fraction"</span><span class="p">]</span>
<span class="p">)</span><span class="o">.</span><span class="n">highlight_min</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="s2">"yellow"</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">subset</span><span class="o">=</span><span class="p">[</span><span class="s2">"fraction"</span><span class="p">])</span><span class="o">.</span><span class="n">highlight_max</span><span class="p">(</span>
    <span class="n">color</span><span class="o">=</span><span class="s2">"red"</span><span class="p">,</span> <span class="n">subset</span><span class="o">=</span><span class="p">[</span><span class="s2">"fraction"</span><span class="p">]</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<style type="text/css">
#T_e1a92_row0_col2 {
  background-color: #dadaeb;
  color: #000000;
}
#T_e1a92_row1_col2 {
  background-color: #dcdcec;
  color: #000000;
}
#T_e1a92_row2_col2 {
  background-color: #bdbedc;
  color: #000000;
}
#T_e1a92_row3_col2 {
  background-color: #8c88bf;
  color: #f1f1f1;
}
#T_e1a92_row4_col2 {
  background-color: #3f007d;
  color: #f1f1f1;
  background-color: red;
}
#T_e1a92_row5_col2 {
  background-color: #fcfbfd;
  color: #000000;
  background-color: yellow;
}
#T_e1a92_row6_col2 {
  background-color: #e1e0ee;
  color: #000000;
}
#T_e1a92_row7_col2 {
  background-color: #4a1486;
  color: #f1f1f1;
}
#T_e1a92_row8_col2 {
  background-color: #5d3897;
  color: #f1f1f1;
}
</style>
<table id="T_e1a92_">
<thead>
<tr>
<th class="blank level0"> </th>
<th class="col_heading level0 col0">total</th>
<th class="col_heading level0 col1">actives</th>
<th class="col_heading level0 col2">fraction</th>
</tr>
</thead>
<tbody>
<tr>
<th class="row_heading level0 row0" id="T_e1a92_level0_row0">EGFR</th>
<td class="data row0 col0" id="T_e1a92_row0_col0">5965</td>
<td class="data row0 col1" id="T_e1a92_row0_col1">3635</td>
<td class="data row0 col2" id="T_e1a92_row0_col2">0.609</td>
</tr>
<tr>
<th class="row_heading level0 row1" id="T_e1a92_level0_row1">ErbB2</th>
<td class="data row1 col0" id="T_e1a92_row1_col0">1700</td>
<td class="data row1 col1" id="T_e1a92_row1_col1">1031</td>
<td class="data row1 col2" id="T_e1a92_row1_col2">0.606</td>
</tr>
<tr>
<th class="row_heading level0 row2" id="T_e1a92_level0_row2">p110a</th>
<td class="data row2 col0" id="T_e1a92_row2_col0">4393</td>
<td class="data row2 col1" id="T_e1a92_row2_col1">2827</td>
<td class="data row2 col2" id="T_e1a92_row2_col2">0.644</td>
</tr>
<tr>
<th class="row_heading level0 row3" id="T_e1a92_level0_row3">KDR</th>
<td class="data row3 col0" id="T_e1a92_row3_col0">7641</td>
<td class="data row3 col1" id="T_e1a92_row3_col1">5328</td>
<td class="data row3 col2" id="T_e1a92_row3_col2">0.697</td>
</tr>
<tr>
<th class="row_heading level0 row4" id="T_e1a92_level0_row4">BRAF</th>
<td class="data row4 col0" id="T_e1a92_row4_col0">3688</td>
<td class="data row4 col1" id="T_e1a92_row4_col1">2992</td>
<td class="data row4 col2" id="T_e1a92_row4_col2">0.811</td>
</tr>
<tr>
<th class="row_heading level0 row5" id="T_e1a92_level0_row5">CDK2</th>
<td class="data row5 col0" id="T_e1a92_row5_col0">1500</td>
<td class="data row5 col1" id="T_e1a92_row5_col1">815</td>
<td class="data row5 col2" id="T_e1a92_row5_col2">0.543</td>
</tr>
<tr>
<th class="row_heading level0 row6" id="T_e1a92_level0_row6">LCK</th>
<td class="data row6 col0" id="T_e1a92_row6_col0">1560</td>
<td class="data row6 col1" id="T_e1a92_row6_col1">935</td>
<td class="data row6 col2" id="T_e1a92_row6_col2">0.599</td>
</tr>
<tr>
<th class="row_heading level0 row7" id="T_e1a92_level0_row7">MET</th>
<td class="data row7 col0" id="T_e1a92_row7_col0">2832</td>
<td class="data row7 col1" id="T_e1a92_row7_col1">2248</td>
<td class="data row7 col2" id="T_e1a92_row7_col2">0.794</td>
</tr>
<tr>
<th class="row_heading level0 row8" id="T_e1a92_level0_row8">p38a</th>
<td class="data row8 col0" id="T_e1a92_row8_col0">3637</td>
<td class="data row8 col1" id="T_e1a92_row8_col1">2778</td>
<td class="data row8 col2" id="T_e1a92_row8_col2">0.764</td>
</tr>
</tbody>
</table></div>
</div>
<p>From the table, we notice that CDK2 is the least (in yellow) and BRAF the most (in red) promiscuous kinase.</p>
</section>
<section id="id2">
<h3 id="id2">Kinase similarity<a class="headerlink" href="#id2" title="Permalink to this heading">¶</a></h3>
<p>We now investigate how we can use the similarity measure discussed in the <em>Theory</em> part to compare kinases.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">similarity_ligand_profile</span><span class="p">(</span><span class="n">uniprot_id1</span><span class="p">,</span> <span class="n">uniprot_id2</span><span class="p">,</span> <span class="n">activity_df</span><span class="p">):</span>
    <span class="sd">"""</span>
<span class="sd">    Compute the similarity between two kinases using ligand profiling data.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    uniprot_id1 : str</span>
<span class="sd">        UniProt ID of first kinase of interest.</span>
<span class="sd">    uniprot_id2 : str</span>
<span class="sd">        UniProt ID of second kinase of interest.</span>
<span class="sd">    activity_df :  pd.DataFrame</span>
<span class="sd">        The DataFrame with activity values for kinases.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    tuple : (int, int, float)</span>
<span class="sd">        The three metrics:</span>
<span class="sd">        1. The total number of measured compounds against both kinases.</span>
<span class="sd">        2. The number of active compounds against both kinases.</span>
<span class="sd">        3. The metric for kinase similarity,</span>
<span class="sd">                i.e. number of active compounds on both kinases</span>
<span class="sd">                over number of measured compounds on both kinases.</span>
<span class="sd">    """</span>
    <span class="k">if</span> <span class="n">uniprot_id1</span> <span class="o">==</span> <span class="n">uniprot_id2</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">kinase_to_activity_numbers</span><span class="p">(</span><span class="n">uniprot_id1</span><span class="p">,</span> <span class="n">activity_df</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># Data for the two kinases only</span>
        <span class="n">reduced_data</span> <span class="o">=</span> <span class="n">activity_df</span><span class="p">[</span><span class="n">activity_df</span><span class="p">[</span><span class="s2">"UniprotID"</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">([</span><span class="n">uniprot_id1</span><span class="p">,</span> <span class="n">uniprot_id2</span><span class="p">])]</span>

        <span class="c1"># Look at active compounds only</span>
        <span class="n">active_entries</span> <span class="o">=</span> <span class="n">reduced_data</span><span class="p">[</span><span class="n">reduced_data</span><span class="p">[</span><span class="s2">"activity_binary"</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span><span class="p">]</span>
        <span class="c1"># Group by compounds</span>
        <span class="n">compounds</span> <span class="o">=</span> <span class="n">active_entries</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">"smiles"</span><span class="p">)</span><span class="o">.</span><span class="n">size</span><span class="p">()</span>
        <span class="c1"># Look at the number of active compounds measured on both kinases</span>
        <span class="n">active_compounds_on_both</span> <span class="o">=</span> <span class="n">compounds</span><span class="p">[</span><span class="n">compounds</span> <span class="o">==</span> <span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

        <span class="c1"># Look at all tested compounds</span>
        <span class="n">compounds</span> <span class="o">=</span> <span class="n">reduced_data</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">"smiles"</span><span class="p">)</span><span class="o">.</span><span class="n">size</span><span class="p">()</span>
        <span class="c1"># Look at the number of compounds measured on both kinases</span>
        <span class="n">measured_compounds_on_both</span> <span class="o">=</span> <span class="n">compounds</span><span class="p">[</span><span class="n">compounds</span> <span class="o">==</span> <span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

        <span class="k">if</span> <span class="n">measured_compounds_on_both</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">fraction</span> <span class="o">=</span> <span class="n">active_compounds_on_both</span> <span class="o">/</span> <span class="n">measured_compounds_on_both</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">"No compounds were measured on both kinases, "</span>
                <span class="sa">f</span><span class="s2">"namely </span><span class="si">{</span><span class="n">uniprot_id1</span><span class="si">}</span><span class="s2"> and </span><span class="si">{</span><span class="n">uniprot_id2</span><span class="si">}</span><span class="s2">."</span>
            <span class="p">)</span>
            <span class="n">fraction</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span>
            <span class="n">measured_compounds_on_both</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span>
            <span class="n">active_compounds_on_both</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span>
        <span class="k">return</span> <span class="p">(</span><span class="n">measured_compounds_on_both</span><span class="p">,</span> <span class="n">active_compounds_on_both</span><span class="p">,</span> <span class="n">fraction</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Let’s look at the values and similarity between two kinases.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">if</span> <span class="n">DEMO</span><span class="p">:</span>
    <span class="n">kinase1</span> <span class="o">=</span> <span class="s2">"EGFR"</span>
    <span class="n">uniprot1</span> <span class="o">=</span> <span class="s2">"P00533"</span>
    <span class="n">kinase2</span> <span class="o">=</span> <span class="s2">"MET"</span>
    <span class="n">uniprot2</span> <span class="o">=</span> <span class="s2">"P08581"</span>
<span class="k">else</span><span class="p">:</span>
    <span class="n">kinase1</span> <span class="o">=</span> <span class="n">kinase_selection_df</span><span class="p">[</span><span class="s2">"kinase_klifs"</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">uniprot1</span> <span class="o">=</span> <span class="n">kinase_selection_df</span><span class="p">[</span><span class="s2">"uniprot_id"</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">kinase2</span> <span class="o">=</span> <span class="n">kinase_selection_df</span><span class="p">[</span><span class="s2">"kinase_klifs"</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">uniprot2</span> <span class="o">=</span> <span class="n">kinase_selection_df</span><span class="p">[</span><span class="s2">"uniprot_id"</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span>

<span class="n">similarity_example</span> <span class="o">=</span> <span class="n">similarity_ligand_profile</span><span class="p">(</span><span class="n">uniprot1</span><span class="p">,</span> <span class="n">uniprot2</span><span class="p">,</span> <span class="n">data</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span>
    <span class="sa">f</span><span class="s2">"Values for </span><span class="si">{</span><span class="n">kinase1</span><span class="si">}</span><span class="s2"> and </span><span class="si">{</span><span class="n">kinase2</span><span class="si">}</span><span class="s2">: </span><span class="se">\n\n</span><span class="s2">"</span>
    <span class="sa">f</span><span class="s2">"</span><span class="si">{</span><span class="s1">'Total number of measured compounds:'</span> <span class="si">:</span><span class="s2"> &lt;50</span><span class="si">}</span><span class="s2">"</span>
    <span class="sa">f</span><span class="s2">"</span><span class="si">{</span><span class="n">similarity_example</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2"> </span><span class="se">\n</span><span class="s2">"</span>
    <span class="sa">f</span><span class="s2">"</span><span class="si">{</span><span class="s1">'Number of active compounds:'</span> <span class="si">:</span><span class="s2"> &lt;50</span><span class="si">}</span><span class="s2">"</span>
    <span class="sa">f</span><span class="s2">"</span><span class="si">{</span><span class="n">similarity_example</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s2"> </span><span class="se">\n</span><span class="s2">"</span>
    <span class="sa">f</span><span class="s2">"Fraction of active compounds or </span><span class="se">\n</span><span class="s2">"</span>
    <span class="sa">f</span><span class="s2">"</span><span class="si">{</span><span class="s1">'ligand profile similarity:'</span> <span class="si">:</span><span class="s2"> &lt;50</span><span class="si">}</span><span class="s2">"</span>
    <span class="sa">f</span><span class="s2">"</span><span class="si">{</span><span class="n">similarity_example</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2"> </span><span class="se">\n</span><span class="s2">"</span>
<span class="p">)</span>
<span class="c1"># NBVAL_CHECK_OUTPUT</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Values for EGFR and MET:

Total number of measured compounds:               92
Number of active compounds:                       21
Fraction of active compounds or
ligand profile similarity:                        0.23

</pre></div></div>
</div>
<section id="Visualize-similarity-as-kinase-matrix">
<h4 id="Visualize-similarity-as-kinase-matrix">Visualize similarity as kinase matrix<a class="headerlink" href="#Visualize-similarity-as-kinase-matrix" title="Permalink to this heading">¶</a></h4>
<p>Let’s first look at the non-reduced fraction of number of active compound against total number of compounds to have an idea of the counts.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">kinase_counts_matrix</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
    <span class="n">index</span><span class="o">=</span><span class="n">kinase_selection_df</span><span class="o">.</span><span class="n">kinase_klifs</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">kinase_selection_df</span><span class="o">.</span><span class="n">kinase_klifs</span>
<span class="p">)</span>
<span class="n">kinase_counts_matrix</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="kc">None</span>
<span class="n">kinase_counts_matrix</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="kc">None</span>

<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">(</span><span class="n">uniprot_id1</span><span class="p">,</span> <span class="n">klifs_name1</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span>
    <span class="nb">zip</span><span class="p">(</span><span class="n">kinase_selection_df</span><span class="o">.</span><span class="n">uniprot_id</span><span class="p">,</span> <span class="n">kinase_selection_df</span><span class="o">.</span><span class="n">kinase_klifs</span><span class="p">)</span>
<span class="p">):</span>
    <span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="p">(</span><span class="n">uniprot_id2</span><span class="p">,</span> <span class="n">klifs_name2</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span>
        <span class="nb">zip</span><span class="p">(</span><span class="n">kinase_selection_df</span><span class="o">.</span><span class="n">uniprot_id</span><span class="p">,</span> <span class="n">kinase_selection_df</span><span class="o">.</span><span class="n">kinase_klifs</span><span class="p">)</span>
    <span class="p">):</span>
        <span class="n">total</span><span class="p">,</span> <span class="n">actives</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">similarity_ligand_profile</span><span class="p">(</span><span class="n">uniprot_id1</span><span class="p">,</span> <span class="n">uniprot_id2</span><span class="p">,</span> <span class="n">data</span><span class="p">)</span>
        <span class="n">integer_ratio</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">"</span><span class="si">{</span><span class="n">actives</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">total</span><span class="si">}</span><span class="s2">"</span>
        <span class="n">kinase_counts_matrix</span><span class="p">[</span><span class="n">klifs_name1</span><span class="p">][</span><span class="n">klifs_name2</span><span class="p">]</span> <span class="o">=</span> <span class="n">integer_ratio</span>
<span class="n">kinase_counts_matrix</span>
<span class="c1"># NBVAL_CHECK_OUTPUT</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
No compounds were measured on both kinases, namely P04626 and P42336.
No compounds were measured on both kinases, namely P42336 and P04626.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
<thead>
<tr style="text-align: right;">
<th></th>
<th>EGFR</th>
<th>ErbB2</th>
<th>p110a</th>
<th>KDR</th>
<th>BRAF</th>
<th>CDK2</th>
<th>LCK</th>
<th>MET</th>
<th>p38a</th>
</tr>
</thead>
<tbody>
<tr>
<th>EGFR</th>
<td>3635/5965</td>
<td>662/1170</td>
<td>13/179</td>
<td>313/893</td>
<td>27/59</td>
<td>5/40</td>
<td>31/126</td>
<td>21/92</td>
<td>18/52</td>
</tr>
<tr>
<th>ErbB2</th>
<td>662/1170</td>
<td>1031/1700</td>
<td>nan/nan</td>
<td>72/180</td>
<td>4/16</td>
<td>4/27</td>
<td>5/33</td>
<td>1/28</td>
<td>2/16</td>
</tr>
<tr>
<th>p110a</th>
<td>13/179</td>
<td>nan/nan</td>
<td>2827/4393</td>
<td>32/174</td>
<td>1/3</td>
<td>4/12</td>
<td>0/3</td>
<td>0/1</td>
<td>0/5</td>
</tr>
<tr>
<th>KDR</th>
<td>313/893</td>
<td>72/180</td>
<td>32/174</td>
<td>5328/7641</td>
<td>199/262</td>
<td>71/115</td>
<td>179/413</td>
<td>184/340</td>
<td>63/122</td>
</tr>
<tr>
<th>BRAF</th>
<td>27/59</td>
<td>4/16</td>
<td>1/3</td>
<td>199/262</td>
<td>2992/3688</td>
<td>1/13</td>
<td>22/40</td>
<td>3/26</td>
<td>29/41</td>
</tr>
<tr>
<th>CDK2</th>
<td>5/40</td>
<td>4/27</td>
<td>4/12</td>
<td>71/115</td>
<td>1/13</td>
<td>815/1500</td>
<td>2/18</td>
<td>2/22</td>
<td>1/9</td>
</tr>
<tr>
<th>LCK</th>
<td>31/126</td>
<td>5/33</td>
<td>0/3</td>
<td>179/413</td>
<td>22/40</td>
<td>2/18</td>
<td>935/1560</td>
<td>17/63</td>
<td>69/138</td>
</tr>
<tr>
<th>MET</th>
<td>21/92</td>
<td>1/28</td>
<td>0/1</td>
<td>184/340</td>
<td>3/26</td>
<td>2/22</td>
<td>17/63</td>
<td>2248/2832</td>
<td>1/20</td>
</tr>
<tr>
<th>p38a</th>
<td>18/52</td>
<td>2/16</td>
<td>0/5</td>
<td>63/122</td>
<td>29/41</td>
<td>1/9</td>
<td>69/138</td>
<td>1/20</td>
<td>2778/3637</td>
</tr>
</tbody>
</table>
</div></div>
</div>
<p>Note that the total number of tested compounds as well as the number of active compounds on two kinases vary largely.</p>
<ul class="simple">
<li><p>For the p110a-ErbB2 pair, there are none.</p></li>
<li><p>For p110a and [BRAF, CDK2, LCK, MET and p38a], there are less than <span class="math notranslate nohighlight">\(13\)</span> commonly tested compounds.</p></li>
<li><p>In contrast, the EGFR-ErbB2 pair has <span class="math notranslate nohighlight">\(1170\)</span> commonly tested compounds, of which <span class="math notranslate nohighlight">\(662\)</span> were active on both.</p></li>
</ul>
<p>Now let’s look at the similarity, in this case, the reduced fraction:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[29]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">kinase_similarity_matrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="n">kinase_selection_df</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="n">kinase_selection_df</span><span class="p">)))</span>
<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">uniprot_id1</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">kinase_selection_df</span><span class="o">.</span><span class="n">uniprot_id</span><span class="p">):</span>
    <span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">uniprot_id2</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">kinase_selection_df</span><span class="o">.</span><span class="n">uniprot_id</span><span class="p">):</span>
        <span class="n">kinase_similarity_matrix</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="n">similarity_ligand_profile</span><span class="p">(</span><span class="n">uniprot_id1</span><span class="p">,</span> <span class="n">uniprot_id2</span><span class="p">,</span> <span class="n">data</span><span class="p">)[</span>
            <span class="mi">2</span>
        <span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
No compounds were measured on both kinases, namely P04626 and P42336.
No compounds were measured on both kinases, namely P42336 and P04626.
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">kinase_similarity_matrix_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
    <span class="n">data</span><span class="o">=</span><span class="n">kinase_similarity_matrix</span><span class="p">,</span>
    <span class="n">index</span><span class="o">=</span><span class="n">kinase_selection_df</span><span class="o">.</span><span class="n">kinase_klifs</span><span class="p">,</span>
    <span class="n">columns</span><span class="o">=</span><span class="n">kinase_selection_df</span><span class="o">.</span><span class="n">kinase_klifs</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">kinase_similarity_matrix_df</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="kc">None</span>
<span class="n">kinase_similarity_matrix_df</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="kc">None</span>
<span class="n">kinase_similarity_matrix_df</span>
<span class="c1"># NBVAL_CHECK_OUTPUT</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
<thead>
<tr style="text-align: right;">
<th></th>
<th>EGFR</th>
<th>ErbB2</th>
<th>p110a</th>
<th>KDR</th>
<th>BRAF</th>
<th>CDK2</th>
<th>LCK</th>
<th>MET</th>
<th>p38a</th>
</tr>
</thead>
<tbody>
<tr>
<th>EGFR</th>
<td>0.609388</td>
<td>0.565812</td>
<td>0.072626</td>
<td>0.350504</td>
<td>0.457627</td>
<td>0.125000</td>
<td>0.246032</td>
<td>0.228261</td>
<td>0.346154</td>
</tr>
<tr>
<th>ErbB2</th>
<td>0.565812</td>
<td>0.606471</td>
<td>NaN</td>
<td>0.400000</td>
<td>0.250000</td>
<td>0.148148</td>
<td>0.151515</td>
<td>0.035714</td>
<td>0.125000</td>
</tr>
<tr>
<th>p110a</th>
<td>0.072626</td>
<td>NaN</td>
<td>0.643524</td>
<td>0.183908</td>
<td>0.333333</td>
<td>0.333333</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
</tr>
<tr>
<th>KDR</th>
<td>0.350504</td>
<td>0.400000</td>
<td>0.183908</td>
<td>0.697291</td>
<td>0.759542</td>
<td>0.617391</td>
<td>0.433414</td>
<td>0.541176</td>
<td>0.516393</td>
</tr>
<tr>
<th>BRAF</th>
<td>0.457627</td>
<td>0.250000</td>
<td>0.333333</td>
<td>0.759542</td>
<td>0.811280</td>
<td>0.076923</td>
<td>0.550000</td>
<td>0.115385</td>
<td>0.707317</td>
</tr>
<tr>
<th>CDK2</th>
<td>0.125000</td>
<td>0.148148</td>
<td>0.333333</td>
<td>0.617391</td>
<td>0.076923</td>
<td>0.543333</td>
<td>0.111111</td>
<td>0.090909</td>
<td>0.111111</td>
</tr>
<tr>
<th>LCK</th>
<td>0.246032</td>
<td>0.151515</td>
<td>0.000000</td>
<td>0.433414</td>
<td>0.550000</td>
<td>0.111111</td>
<td>0.599359</td>
<td>0.269841</td>
<td>0.500000</td>
</tr>
<tr>
<th>MET</th>
<td>0.228261</td>
<td>0.035714</td>
<td>0.000000</td>
<td>0.541176</td>
<td>0.115385</td>
<td>0.090909</td>
<td>0.269841</td>
<td>0.793785</td>
<td>0.050000</td>
</tr>
<tr>
<th>p38a</th>
<td>0.346154</td>
<td>0.125000</td>
<td>0.000000</td>
<td>0.516393</td>
<td>0.707317</td>
<td>0.111111</td>
<td>0.500000</td>
<td>0.050000</td>
<td>0.763816</td>
</tr>
</tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[31]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Show matrix with background gradient</span>
<span class="n">cm</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">light_palette</span><span class="p">(</span><span class="s2">"green"</span><span class="p">,</span> <span class="n">as_cmap</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">kinase_similarity_matrix_df</span><span class="o">.</span><span class="n">style</span><span class="o">.</span><span class="n">background_gradient</span><span class="p">(</span><span class="n">cmap</span><span class="o">=</span><span class="n">cm</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s2">"</span><span class="si">{:.3f}</span><span class="s2">"</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[31]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<style type="text/css">
#T_9049f_row0_col0, #T_9049f_row1_col1, #T_9049f_row2_col2, #T_9049f_row3_col5, #T_9049f_row4_col3, #T_9049f_row4_col4, #T_9049f_row6_col6, #T_9049f_row7_col7, #T_9049f_row8_col8 {
  background-color: #008000;
  color: #f1f1f1;
}
#T_9049f_row0_col1, #T_9049f_row3_col4, #T_9049f_row4_col8 {
  background-color: #118811;
  color: #f1f1f1;
}
#T_9049f_row0_col2 {
  background-color: #d1e6d1;
  color: #000000;
}
#T_9049f_row0_col3, #T_9049f_row7_col0 {
  background-color: #a6d2a6;
  color: #000000;
}
#T_9049f_row0_col4, #T_9049f_row3_col0, #T_9049f_row4_col2, #T_9049f_row5_col2 {
  background-color: #71b771;
  color: #f1f1f1;
}
#T_9049f_row0_col5 {
  background-color: #d6e9d6;
  color: #000000;
}
#T_9049f_row0_col6 {
  background-color: #8ac48a;
  color: #000000;
}
#T_9049f_row0_col7, #T_9049f_row3_col2 {
  background-color: #a7d2a7;
  color: #000000;
}
#T_9049f_row0_col8 {
  background-color: #80bf80;
  color: #000000;
}
#T_9049f_row1_col0 {
  background-color: #128912;
  color: #f1f1f1;
}
#T_9049f_row1_col2, #T_9049f_row2_col1 {
  background-color: #000000;
  color: #f1f1f1;
}
#T_9049f_row1_col3, #T_9049f_row4_col1 {
  background-color: #92c892;
  color: #000000;
}
#T_9049f_row1_col4 {
  background-color: #b3d8b3;
  color: #000000;
}
#T_9049f_row1_col5 {
  background-color: #cce4cc;
  color: #000000;
}
#T_9049f_row1_col6 {
  background-color: #b0d6b0;
  color: #000000;
}
#T_9049f_row1_col7 {
  background-color: #e0eee0;
  color: #000000;
}
#T_9049f_row1_col8 {
  background-color: #c5e0c5;
  color: #000000;
}
#T_9049f_row2_col0, #T_9049f_row2_col3, #T_9049f_row2_col6, #T_9049f_row2_col7, #T_9049f_row2_col8, #T_9049f_row4_col5, #T_9049f_row5_col4, #T_9049f_row6_col2, #T_9049f_row7_col1, #T_9049f_row7_col2, #T_9049f_row8_col2 {
  background-color: #ebf3eb;
  color: #000000;
}
#T_9049f_row2_col4 {
  background-color: #99cb99;
  color: #000000;
}
#T_9049f_row2_col5 {
  background-color: #7bbc7b;
  color: #000000;
}
#T_9049f_row3_col1 {
  background-color: #55a955;
  color: #f1f1f1;
}
#T_9049f_row3_col3 {
  background-color: #198c19;
  color: #f1f1f1;
}
#T_9049f_row3_col6 {
  background-color: #40a040;
  color: #f1f1f1;
}
#T_9049f_row3_col7 {
  background-color: #4ba54a;
  color: #f1f1f1;
}
#T_9049f_row3_col8 {
  background-color: #4ba54b;
  color: #f1f1f1;
}
#T_9049f_row4_col0 {
  background-color: #42a042;
  color: #f1f1f1;
}
#T_9049f_row4_col6 {
  background-color: #138913;
  color: #f1f1f1;
}
#T_9049f_row4_col7, #T_9049f_row5_col8 {
  background-color: #c9e2c8;
  color: #000000;
}
#T_9049f_row5_col0 {
  background-color: #d4e8d4;
  color: #000000;
}
#T_9049f_row5_col1 {
  background-color: #bddcbd;
  color: #000000;
}
#T_9049f_row5_col3 {
  background-color: #3a9c3a;
  color: #f1f1f1;
}
#T_9049f_row5_col5 {
  background-color: #209020;
  color: #f1f1f1;
}
#T_9049f_row5_col6 {
  background-color: #bfdebf;
  color: #000000;
}
#T_9049f_row5_col7 {
  background-color: #d0e6d0;
  color: #000000;
}
#T_9049f_row6_col0 {
  background-color: #9fce9f;
  color: #000000;
}
#T_9049f_row6_col1 {
  background-color: #bcdcbc;
  color: #000000;
}
#T_9049f_row6_col3 {
  background-color: #85c185;
  color: #000000;
}
#T_9049f_row6_col4 {
  background-color: #54a954;
  color: #f1f1f1;
}
#T_9049f_row6_col5, #T_9049f_row7_col8, #T_9049f_row8_col5, #T_9049f_row8_col7 {
  background-color: #dcecdc;
  color: #000000;
}
#T_9049f_row6_col7 {
  background-color: #9bcc9b;
  color: #000000;
}
#T_9049f_row6_col8 {
  background-color: #51a851;
  color: #f1f1f1;
}
#T_9049f_row7_col3 {
  background-color: #59ac59;
  color: #f1f1f1;
}
#T_9049f_row7_col4 {
  background-color: #dfeddf;
  color: #000000;
}
#T_9049f_row7_col5 {
  background-color: #e5f0e5;
  color: #000000;
}
#T_9049f_row7_col6 {
  background-color: #81bf81;
  color: #000000;
}
#T_9049f_row8_col0 {
  background-color: #73b873;
  color: #f1f1f1;
}
#T_9049f_row8_col1 {
  background-color: #c6e1c6;
  color: #000000;
}
#T_9049f_row8_col3 {
  background-color: #63b163;
  color: #f1f1f1;
}
#T_9049f_row8_col4 {
  background-color: #219021;
  color: #f1f1f1;
}
#T_9049f_row8_col6 {
  background-color: #279327;
  color: #f1f1f1;
}
</style>
<table id="T_9049f_">
<thead>
<tr>
<th class="blank level0"> </th>
<th class="col_heading level0 col0">EGFR</th>
<th class="col_heading level0 col1">ErbB2</th>
<th class="col_heading level0 col2">p110a</th>
<th class="col_heading level0 col3">KDR</th>
<th class="col_heading level0 col4">BRAF</th>
<th class="col_heading level0 col5">CDK2</th>
<th class="col_heading level0 col6">LCK</th>
<th class="col_heading level0 col7">MET</th>
<th class="col_heading level0 col8">p38a</th>
</tr>
</thead>
<tbody>
<tr>
<th class="row_heading level0 row0" id="T_9049f_level0_row0">EGFR</th>
<td class="data row0 col0" id="T_9049f_row0_col0">0.609</td>
<td class="data row0 col1" id="T_9049f_row0_col1">0.566</td>
<td class="data row0 col2" id="T_9049f_row0_col2">0.073</td>
<td class="data row0 col3" id="T_9049f_row0_col3">0.351</td>
<td class="data row0 col4" id="T_9049f_row0_col4">0.458</td>
<td class="data row0 col5" id="T_9049f_row0_col5">0.125</td>
<td class="data row0 col6" id="T_9049f_row0_col6">0.246</td>
<td class="data row0 col7" id="T_9049f_row0_col7">0.228</td>
<td class="data row0 col8" id="T_9049f_row0_col8">0.346</td>
</tr>
<tr>
<th class="row_heading level0 row1" id="T_9049f_level0_row1">ErbB2</th>
<td class="data row1 col0" id="T_9049f_row1_col0">0.566</td>
<td class="data row1 col1" id="T_9049f_row1_col1">0.606</td>
<td class="data row1 col2" id="T_9049f_row1_col2">nan</td>
<td class="data row1 col3" id="T_9049f_row1_col3">0.400</td>
<td class="data row1 col4" id="T_9049f_row1_col4">0.250</td>
<td class="data row1 col5" id="T_9049f_row1_col5">0.148</td>
<td class="data row1 col6" id="T_9049f_row1_col6">0.152</td>
<td class="data row1 col7" id="T_9049f_row1_col7">0.036</td>
<td class="data row1 col8" id="T_9049f_row1_col8">0.125</td>
</tr>
<tr>
<th class="row_heading level0 row2" id="T_9049f_level0_row2">p110a</th>
<td class="data row2 col0" id="T_9049f_row2_col0">0.073</td>
<td class="data row2 col1" id="T_9049f_row2_col1">nan</td>
<td class="data row2 col2" id="T_9049f_row2_col2">0.644</td>
<td class="data row2 col3" id="T_9049f_row2_col3">0.184</td>
<td class="data row2 col4" id="T_9049f_row2_col4">0.333</td>
<td class="data row2 col5" id="T_9049f_row2_col5">0.333</td>
<td class="data row2 col6" id="T_9049f_row2_col6">0.000</td>
<td class="data row2 col7" id="T_9049f_row2_col7">0.000</td>
<td class="data row2 col8" id="T_9049f_row2_col8">0.000</td>
</tr>
<tr>
<th class="row_heading level0 row3" id="T_9049f_level0_row3">KDR</th>
<td class="data row3 col0" id="T_9049f_row3_col0">0.351</td>
<td class="data row3 col1" id="T_9049f_row3_col1">0.400</td>
<td class="data row3 col2" id="T_9049f_row3_col2">0.184</td>
<td class="data row3 col3" id="T_9049f_row3_col3">0.697</td>
<td class="data row3 col4" id="T_9049f_row3_col4">0.760</td>
<td class="data row3 col5" id="T_9049f_row3_col5">0.617</td>
<td class="data row3 col6" id="T_9049f_row3_col6">0.433</td>
<td class="data row3 col7" id="T_9049f_row3_col7">0.541</td>
<td class="data row3 col8" id="T_9049f_row3_col8">0.516</td>
</tr>
<tr>
<th class="row_heading level0 row4" id="T_9049f_level0_row4">BRAF</th>
<td class="data row4 col0" id="T_9049f_row4_col0">0.458</td>
<td class="data row4 col1" id="T_9049f_row4_col1">0.250</td>
<td class="data row4 col2" id="T_9049f_row4_col2">0.333</td>
<td class="data row4 col3" id="T_9049f_row4_col3">0.760</td>
<td class="data row4 col4" id="T_9049f_row4_col4">0.811</td>
<td class="data row4 col5" id="T_9049f_row4_col5">0.077</td>
<td class="data row4 col6" id="T_9049f_row4_col6">0.550</td>
<td class="data row4 col7" id="T_9049f_row4_col7">0.115</td>
<td class="data row4 col8" id="T_9049f_row4_col8">0.707</td>
</tr>
<tr>
<th class="row_heading level0 row5" id="T_9049f_level0_row5">CDK2</th>
<td class="data row5 col0" id="T_9049f_row5_col0">0.125</td>
<td class="data row5 col1" id="T_9049f_row5_col1">0.148</td>
<td class="data row5 col2" id="T_9049f_row5_col2">0.333</td>
<td class="data row5 col3" id="T_9049f_row5_col3">0.617</td>
<td class="data row5 col4" id="T_9049f_row5_col4">0.077</td>
<td class="data row5 col5" id="T_9049f_row5_col5">0.543</td>
<td class="data row5 col6" id="T_9049f_row5_col6">0.111</td>
<td class="data row5 col7" id="T_9049f_row5_col7">0.091</td>
<td class="data row5 col8" id="T_9049f_row5_col8">0.111</td>
</tr>
<tr>
<th class="row_heading level0 row6" id="T_9049f_level0_row6">LCK</th>
<td class="data row6 col0" id="T_9049f_row6_col0">0.246</td>
<td class="data row6 col1" id="T_9049f_row6_col1">0.152</td>
<td class="data row6 col2" id="T_9049f_row6_col2">0.000</td>
<td class="data row6 col3" id="T_9049f_row6_col3">0.433</td>
<td class="data row6 col4" id="T_9049f_row6_col4">0.550</td>
<td class="data row6 col5" id="T_9049f_row6_col5">0.111</td>
<td class="data row6 col6" id="T_9049f_row6_col6">0.599</td>
<td class="data row6 col7" id="T_9049f_row6_col7">0.270</td>
<td class="data row6 col8" id="T_9049f_row6_col8">0.500</td>
</tr>
<tr>
<th class="row_heading level0 row7" id="T_9049f_level0_row7">MET</th>
<td class="data row7 col0" id="T_9049f_row7_col0">0.228</td>
<td class="data row7 col1" id="T_9049f_row7_col1">0.036</td>
<td class="data row7 col2" id="T_9049f_row7_col2">0.000</td>
<td class="data row7 col3" id="T_9049f_row7_col3">0.541</td>
<td class="data row7 col4" id="T_9049f_row7_col4">0.115</td>
<td class="data row7 col5" id="T_9049f_row7_col5">0.091</td>
<td class="data row7 col6" id="T_9049f_row7_col6">0.270</td>
<td class="data row7 col7" id="T_9049f_row7_col7">0.794</td>
<td class="data row7 col8" id="T_9049f_row7_col8">0.050</td>
</tr>
<tr>
<th class="row_heading level0 row8" id="T_9049f_level0_row8">p38a</th>
<td class="data row8 col0" id="T_9049f_row8_col0">0.346</td>
<td class="data row8 col1" id="T_9049f_row8_col1">0.125</td>
<td class="data row8 col2" id="T_9049f_row8_col2">0.000</td>
<td class="data row8 col3" id="T_9049f_row8_col3">0.516</td>
<td class="data row8 col4" id="T_9049f_row8_col4">0.707</td>
<td class="data row8 col5" id="T_9049f_row8_col5">0.111</td>
<td class="data row8 col6" id="T_9049f_row8_col6">0.500</td>
<td class="data row8 col7" id="T_9049f_row8_col7">0.050</td>
<td class="data row8 col8" id="T_9049f_row8_col8">0.764</td>
</tr>
</tbody>
</table></div>
</div>
<p>Note that the diagonal contains the previously discussed promiscuity values.</p>
<p>As mentioned above, no compounds were measured on both ErbB2 and p110a and therefore creates a <code class="docutils literal notranslate"><span class="pre">np.nan</span></code> entry which can be problematic for algorithmic reason.</p>
<p>As a simple workaround, we will fill the NA values with zero.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[32]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">kinase_similarity_matrix_df</span> <span class="o">=</span> <span class="n">kinase_similarity_matrix_df</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>

<span class="n">kinase_similarity_matrix_df</span><span class="o">.</span><span class="n">style</span><span class="o">.</span><span class="n">background_gradient</span><span class="p">(</span><span class="n">cmap</span><span class="o">=</span><span class="n">cm</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s2">"</span><span class="si">{:.3f}</span><span class="s2">"</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[32]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<style type="text/css">
#T_6f48b_row0_col0, #T_6f48b_row1_col1, #T_6f48b_row2_col2, #T_6f48b_row3_col5, #T_6f48b_row4_col3, #T_6f48b_row4_col4, #T_6f48b_row6_col6, #T_6f48b_row7_col7, #T_6f48b_row8_col8 {
  background-color: #008000;
  color: #f1f1f1;
}
#T_6f48b_row0_col1 {
  background-color: #108810;
  color: #f1f1f1;
}
#T_6f48b_row0_col2 {
  background-color: #d1e6d1;
  color: #000000;
}
#T_6f48b_row0_col3, #T_6f48b_row7_col0 {
  background-color: #a6d2a6;
  color: #000000;
}
#T_6f48b_row0_col4, #T_6f48b_row3_col0, #T_6f48b_row4_col2, #T_6f48b_row5_col2 {
  background-color: #71b771;
  color: #f1f1f1;
}
#T_6f48b_row0_col5 {
  background-color: #d6e9d6;
  color: #000000;
}
#T_6f48b_row0_col6, #T_6f48b_row4_col1 {
  background-color: #8ac48a;
  color: #000000;
}
#T_6f48b_row0_col7, #T_6f48b_row3_col2 {
  background-color: #a7d2a7;
  color: #000000;
}
#T_6f48b_row0_col8 {
  background-color: #80bf80;
  color: #000000;
}
#T_6f48b_row1_col0 {
  background-color: #128912;
  color: #f1f1f1;
}
#T_6f48b_row1_col2, #T_6f48b_row2_col0, #T_6f48b_row2_col1, #T_6f48b_row2_col3, #T_6f48b_row2_col6, #T_6f48b_row2_col7, #T_6f48b_row2_col8, #T_6f48b_row4_col5, #T_6f48b_row5_col4, #T_6f48b_row6_col2, #T_6f48b_row7_col2, #T_6f48b_row8_col2 {
  background-color: #ebf3eb;
  color: #000000;
}
#T_6f48b_row1_col3 {
  background-color: #92c892;
  color: #000000;
}
#T_6f48b_row1_col4 {
  background-color: #b3d8b3;
  color: #000000;
}
#T_6f48b_row1_col5 {
  background-color: #cce4cc;
  color: #000000;
}
#T_6f48b_row1_col6 {
  background-color: #b0d6b0;
  color: #000000;
}
#T_6f48b_row1_col7 {
  background-color: #e0eee0;
  color: #000000;
}
#T_6f48b_row1_col8 {
  background-color: #c5e0c5;
  color: #000000;
}
#T_6f48b_row2_col4 {
  background-color: #99cb99;
  color: #000000;
}
#T_6f48b_row2_col5 {
  background-color: #7bbc7b;
  color: #000000;
}
#T_6f48b_row3_col1 {
  background-color: #50a750;
  color: #f1f1f1;
}
#T_6f48b_row3_col3 {
  background-color: #198c19;
  color: #f1f1f1;
}
#T_6f48b_row3_col4, #T_6f48b_row4_col8 {
  background-color: #118811;
  color: #f1f1f1;
}
#T_6f48b_row3_col6 {
  background-color: #40a040;
  color: #f1f1f1;
}
#T_6f48b_row3_col7 {
  background-color: #4ba54a;
  color: #f1f1f1;
}
#T_6f48b_row3_col8 {
  background-color: #4ba54b;
  color: #f1f1f1;
}
#T_6f48b_row4_col0 {
  background-color: #42a042;
  color: #f1f1f1;
}
#T_6f48b_row4_col6 {
  background-color: #138913;
  color: #f1f1f1;
}
#T_6f48b_row4_col7, #T_6f48b_row5_col8 {
  background-color: #c9e2c8;
  color: #000000;
}
#T_6f48b_row5_col0 {
  background-color: #d4e8d4;
  color: #000000;
}
#T_6f48b_row5_col1 {
  background-color: #b2d7b1;
  color: #000000;
}
#T_6f48b_row5_col3 {
  background-color: #3a9c3a;
  color: #f1f1f1;
}
#T_6f48b_row5_col5 {
  background-color: #209020;
  color: #f1f1f1;
}
#T_6f48b_row5_col6 {
  background-color: #bfdebf;
  color: #000000;
}
#T_6f48b_row5_col7 {
  background-color: #d0e6d0;
  color: #000000;
}
#T_6f48b_row6_col0 {
  background-color: #9fce9f;
  color: #000000;
}
#T_6f48b_row6_col1 {
  background-color: #b1d7b1;
  color: #000000;
}
#T_6f48b_row6_col3 {
  background-color: #85c185;
  color: #000000;
}
#T_6f48b_row6_col4 {
  background-color: #54a954;
  color: #f1f1f1;
}
#T_6f48b_row6_col5, #T_6f48b_row7_col8, #T_6f48b_row8_col5, #T_6f48b_row8_col7 {
  background-color: #dcecdc;
  color: #000000;
}
#T_6f48b_row6_col7 {
  background-color: #9bcc9b;
  color: #000000;
}
#T_6f48b_row6_col8 {
  background-color: #51a851;
  color: #f1f1f1;
}
#T_6f48b_row7_col1 {
  background-color: #ddecdd;
  color: #000000;
}
#T_6f48b_row7_col3 {
  background-color: #59ac59;
  color: #f1f1f1;
}
#T_6f48b_row7_col4 {
  background-color: #dfeddf;
  color: #000000;
}
#T_6f48b_row7_col5 {
  background-color: #e5f0e5;
  color: #000000;
}
#T_6f48b_row7_col6 {
  background-color: #81bf81;
  color: #000000;
}
#T_6f48b_row8_col0 {
  background-color: #73b873;
  color: #f1f1f1;
}
#T_6f48b_row8_col1 {
  background-color: #bbdcbb;
  color: #000000;
}
#T_6f48b_row8_col3 {
  background-color: #63b163;
  color: #f1f1f1;
}
#T_6f48b_row8_col4 {
  background-color: #219021;
  color: #f1f1f1;
}
#T_6f48b_row8_col6 {
  background-color: #279327;
  color: #f1f1f1;
}
</style>
<table id="T_6f48b_">
<thead>
<tr>
<th class="blank level0"> </th>
<th class="col_heading level0 col0">EGFR</th>
<th class="col_heading level0 col1">ErbB2</th>
<th class="col_heading level0 col2">p110a</th>
<th class="col_heading level0 col3">KDR</th>
<th class="col_heading level0 col4">BRAF</th>
<th class="col_heading level0 col5">CDK2</th>
<th class="col_heading level0 col6">LCK</th>
<th class="col_heading level0 col7">MET</th>
<th class="col_heading level0 col8">p38a</th>
</tr>
</thead>
<tbody>
<tr>
<th class="row_heading level0 row0" id="T_6f48b_level0_row0">EGFR</th>
<td class="data row0 col0" id="T_6f48b_row0_col0">0.609</td>
<td class="data row0 col1" id="T_6f48b_row0_col1">0.566</td>
<td class="data row0 col2" id="T_6f48b_row0_col2">0.073</td>
<td class="data row0 col3" id="T_6f48b_row0_col3">0.351</td>
<td class="data row0 col4" id="T_6f48b_row0_col4">0.458</td>
<td class="data row0 col5" id="T_6f48b_row0_col5">0.125</td>
<td class="data row0 col6" id="T_6f48b_row0_col6">0.246</td>
<td class="data row0 col7" id="T_6f48b_row0_col7">0.228</td>
<td class="data row0 col8" id="T_6f48b_row0_col8">0.346</td>
</tr>
<tr>
<th class="row_heading level0 row1" id="T_6f48b_level0_row1">ErbB2</th>
<td class="data row1 col0" id="T_6f48b_row1_col0">0.566</td>
<td class="data row1 col1" id="T_6f48b_row1_col1">0.606</td>
<td class="data row1 col2" id="T_6f48b_row1_col2">0.000</td>
<td class="data row1 col3" id="T_6f48b_row1_col3">0.400</td>
<td class="data row1 col4" id="T_6f48b_row1_col4">0.250</td>
<td class="data row1 col5" id="T_6f48b_row1_col5">0.148</td>
<td class="data row1 col6" id="T_6f48b_row1_col6">0.152</td>
<td class="data row1 col7" id="T_6f48b_row1_col7">0.036</td>
<td class="data row1 col8" id="T_6f48b_row1_col8">0.125</td>
</tr>
<tr>
<th class="row_heading level0 row2" id="T_6f48b_level0_row2">p110a</th>
<td class="data row2 col0" id="T_6f48b_row2_col0">0.073</td>
<td class="data row2 col1" id="T_6f48b_row2_col1">0.000</td>
<td class="data row2 col2" id="T_6f48b_row2_col2">0.644</td>
<td class="data row2 col3" id="T_6f48b_row2_col3">0.184</td>
<td class="data row2 col4" id="T_6f48b_row2_col4">0.333</td>
<td class="data row2 col5" id="T_6f48b_row2_col5">0.333</td>
<td class="data row2 col6" id="T_6f48b_row2_col6">0.000</td>
<td class="data row2 col7" id="T_6f48b_row2_col7">0.000</td>
<td class="data row2 col8" id="T_6f48b_row2_col8">0.000</td>
</tr>
<tr>
<th class="row_heading level0 row3" id="T_6f48b_level0_row3">KDR</th>
<td class="data row3 col0" id="T_6f48b_row3_col0">0.351</td>
<td class="data row3 col1" id="T_6f48b_row3_col1">0.400</td>
<td class="data row3 col2" id="T_6f48b_row3_col2">0.184</td>
<td class="data row3 col3" id="T_6f48b_row3_col3">0.697</td>
<td class="data row3 col4" id="T_6f48b_row3_col4">0.760</td>
<td class="data row3 col5" id="T_6f48b_row3_col5">0.617</td>
<td class="data row3 col6" id="T_6f48b_row3_col6">0.433</td>
<td class="data row3 col7" id="T_6f48b_row3_col7">0.541</td>
<td class="data row3 col8" id="T_6f48b_row3_col8">0.516</td>
</tr>
<tr>
<th class="row_heading level0 row4" id="T_6f48b_level0_row4">BRAF</th>
<td class="data row4 col0" id="T_6f48b_row4_col0">0.458</td>
<td class="data row4 col1" id="T_6f48b_row4_col1">0.250</td>
<td class="data row4 col2" id="T_6f48b_row4_col2">0.333</td>
<td class="data row4 col3" id="T_6f48b_row4_col3">0.760</td>
<td class="data row4 col4" id="T_6f48b_row4_col4">0.811</td>
<td class="data row4 col5" id="T_6f48b_row4_col5">0.077</td>
<td class="data row4 col6" id="T_6f48b_row4_col6">0.550</td>
<td class="data row4 col7" id="T_6f48b_row4_col7">0.115</td>
<td class="data row4 col8" id="T_6f48b_row4_col8">0.707</td>
</tr>
<tr>
<th class="row_heading level0 row5" id="T_6f48b_level0_row5">CDK2</th>
<td class="data row5 col0" id="T_6f48b_row5_col0">0.125</td>
<td class="data row5 col1" id="T_6f48b_row5_col1">0.148</td>
<td class="data row5 col2" id="T_6f48b_row5_col2">0.333</td>
<td class="data row5 col3" id="T_6f48b_row5_col3">0.617</td>
<td class="data row5 col4" id="T_6f48b_row5_col4">0.077</td>
<td class="data row5 col5" id="T_6f48b_row5_col5">0.543</td>
<td class="data row5 col6" id="T_6f48b_row5_col6">0.111</td>
<td class="data row5 col7" id="T_6f48b_row5_col7">0.091</td>
<td class="data row5 col8" id="T_6f48b_row5_col8">0.111</td>
</tr>
<tr>
<th class="row_heading level0 row6" id="T_6f48b_level0_row6">LCK</th>
<td class="data row6 col0" id="T_6f48b_row6_col0">0.246</td>
<td class="data row6 col1" id="T_6f48b_row6_col1">0.152</td>
<td class="data row6 col2" id="T_6f48b_row6_col2">0.000</td>
<td class="data row6 col3" id="T_6f48b_row6_col3">0.433</td>
<td class="data row6 col4" id="T_6f48b_row6_col4">0.550</td>
<td class="data row6 col5" id="T_6f48b_row6_col5">0.111</td>
<td class="data row6 col6" id="T_6f48b_row6_col6">0.599</td>
<td class="data row6 col7" id="T_6f48b_row6_col7">0.270</td>
<td class="data row6 col8" id="T_6f48b_row6_col8">0.500</td>
</tr>
<tr>
<th class="row_heading level0 row7" id="T_6f48b_level0_row7">MET</th>
<td class="data row7 col0" id="T_6f48b_row7_col0">0.228</td>
<td class="data row7 col1" id="T_6f48b_row7_col1">0.036</td>
<td class="data row7 col2" id="T_6f48b_row7_col2">0.000</td>
<td class="data row7 col3" id="T_6f48b_row7_col3">0.541</td>
<td class="data row7 col4" id="T_6f48b_row7_col4">0.115</td>
<td class="data row7 col5" id="T_6f48b_row7_col5">0.091</td>
<td class="data row7 col6" id="T_6f48b_row7_col6">0.270</td>
<td class="data row7 col7" id="T_6f48b_row7_col7">0.794</td>
<td class="data row7 col8" id="T_6f48b_row7_col8">0.050</td>
</tr>
<tr>
<th class="row_heading level0 row8" id="T_6f48b_level0_row8">p38a</th>
<td class="data row8 col0" id="T_6f48b_row8_col0">0.346</td>
<td class="data row8 col1" id="T_6f48b_row8_col1">0.125</td>
<td class="data row8 col2" id="T_6f48b_row8_col2">0.000</td>
<td class="data row8 col3" id="T_6f48b_row8_col3">0.516</td>
<td class="data row8 col4" id="T_6f48b_row8_col4">0.707</td>
<td class="data row8 col5" id="T_6f48b_row8_col5">0.111</td>
<td class="data row8 col6" id="T_6f48b_row8_col6">0.500</td>
<td class="data row8 col7" id="T_6f48b_row8_col7">0.050</td>
<td class="data row8 col8" id="T_6f48b_row8_col8">0.764</td>
</tr>
</tbody>
</table></div>
</div>
</section>
<section id="Save-kinase-similarity-matrix">
<h4 id="Save-kinase-similarity-matrix">Save kinase similarity matrix<a class="headerlink" href="#Save-kinase-similarity-matrix" title="Permalink to this heading">¶</a></h4>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[33]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">kinase_similarity_matrix_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">DATA</span> <span class="o">/</span> <span class="s2">"kinase_similarity_matrix.csv"</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
</section>
<section id="Kinase-distance-matrix">
<h3 id="Kinase-distance-matrix">Kinase distance matrix<a class="headerlink" href="#Kinase-distance-matrix" title="Permalink to this heading">¶</a></h3>
<p>The similarity matrix <span class="math notranslate nohighlight">\(SM\)</span> is converted to a pseudo-distance matrix (all entries of the similarity matrix are between <span class="math notranslate nohighlight">\(0\)</span> and <span class="math notranslate nohighlight">\(1\)</span>):</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[34]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span>
    <span class="sa">f</span><span class="s2">"The values of the similarity matrix lie between: "</span>
    <span class="sa">f</span><span class="s2">"</span><span class="si">{</span><span class="n">kinase_similarity_matrix_df</span><span class="o">.</span><span class="n">min</span><span class="p">()</span><span class="o">.</span><span class="n">min</span><span class="p">()</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">"</span>
    <span class="sa">f</span><span class="s2">" and </span><span class="si">{</span><span class="n">kinase_similarity_matrix_df</span><span class="o">.</span><span class="n">max</span><span class="p">()</span><span class="o">.</span><span class="n">max</span><span class="p">()</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">"</span>
<span class="p">)</span>
<span class="c1"># NBVAL_CHECK_OUTPUT</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
The values of the similarity matrix lie between: 0.00 and 0.81
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[35]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">kinase_distance_matrix_df</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">-</span> <span class="n">kinase_similarity_matrix_df</span>
</pre></div>
</div>
</div>
<p>Finally, we set the diagonal values to <span class="math notranslate nohighlight">\(0\)</span> and we obtain the kinase distance matrix:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[36]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">np</span><span class="o">.</span><span class="n">fill_diagonal</span><span class="p">(</span><span class="n">kinase_distance_matrix_df</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[37]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">kinase_distance_matrix_df</span><span class="o">.</span><span class="n">style</span><span class="o">.</span><span class="n">background_gradient</span><span class="p">(</span><span class="n">cmap</span><span class="o">=</span><span class="n">cm</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s2">"</span><span class="si">{:.3f}</span><span class="s2">"</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[37]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<style type="text/css">
#T_1fe83_row0_col0, #T_1fe83_row1_col1, #T_1fe83_row2_col2, #T_1fe83_row3_col3, #T_1fe83_row4_col4, #T_1fe83_row5_col5, #T_1fe83_row6_col6, #T_1fe83_row7_col7, #T_1fe83_row8_col8 {
  background-color: #ebf3eb;
  color: #000000;
}
#T_1fe83_row0_col1 {
  background-color: #84c184;
  color: #000000;
}
#T_1fe83_row0_col2 {
  background-color: #118811;
  color: #f1f1f1;
}
#T_1fe83_row0_col3 {
  background-color: #309730;
  color: #f1f1f1;
}
#T_1fe83_row0_col4 {
  background-color: #61af61;
  color: #f1f1f1;
}
#T_1fe83_row0_col5 {
  background-color: #0c860c;
  color: #f1f1f1;
}
#T_1fe83_row0_col6 {
  background-color: #399c39;
  color: #f1f1f1;
}
#T_1fe83_row0_col7 {
  background-color: #359a35;
  color: #f1f1f1;
}
#T_1fe83_row0_col8 {
  background-color: #51a851;
  color: #f1f1f1;
}
#T_1fe83_row1_col0 {
  background-color: #7dbd7d;
  color: #000000;
}
#T_1fe83_row1_col2, #T_1fe83_row2_col0, #T_1fe83_row2_col1, #T_1fe83_row2_col3, #T_1fe83_row2_col6, #T_1fe83_row2_col7, #T_1fe83_row2_col8, #T_1fe83_row4_col5, #T_1fe83_row5_col4, #T_1fe83_row6_col2, #T_1fe83_row7_col2, #T_1fe83_row8_col2 {
  background-color: #008000;
  color: #f1f1f1;
}
#T_1fe83_row1_col3 {
  background-color: #3e9e3e;
  color: #f1f1f1;
}
#T_1fe83_row1_col4, #T_1fe83_row3_col2, #T_1fe83_row6_col0 {
  background-color: #2b952b;
  color: #f1f1f1;
}
#T_1fe83_row1_col5 {
  background-color: #118911;
  color: #f1f1f1;
}
#T_1fe83_row1_col6, #T_1fe83_row6_col1 {
  background-color: #239123;
  color: #f1f1f1;
}
#T_1fe83_row1_col7, #T_1fe83_row6_col5, #T_1fe83_row7_col1, #T_1fe83_row8_col5 {
  background-color: #088408;
  color: #f1f1f1;
}
#T_1fe83_row1_col8, #T_1fe83_row8_col1 {
  background-color: #1d8e1d;
  color: #f1f1f1;
}
#T_1fe83_row2_col4, #T_1fe83_row2_col5 {
  background-color: #41a041;
  color: #f1f1f1;
}
#T_1fe83_row3_col0 {
  background-color: #46a246;
  color: #f1f1f1;
}
#T_1fe83_row3_col1 {
  background-color: #5eae5e;
  color: #f1f1f1;
}
#T_1fe83_row3_col4 {
  background-color: #aed5ae;
  color: #000000;
}
#T_1fe83_row3_col5 {
  background-color: #89c389;
  color: #000000;
}
#T_1fe83_row3_col6 {
  background-color: #65b265;
  color: #f1f1f1;
}
#T_1fe83_row3_col7 {
  background-color: #7fbe7f;
  color: #000000;
}
#T_1fe83_row3_col8 {
  background-color: #79bc79;
  color: #000000;
}
#T_1fe83_row4_col0 {
  background-color: #61b061;
  color: #f1f1f1;
}
#T_1fe83_row4_col1 {
  background-color: #3a9c3a;
  color: #f1f1f1;
}
#T_1fe83_row4_col2, #T_1fe83_row5_col2 {
  background-color: #4ea64e;
  color: #f1f1f1;
}
#T_1fe83_row4_col3 {
  background-color: #a6d1a6;
  color: #000000;
}
#T_1fe83_row4_col6 {
  background-color: #81bf81;
  color: #000000;
}
#T_1fe83_row4_col7 {
  background-color: #1b8d1b;
  color: #f1f1f1;
}
#T_1fe83_row4_col8 {
  background-color: #a6d2a6;
  color: #000000;
}
#T_1fe83_row5_col0 {
  background-color: #0d860d;
  color: #f1f1f1;
}
#T_1fe83_row5_col1 {
  background-color: #229122;
  color: #f1f1f1;
}
#T_1fe83_row5_col3 {
  background-color: #7cbd7c;
  color: #000000;
}
#T_1fe83_row5_col6, #T_1fe83_row5_col8 {
  background-color: #1a8d1a;
  color: #f1f1f1;
}
#T_1fe83_row5_col7 {
  background-color: #158a15;
  color: #f1f1f1;
}
#T_1fe83_row6_col3 {
  background-color: #48a348;
  color: #f1f1f1;
}
#T_1fe83_row6_col4 {
  background-color: #78bb78;
  color: #000000;
}
#T_1fe83_row6_col7, #T_1fe83_row7_col6 {
  background-color: #3f9f3f;
  color: #f1f1f1;
}
#T_1fe83_row6_col8, #T_1fe83_row8_col6 {
  background-color: #75b975;
  color: #f1f1f1;
}
#T_1fe83_row7_col0 {
  background-color: #279327;
  color: #f1f1f1;
}
#T_1fe83_row7_col3 {
  background-color: #67b267;
  color: #f1f1f1;
}
#T_1fe83_row7_col4 {
  background-color: #098509;
  color: #f1f1f1;
}
#T_1fe83_row7_col5 {
  background-color: #038103;
  color: #f1f1f1;
}
#T_1fe83_row7_col8, #T_1fe83_row8_col7 {
  background-color: #0b850b;
  color: #f1f1f1;
}
#T_1fe83_row8_col0 {
  background-color: #45a245;
  color: #f1f1f1;
}
#T_1fe83_row8_col3 {
  background-color: #60af60;
  color: #f1f1f1;
}
#T_1fe83_row8_col4 {
  background-color: #a0cea0;
  color: #000000;
}
</style>
<table id="T_1fe83_">
<thead>
<tr>
<th class="blank level0"> </th>
<th class="col_heading level0 col0">EGFR</th>
<th class="col_heading level0 col1">ErbB2</th>
<th class="col_heading level0 col2">p110a</th>
<th class="col_heading level0 col3">KDR</th>
<th class="col_heading level0 col4">BRAF</th>
<th class="col_heading level0 col5">CDK2</th>
<th class="col_heading level0 col6">LCK</th>
<th class="col_heading level0 col7">MET</th>
<th class="col_heading level0 col8">p38a</th>
</tr>
</thead>
<tbody>
<tr>
<th class="row_heading level0 row0" id="T_1fe83_level0_row0">EGFR</th>
<td class="data row0 col0" id="T_1fe83_row0_col0">0.000</td>
<td class="data row0 col1" id="T_1fe83_row0_col1">0.434</td>
<td class="data row0 col2" id="T_1fe83_row0_col2">0.927</td>
<td class="data row0 col3" id="T_1fe83_row0_col3">0.649</td>
<td class="data row0 col4" id="T_1fe83_row0_col4">0.542</td>
<td class="data row0 col5" id="T_1fe83_row0_col5">0.875</td>
<td class="data row0 col6" id="T_1fe83_row0_col6">0.754</td>
<td class="data row0 col7" id="T_1fe83_row0_col7">0.772</td>
<td class="data row0 col8" id="T_1fe83_row0_col8">0.654</td>
</tr>
<tr>
<th class="row_heading level0 row1" id="T_1fe83_level0_row1">ErbB2</th>
<td class="data row1 col0" id="T_1fe83_row1_col0">0.434</td>
<td class="data row1 col1" id="T_1fe83_row1_col1">0.000</td>
<td class="data row1 col2" id="T_1fe83_row1_col2">1.000</td>
<td class="data row1 col3" id="T_1fe83_row1_col3">0.600</td>
<td class="data row1 col4" id="T_1fe83_row1_col4">0.750</td>
<td class="data row1 col5" id="T_1fe83_row1_col5">0.852</td>
<td class="data row1 col6" id="T_1fe83_row1_col6">0.848</td>
<td class="data row1 col7" id="T_1fe83_row1_col7">0.964</td>
<td class="data row1 col8" id="T_1fe83_row1_col8">0.875</td>
</tr>
<tr>
<th class="row_heading level0 row2" id="T_1fe83_level0_row2">p110a</th>
<td class="data row2 col0" id="T_1fe83_row2_col0">0.927</td>
<td class="data row2 col1" id="T_1fe83_row2_col1">1.000</td>
<td class="data row2 col2" id="T_1fe83_row2_col2">0.000</td>
<td class="data row2 col3" id="T_1fe83_row2_col3">0.816</td>
<td class="data row2 col4" id="T_1fe83_row2_col4">0.667</td>
<td class="data row2 col5" id="T_1fe83_row2_col5">0.667</td>
<td class="data row2 col6" id="T_1fe83_row2_col6">1.000</td>
<td class="data row2 col7" id="T_1fe83_row2_col7">1.000</td>
<td class="data row2 col8" id="T_1fe83_row2_col8">1.000</td>
</tr>
<tr>
<th class="row_heading level0 row3" id="T_1fe83_level0_row3">KDR</th>
<td class="data row3 col0" id="T_1fe83_row3_col0">0.649</td>
<td class="data row3 col1" id="T_1fe83_row3_col1">0.600</td>
<td class="data row3 col2" id="T_1fe83_row3_col2">0.816</td>
<td class="data row3 col3" id="T_1fe83_row3_col3">0.000</td>
<td class="data row3 col4" id="T_1fe83_row3_col4">0.240</td>
<td class="data row3 col5" id="T_1fe83_row3_col5">0.383</td>
<td class="data row3 col6" id="T_1fe83_row3_col6">0.567</td>
<td class="data row3 col7" id="T_1fe83_row3_col7">0.459</td>
<td class="data row3 col8" id="T_1fe83_row3_col8">0.484</td>
</tr>
<tr>
<th class="row_heading level0 row4" id="T_1fe83_level0_row4">BRAF</th>
<td class="data row4 col0" id="T_1fe83_row4_col0">0.542</td>
<td class="data row4 col1" id="T_1fe83_row4_col1">0.750</td>
<td class="data row4 col2" id="T_1fe83_row4_col2">0.667</td>
<td class="data row4 col3" id="T_1fe83_row4_col3">0.240</td>
<td class="data row4 col4" id="T_1fe83_row4_col4">0.000</td>
<td class="data row4 col5" id="T_1fe83_row4_col5">0.923</td>
<td class="data row4 col6" id="T_1fe83_row4_col6">0.450</td>
<td class="data row4 col7" id="T_1fe83_row4_col7">0.885</td>
<td class="data row4 col8" id="T_1fe83_row4_col8">0.293</td>
</tr>
<tr>
<th class="row_heading level0 row5" id="T_1fe83_level0_row5">CDK2</th>
<td class="data row5 col0" id="T_1fe83_row5_col0">0.875</td>
<td class="data row5 col1" id="T_1fe83_row5_col1">0.852</td>
<td class="data row5 col2" id="T_1fe83_row5_col2">0.667</td>
<td class="data row5 col3" id="T_1fe83_row5_col3">0.383</td>
<td class="data row5 col4" id="T_1fe83_row5_col4">0.923</td>
<td class="data row5 col5" id="T_1fe83_row5_col5">0.000</td>
<td class="data row5 col6" id="T_1fe83_row5_col6">0.889</td>
<td class="data row5 col7" id="T_1fe83_row5_col7">0.909</td>
<td class="data row5 col8" id="T_1fe83_row5_col8">0.889</td>
</tr>
<tr>
<th class="row_heading level0 row6" id="T_1fe83_level0_row6">LCK</th>
<td class="data row6 col0" id="T_1fe83_row6_col0">0.754</td>
<td class="data row6 col1" id="T_1fe83_row6_col1">0.848</td>
<td class="data row6 col2" id="T_1fe83_row6_col2">1.000</td>
<td class="data row6 col3" id="T_1fe83_row6_col3">0.567</td>
<td class="data row6 col4" id="T_1fe83_row6_col4">0.450</td>
<td class="data row6 col5" id="T_1fe83_row6_col5">0.889</td>
<td class="data row6 col6" id="T_1fe83_row6_col6">0.000</td>
<td class="data row6 col7" id="T_1fe83_row6_col7">0.730</td>
<td class="data row6 col8" id="T_1fe83_row6_col8">0.500</td>
</tr>
<tr>
<th class="row_heading level0 row7" id="T_1fe83_level0_row7">MET</th>
<td class="data row7 col0" id="T_1fe83_row7_col0">0.772</td>
<td class="data row7 col1" id="T_1fe83_row7_col1">0.964</td>
<td class="data row7 col2" id="T_1fe83_row7_col2">1.000</td>
<td class="data row7 col3" id="T_1fe83_row7_col3">0.459</td>
<td class="data row7 col4" id="T_1fe83_row7_col4">0.885</td>
<td class="data row7 col5" id="T_1fe83_row7_col5">0.909</td>
<td class="data row7 col6" id="T_1fe83_row7_col6">0.730</td>
<td class="data row7 col7" id="T_1fe83_row7_col7">0.000</td>
<td class="data row7 col8" id="T_1fe83_row7_col8">0.950</td>
</tr>
<tr>
<th class="row_heading level0 row8" id="T_1fe83_level0_row8">p38a</th>
<td class="data row8 col0" id="T_1fe83_row8_col0">0.654</td>
<td class="data row8 col1" id="T_1fe83_row8_col1">0.875</td>
<td class="data row8 col2" id="T_1fe83_row8_col2">1.000</td>
<td class="data row8 col3" id="T_1fe83_row8_col3">0.484</td>
<td class="data row8 col4" id="T_1fe83_row8_col4">0.293</td>
<td class="data row8 col5" id="T_1fe83_row8_col5">0.889</td>
<td class="data row8 col6" id="T_1fe83_row8_col6">0.500</td>
<td class="data row8 col7" id="T_1fe83_row8_col7">0.950</td>
<td class="data row8 col8" id="T_1fe83_row8_col8">0.000</td>
</tr>
</tbody>
</table></div>
</div>
<section id="Save-kinase-distance-matrix">
<h4 id="Save-kinase-distance-matrix">Save kinase distance matrix<a class="headerlink" href="#Save-kinase-distance-matrix" title="Permalink to this heading">¶</a></h4>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[38]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">kinase_distance_matrix_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">DATA</span> <span class="o">/</span> <span class="s2">"kinase_distance_matrix.csv"</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
</section>
</section>
<section id="Discussion">
<h2 id="Discussion">Discussion<a class="headerlink" href="#Discussion" title="Permalink to this heading">¶</a></h2>
<p>In this talktorial, we investigate how activity data can be used as a measure of similarity between kinases. The fraction of compounds tested as actives over the total number of measured compounds is a way of accessing the similarity. Moreover, using the same rationale, the promiscuity of a kinase can be quantified using the ratio of active compounds over measured compounds.</p>
<p>When working with these data, we have to keep in mind that some kinases have much higher coverage with respect to the number of compounds that were tested against them, leading to an imbalance in information content. This cannot be inferred from the calculated fraction. For example, the pairs EGFR-KDR and EGFR-p38a both have a profile similarity of <span class="math notranslate nohighlight">\(0.35\)</span>. However, the first was calculated based on <span class="math notranslate nohighlight">\(893\)</span> tested compounds, whereas the latter on <span class="math notranslate nohighlight">\(52\)</span> only.</p>
<p>The kinase distance matrix above will be reloaded in <strong>Talktorial T028</strong>, where we compare kinase similarities from different perspectives, including the ligand profile perspective we have talked about in this talktorial.</p>
</section>
<section id="Quiz">
<h2 id="Quiz">Quiz<a class="headerlink" href="#Quiz" title="Permalink to this heading">¶</a></h2>
<ol class="arabic simple">
<li><p>Is there an optimal way to deal with multiple kinase-ligand measurements?</p></li>
<li><p>Can promiscuity be fairly compared between two kinases if one has been tested against many compounds whereas the other only against very few?</p></li>
<li><p>Using the similarity described in this talktorial, what does it mean that two kinases have a similarity of <span class="math notranslate nohighlight">\(0\)</span>, as is the case for p110a and LCK?</p></li>
</ol>
<script type="application/vnd.jupyter.widget-state+json">
{"state": {}, "version_major": 2, "version_minor": 0}
</script></section>
</section>


          </article>
        </div>
      </div>
    </main>
  </div>
  <footer class="md-footer">
    <div class="md-footer-nav">
      <nav class="md-footer-nav__inner md-grid">
          
            <a href="T026_kinase_similarity_ifp.html" title="T026 · Kinase similarity: Interaction fingerprints"
               class="md-flex md-footer-nav__link md-footer-nav__link--prev"
               rel="prev">
              <div class="md-flex__cell md-flex__cell--shrink">
                <i class="md-icon md-icon--arrow-back md-footer-nav__button"></i>
              </div>
              <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
                <span class="md-flex__ellipsis">
                  <span
                      class="md-footer-nav__direction"> Previous </span> T026 · Kinase similarity: Interaction fingerprints </span>
              </div>
            </a>
          
          
            <a href="T028_kinase_similarity_compare_perspectives.html" title="T028 · Kinase similarity: Compare different perspectives"
               class="md-flex md-footer-nav__link md-footer-nav__link--next"
               rel="next">
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title"><span
                class="md-flex__ellipsis"> <span
                class="md-footer-nav__direction"> Next </span> T028 · Kinase similarity: Compare different perspectives </span>
            </div>
            <div class="md-flex__cell md-flex__cell--shrink"><i
                class="md-icon md-icon--arrow-forward md-footer-nav__button"></i>
            </div>
          
        </a>
        
      </nav>
    </div>
    <div class="md-footer-meta md-typeset">
      <div class="md-footer-meta__inner md-grid">
        <div class="md-footer-copyright">
          <div class="md-footer-copyright__highlight">
              &#169; Copyright 2018-2023, Volkamer Lab. Project structure based on the Computational Molecular Science Python Cookiecutter version 1.1.
              
          </div>
            Created using
            <a href="http://www.sphinx-doc.org/">Sphinx</a> 6.0.0.
             and
            <a href="https://github.com/bashtage/sphinx-material/">Material for
              Sphinx</a>
        </div>
      </div>
    </div>
  </footer>
  <script src="../_static/javascripts/application.js"></script>
  <script>app.initialize({version: "1.0.4", url: {base: ".."}})</script>
  </body>
</html>