
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />
<meta property="og:title" content="T028 · Kinase similarity: Compare different perspectives" />
<meta property="og:type" content="website" />
<meta property="og:url" content="https://projects.volkamerlab.org/teachopencadd/talktorials/T028_kinase_similarity_compare_perspectives.html" />
<meta property="og:site_name" content="TeachOpenCADD" />
<meta property="og:description" content="Note: This talktorial is a part of TeachOpenCADD, a platform that aims to teach domain-specific skills and to provide pipeline templates as starting points for research projects. Authors: Talia B. Kimber, 2021, Volkamer lab, Charité, Dominique Sydow, 2021, Volkamer lab, Charité, Andrea Volkamer, ..." />
<meta property="og:image" content="https://raw.githubusercontent.com/volkamerlab/teachopencadd/master/docs/_static/images/TeachOpenCADD_topics.png" />
<meta property="og:image:alt" content="TeachOpenCADD" />
<meta name="description" content="Note: This talktorial is a part of TeachOpenCADD, a platform that aims to teach domain-specific skills and to provide pipeline templates as starting points for research projects. Authors: Talia B. Kimber, 2021, Volkamer lab, Charité, Dominique Sydow, 2021, Volkamer lab, Charité, Andrea Volkamer, ..." />

  <meta name="viewport" content="width=device-width,initial-scale=1">
  <meta http-equiv="x-ua-compatible" content="ie=edge">
  <meta name="lang:clipboard.copy" content="Copy to clipboard">
  <meta name="lang:clipboard.copied" content="Copied to clipboard">
  <meta name="lang:search.language" content="en">
  <meta name="lang:search.pipeline.stopwords" content="True">
  <meta name="lang:search.pipeline.trimmer" content="True">
  <meta name="lang:search.result.none" content="No matching documents">
  <meta name="lang:search.result.one" content="1 matching document">
  <meta name="lang:search.result.other" content="# matching documents">
  <meta name="lang:search.tokenizer" content="[\s\-]+">

  
    <link href="https://fonts.gstatic.com/" rel="preconnect" crossorigin>
    <link href="https://fonts.googleapis.com/css?family=Roboto+Mono:400,500,700|Roboto:300,400,400i,700&display=fallback" rel="stylesheet">

    <style>
      body,
      input {
        font-family: "Roboto", "Helvetica Neue", Helvetica, Arial, sans-serif
      }

      code,
      kbd,
      pre {
        font-family: "Roboto Mono", "Courier New", Courier, monospace
      }
    </style>
  

  <link rel="stylesheet" href="../_static/stylesheets/application.css"/>
  <link rel="stylesheet" href="../_static/stylesheets/application-palette.css"/>
  <link rel="stylesheet" href="../_static/stylesheets/application-fixes.css"/>
  
  <link rel="stylesheet" href="../_static/fonts/material-icons.css"/>
  
  <meta name="theme-color" content="#2196f3">
  <script src="../_static/javascripts/modernizr.js"></script>
  
    <script async src="../_static/cookieconsent.min.js"></script>
    <script>
        window.addEventListener("load", function(){
        window.cookieconsent.initialise({
            "palette": {
            "popup": {
                "background": "#f0f0f0",
                "text": "#999"
            },
            "button": {
                "text": "#fff",
                "background": "#009688"
            }
            },
            "theme": "classic"
        })});
    </script>
    <!-- Global site tag (gtag.js) - Google Analytics -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=G-Q6ZE82CNZB"></script>
    <script>
        window.dataLayer = window.dataLayer || [];
        function gtag(){dataLayer.push(arguments);}
        gtag('js', new Date());

        gtag('config', 'G-Q6ZE82CNZB');
    </script>
  
    <link rel="apple-touch-icon" href="../_static/images/apple-icon-152x152.png"/>
  
  
    <title>T028 · Kinase similarity: Compare different perspectives &#8212; TeachOpenCADD 0 documentation</title>
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../_static/material.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/custom.css" />
    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/sphinx_highlight.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="https://unpkg.com/@jupyter-widgets/html-manager@^0.20.1/dist/embed-amd.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Talktorials by collection" href="../talktorials.html" />
    <link rel="prev" title="T027 · Kinase similarity: Ligand profile" href="T027_kinase_similarity_ligand_profile.html" />
  
    <link rel="apple-touch-icon" href="../_static/images/apple-icon-152x152.png"/>
  
   

  </head>
  <body dir=ltr
        data-md-color-primary=teal data-md-color-accent=cyan>
  
  <svg class="md-svg">
    <defs data-children-count="0">
      
      <svg xmlns="http://www.w3.org/2000/svg" width="416" height="448" viewBox="0 0 416 448" id="__github"><path fill="currentColor" d="M160 304q0 10-3.125 20.5t-10.75 19T128 352t-18.125-8.5-10.75-19T96 304t3.125-20.5 10.75-19T128 256t18.125 8.5 10.75 19T160 304zm160 0q0 10-3.125 20.5t-10.75 19T288 352t-18.125-8.5-10.75-19T256 304t3.125-20.5 10.75-19T288 256t18.125 8.5 10.75 19T320 304zm40 0q0-30-17.25-51T296 232q-10.25 0-48.75 5.25Q229.5 240 208 240t-39.25-2.75Q130.75 232 120 232q-29.5 0-46.75 21T56 304q0 22 8 38.375t20.25 25.75 30.5 15 35 7.375 37.25 1.75h42q20.5 0 37.25-1.75t35-7.375 30.5-15 20.25-25.75T360 304zm56-44q0 51.75-15.25 82.75-9.5 19.25-26.375 33.25t-35.25 21.5-42.5 11.875-42.875 5.5T212 416q-19.5 0-35.5-.75t-36.875-3.125-38.125-7.5-34.25-12.875T37 371.5t-21.5-28.75Q0 312 0 260q0-59.25 34-99-6.75-20.5-6.75-42.5 0-29 12.75-54.5 27 0 47.5 9.875t47.25 30.875Q171.5 96 212 96q37 0 70 8 26.25-20.5 46.75-30.25T376 64q12.75 25.5 12.75 54.5 0 21.75-6.75 42 34 40 34 99.5z"/></svg>
      
    </defs>
  </svg>
  
  <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer">
  <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search">
  <label class="md-overlay" data-md-component="overlay" for="__drawer"></label>
  <a href="#talktorials/T028_kinase_similarity_compare_perspectives" tabindex="1" class="md-skip"> Skip to content </a>
  <header class="md-header" data-md-component="header">
  <nav class="md-header-nav md-grid">
    <div class="md-flex navheader">
      <div class="md-flex__cell md-flex__cell--shrink">
        <a href="../index.html" title="TeachOpenCADD 0 documentation"
           class="md-header-nav__button md-logo">
          
            <i class="md-icon">school</i>
          
        </a>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        <label class="md-icon md-icon--menu md-header-nav__button" for="__drawer"></label>
      </div>
      <div class="md-flex__cell md-flex__cell--stretch">
        <div class="md-flex__ellipsis md-header-nav__title" data-md-component="title">
          <span class="md-header-nav__topic">TeachOpenCADD</span>
          <span class="md-header-nav__topic"> T028 · Kinase similarity: Compare different perspectives </span>
        </div>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        <label class="md-icon md-icon--search md-header-nav__button" for="__search"></label>
        
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" action="../search.html" method="get" name="search">
      <input type="text" class="md-search__input" name="q" placeholder="Search"
             autocapitalize="off" autocomplete="off" spellcheck="false"
             data-md-component="query" data-md-state="active">
      <label class="md-icon md-search__icon" for="__search"></label>
      <button type="reset" class="md-icon md-search__icon" data-md-component="reset" tabindex="-1">
        &#xE5CD;
      </button>
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="result">
          <div class="md-search-result__meta">
            Type to start searching
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>

      </div>
      
        <div class="md-flex__cell md-flex__cell--shrink">
          <div class="md-header-nav__source">
            <a href="https://github.com/volkamerlab/teachopencadd/" title="Go to repository" class="md-source" data-md-source="github">

    <div class="md-source__icon">
      <svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 24 24" width="28" height="28">
        <use xlink:href="#__github" width="24" height="24"></use>
      </svg>
    </div>
  
  <div class="md-source__repository">
    TeachOpenCADD
  </div>
</a>
          </div>
        </div>
      
      
    </div>
  </nav>
</header>

  
  <div class="md-container">
    
    
    
  <nav class="md-tabs" data-md-component="tabs">
    <div class="md-tabs__inner md-grid">
      <ul class="md-tabs__list">
            
            <li class="md-tabs__item"><a href="../talktorials.html" class="md-tabs__link">Our talktorials</a></li>
            
            <li class="md-tabs__item"><a href="../installing.html" class="md-tabs__link">Run locally</a></li>
            
            <li class="md-tabs__item"><a href="../contribute.html" class="md-tabs__link">Development</a></li>
            
            <li class="md-tabs__item"><a href="../contact.html" class="md-tabs__link">Contact</a></li>
            
            <li class="md-tabs__item"><a href="../citation.html" class="md-tabs__link">Citation</a></li>
          <li class="md-tabs__item"><a href="../all_talktorials.html" class="md-tabs__link">Complete list of talktorials</a></li>
      </ul>
    </div>
  </nav>
    <main class="md-main">
      <div class="md-main__inner md-grid" data-md-component="container">
        
          <div class="md-sidebar md-sidebar--primary" data-md-component="navigation">
            <div class="md-sidebar__scrollwrap">
              <div class="md-sidebar__inner">
                <nav class="md-nav md-nav--primary" data-md-level="0">
  <label class="md-nav__title md-nav__title--site" for="__drawer">
    <a href="../index.html" title="TeachOpenCADD 0 documentation" class="md-nav__button md-logo">
      
        <i class="md-icon">school</i>
      
    </a>
    <a href="../index.html"
       title="TeachOpenCADD 0 documentation">TeachOpenCADD</a>
  </label>
    <div class="md-nav__source">
      <a href="https://github.com/volkamerlab/teachopencadd/" title="Go to repository" class="md-source" data-md-source="github">

    <div class="md-source__icon">
      <svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 24 24" width="28" height="28">
        <use xlink:href="#__github" width="24" height="24"></use>
      </svg>
    </div>
  
  <div class="md-source__repository">
    TeachOpenCADD
  </div>
</a>
    </div>
  
  

  
  <ul class="md-nav__list">
    <li class="md-nav__item">
    
      <span class="md-nav__link caption"><span class="caption-text">Our talktorials</span></span>
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="../all_talktorials.html" class="md-nav__link">Complete list of talktorials</a>
      <ul class="md-nav__list"> 
    <li class="md-nav__item">
    
    
      <a href="T001_query_chembl.html" class="md-nav__link">T001 · Compound data acquisition (ChEMBL)</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T002_compound_adme.html" class="md-nav__link">T002 · Molecular filtering: ADME and lead-likeness criteria</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T003_compound_unwanted_substructures.html" class="md-nav__link">T003 · Molecular filtering: unwanted substructures</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T004_compound_similarity.html" class="md-nav__link">T004 · Ligand-based screening: compound similarity</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T005_compound_clustering.html" class="md-nav__link">T005 · Compound clustering</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T006_compound_maximum_common_substructures.html" class="md-nav__link">T006 · Maximum common substructure</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T007_compound_activity_machine_learning.html" class="md-nav__link">T007 · Ligand-based screening: machine learning</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T008_query_pdb.html" class="md-nav__link">T008 · Protein data acquisition: Protein Data Bank (PDB)</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T009_compound_ensemble_pharmacophores.html" class="md-nav__link">T009 · Ligand-based pharmacophores</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T010_binding_site_comparison.html" class="md-nav__link">T010 · Binding site similarity and off-target prediction</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T011_query_online_api_webservices.html" class="md-nav__link">T011 · Querying online API webservices</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T012_query_klifs.html" class="md-nav__link">T012 · Data acquisition from KLIFS</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T013_query_pubchem.html" class="md-nav__link">T013 · Data acquisition from PubChem</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T014_binding_site_detection.html" class="md-nav__link">T014 · Binding site detection</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T015_protein_ligand_docking.html" class="md-nav__link">T015 · Protein ligand docking</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T016_protein_ligand_interactions.html" class="md-nav__link">T016 · Protein-ligand interactions</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T017_advanced_nglview_usage.html" class="md-nav__link">T017 · Advanced NGLview usage</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T018_automated_cadd_pipeline.html" class="md-nav__link">T018 · Automated pipeline for lead optimization</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T019_md_simulation.html" class="md-nav__link">T019 · Molecular dynamics simulation</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T020_md_analysis.html" class="md-nav__link">T020 · Analyzing molecular dynamics simulations</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T021_one_hot_encoding.html" class="md-nav__link">T021 · One-Hot Encoding</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T022_ligand_based_screening_neural_network.html" class="md-nav__link">T022 · Ligand-based screening: neural networks</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T023_what_is_a_kinase.html" class="md-nav__link">T023 · What is a kinase?</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T024_kinase_similarity_sequence.html" class="md-nav__link">T024 · Kinase similarity: Sequence</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T025_kinase_similarity_kissim.html" class="md-nav__link">T025 · Kinase similarity: Kinase pocket (KiSSim fingerprint)</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T026_kinase_similarity_ifp.html" class="md-nav__link">T026 · Kinase similarity: Interaction fingerprints</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="T027_kinase_similarity_ligand_profile.html" class="md-nav__link">T027 · Kinase similarity: Ligand profile</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
    <input class="md-toggle md-nav__toggle" data-md-toggle="toc" type="checkbox" id="__toc">
    <label class="md-nav__link md-nav__link--active" for="__toc"> T028 · Kinase similarity: Compare different perspectives </label>
    
      <a href="#" class="md-nav__link md-nav__link--active">T028 · Kinase similarity: Compare different perspectives</a>
      
        
<nav class="md-nav md-nav--secondary">
    <label class="md-nav__title" for="__toc">Contents</label>
  <ul class="md-nav__list" data-md-scrollfix="">
        <li class="md-nav__item"><a href="#talktorials-t028-kinase-similarity-compare-perspectives--page-root" class="md-nav__link">T028 · Kinase similarity: Compare different perspectives</a><nav class="md-nav">
              <ul class="md-nav__list">
        <li class="md-nav__item"><a href="#Aim-of-this-talktorial" class="md-nav__link">Aim of this talktorial</a><nav class="md-nav">
              <ul class="md-nav__list">
        <li class="md-nav__item"><a href="#Contents-in-Theory" class="md-nav__link">Contents in <em>Theory</em></a>
        </li>
        <li class="md-nav__item"><a href="#Contents-in-Practical" class="md-nav__link">Contents in <em>Practical</em></a>
        </li>
        <li class="md-nav__item"><a href="#References" class="md-nav__link">References</a>
        </li></ul>
            </nav>
        </li>
        <li class="md-nav__item"><a href="#Theory" class="md-nav__link">Theory</a><nav class="md-nav">
              <ul class="md-nav__list">
        <li class="md-nav__item"><a href="#Kinase-dataset" class="md-nav__link">Kinase dataset</a>
        </li>
        <li class="md-nav__item"><a href="#Kinase-similarity-descriptor-(considering-4-different-methods)" class="md-nav__link">Kinase similarity descriptor (considering 4 different methods)</a>
        </li>
        <li class="md-nav__item"><a href="#Distance-matrix-conditions" class="md-nav__link">Distance matrix conditions</a>
        </li></ul>
            </nav>
        </li>
        <li class="md-nav__item"><a href="#Practical" class="md-nav__link">Practical</a><nav class="md-nav">
              <ul class="md-nav__list">
        <li class="md-nav__item"><a href="#Load-kinase-similarity-and-distance-matrices" class="md-nav__link">Load kinase similarity and distance matrices</a>
        </li>
        <li class="md-nav__item"><a href="#id1" class="md-nav__link">Distance matrix conditions</a>
        </li>
        <li class="md-nav__item"><a href="#Visualize-similarity-for-example-perspective" class="md-nav__link">Visualize similarity for example perspective</a><nav class="md-nav">
              <ul class="md-nav__list">
        <li class="md-nav__item"><a href="#Visualize-kinase-similarity-matrix-as-heatmap" class="md-nav__link">Visualize kinase similarity matrix as heatmap</a>
        </li>
        <li class="md-nav__item"><a href="#Visualize-similarity-as-dendrogram" class="md-nav__link">Visualize similarity as dendrogram</a>
        </li></ul>
            </nav>
        </li>
        <li class="md-nav__item"><a href="#Visualize-similarities-from-the-four-different-perspectives" class="md-nav__link">Visualize similarities from the four different perspectives</a><nav class="md-nav">
              <ul class="md-nav__list">
        <li class="md-nav__item"><a href="#Preprocess-distance-matrices" class="md-nav__link">Preprocess distance matrices</a><nav class="md-nav">
              <ul class="md-nav__list">
        <li class="md-nav__item"><a href="#Normalize-matrices" class="md-nav__link">Normalize matrices</a>
        </li>
        <li class="md-nav__item"><a href="#Define-kinase-order" class="md-nav__link">Define kinase order</a>
        </li></ul>
            </nav>
        </li>
        <li class="md-nav__item"><a href="#Visualize-kinase-similarities" class="md-nav__link">Visualize kinase similarities</a>
        </li>
        <li class="md-nav__item"><a href="#Analysis-of-results" class="md-nav__link">Analysis of results</a>
        </li></ul>
            </nav>
        </li></ul>
            </nav>
        </li>
        <li class="md-nav__item"><a href="#Discussion" class="md-nav__link">Discussion</a>
        </li>
        <li class="md-nav__item"><a href="#Quiz" class="md-nav__link">Quiz</a>
        </li>
        <li class="md-nav__item"><a href="#Appendix" class="md-nav__link">Appendix</a><nav class="md-nav">
              <ul class="md-nav__list">
        <li class="md-nav__item"><a href="#One-single-measure-of-comparison" class="md-nav__link">One single measure of comparison</a>
        </li></ul>
            </nav>
        </li></ul>
            </nav>
        </li>
    
<li class="md-nav__item"><a class="md-nav__extra_link" href="../_sources/talktorials/T028_kinase_similarity_compare_perspectives.nblink.txt">Show Source</a> </li>

  </ul>
</nav>
      
    
    </li></ul>
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="../talktorials.html" class="md-nav__link">Talktorials by collection</a>
      
    
    </li>
    <li class="md-nav__item">
    
      <span class="md-nav__link caption"><span class="caption-text">Run locally</span></span>
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="../installing.html" class="md-nav__link">Installing</a>
      
    
    </li>
    <li class="md-nav__item">
    
      <span class="md-nav__link caption"><span class="caption-text">Development</span></span>
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="../contribute.html" class="md-nav__link">For contributors</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="../api.html" class="md-nav__link">API Documentation</a>
      
    
    </li>
    <li class="md-nav__item">
    
      <span class="md-nav__link caption"><span class="caption-text">About TeachOpenCADD</span></span>
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="../contact.html" class="md-nav__link">Contact</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="../acknowledgments.html" class="md-nav__link">Acknowledgments</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="../citation.html" class="md-nav__link">Citation</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="../license.html" class="md-nav__link">License</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="../funding.html" class="md-nav__link">Funding</a>
      
    
    </li>
    <li class="md-nav__item">
    
      <span class="md-nav__link caption"><span class="caption-text">External resources</span></span>
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="../external_dependencies.html" class="md-nav__link">Packages and webservers used in TeachOpenCADD</a>
      
    
    </li>
    <li class="md-nav__item">
    
    
      <a href="../external_tutorials_collections.html" class="md-nav__link">External tutorials and collections</a>
      
    
    </li>
  </ul>
  

</nav>
              </div>
            </div>
          </div>
          <div class="md-sidebar md-sidebar--secondary" data-md-component="toc">
            <div class="md-sidebar__scrollwrap">
              <div class="md-sidebar__inner">
                
<nav class="md-nav md-nav--secondary">
    <label class="md-nav__title" for="__toc">Contents</label>
  <ul class="md-nav__list" data-md-scrollfix="">
        <li class="md-nav__item"><a href="#talktorials-t028-kinase-similarity-compare-perspectives--page-root" class="md-nav__link">T028 · Kinase similarity: Compare different perspectives</a><nav class="md-nav">
              <ul class="md-nav__list">
        <li class="md-nav__item"><a href="#Aim-of-this-talktorial" class="md-nav__link">Aim of this talktorial</a><nav class="md-nav">
              <ul class="md-nav__list">
        <li class="md-nav__item"><a href="#Contents-in-Theory" class="md-nav__link">Contents in <em>Theory</em></a>
        </li>
        <li class="md-nav__item"><a href="#Contents-in-Practical" class="md-nav__link">Contents in <em>Practical</em></a>
        </li>
        <li class="md-nav__item"><a href="#References" class="md-nav__link">References</a>
        </li></ul>
            </nav>
        </li>
        <li class="md-nav__item"><a href="#Theory" class="md-nav__link">Theory</a><nav class="md-nav">
              <ul class="md-nav__list">
        <li class="md-nav__item"><a href="#Kinase-dataset" class="md-nav__link">Kinase dataset</a>
        </li>
        <li class="md-nav__item"><a href="#Kinase-similarity-descriptor-(considering-4-different-methods)" class="md-nav__link">Kinase similarity descriptor (considering 4 different methods)</a>
        </li>
        <li class="md-nav__item"><a href="#Distance-matrix-conditions" class="md-nav__link">Distance matrix conditions</a>
        </li></ul>
            </nav>
        </li>
        <li class="md-nav__item"><a href="#Practical" class="md-nav__link">Practical</a><nav class="md-nav">
              <ul class="md-nav__list">
        <li class="md-nav__item"><a href="#Load-kinase-similarity-and-distance-matrices" class="md-nav__link">Load kinase similarity and distance matrices</a>
        </li>
        <li class="md-nav__item"><a href="#id1" class="md-nav__link">Distance matrix conditions</a>
        </li>
        <li class="md-nav__item"><a href="#Visualize-similarity-for-example-perspective" class="md-nav__link">Visualize similarity for example perspective</a><nav class="md-nav">
              <ul class="md-nav__list">
        <li class="md-nav__item"><a href="#Visualize-kinase-similarity-matrix-as-heatmap" class="md-nav__link">Visualize kinase similarity matrix as heatmap</a>
        </li>
        <li class="md-nav__item"><a href="#Visualize-similarity-as-dendrogram" class="md-nav__link">Visualize similarity as dendrogram</a>
        </li></ul>
            </nav>
        </li>
        <li class="md-nav__item"><a href="#Visualize-similarities-from-the-four-different-perspectives" class="md-nav__link">Visualize similarities from the four different perspectives</a><nav class="md-nav">
              <ul class="md-nav__list">
        <li class="md-nav__item"><a href="#Preprocess-distance-matrices" class="md-nav__link">Preprocess distance matrices</a><nav class="md-nav">
              <ul class="md-nav__list">
        <li class="md-nav__item"><a href="#Normalize-matrices" class="md-nav__link">Normalize matrices</a>
        </li>
        <li class="md-nav__item"><a href="#Define-kinase-order" class="md-nav__link">Define kinase order</a>
        </li></ul>
            </nav>
        </li>
        <li class="md-nav__item"><a href="#Visualize-kinase-similarities" class="md-nav__link">Visualize kinase similarities</a>
        </li>
        <li class="md-nav__item"><a href="#Analysis-of-results" class="md-nav__link">Analysis of results</a>
        </li></ul>
            </nav>
        </li></ul>
            </nav>
        </li>
        <li class="md-nav__item"><a href="#Discussion" class="md-nav__link">Discussion</a>
        </li>
        <li class="md-nav__item"><a href="#Quiz" class="md-nav__link">Quiz</a>
        </li>
        <li class="md-nav__item"><a href="#Appendix" class="md-nav__link">Appendix</a><nav class="md-nav">
              <ul class="md-nav__list">
        <li class="md-nav__item"><a href="#One-single-measure-of-comparison" class="md-nav__link">One single measure of comparison</a>
        </li></ul>
            </nav>
        </li></ul>
            </nav>
        </li>
    
<li class="md-nav__item"><a class="md-nav__extra_link" href="../_sources/talktorials/T028_kinase_similarity_compare_perspectives.nblink.txt">Show Source</a> </li>

<li id="searchbox" class="md-nav__item"></li>

  </ul>
</nav>
              </div>
            </div>
          </div>
        
        <div class="md-content">
          <article class="md-content__inner md-typeset" role="main">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars and line breaks on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
    white-space: pre;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt .copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
.jp-RenderedHTMLCommon table,
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
.jp-RenderedHTMLCommon thead,
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
.jp-RenderedHTMLCommon tr,
.jp-RenderedHTMLCommon th,
.jp-RenderedHTMLCommon td,
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
.jp-RenderedHTMLCommon th,
div.rendered_html th {
  font-weight: bold;
}
.jp-RenderedHTMLCommon tbody tr:nth-child(odd),
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
.jp-RenderedHTMLCommon tbody tr:hover,
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}
</style>
<section id="T028-·-Kinase-similarity:-Compare-different-perspectives">
<h1 id="talktorials-t028-kinase-similarity-compare-perspectives--page-root">T028 · Kinase similarity: Compare different perspectives<a class="headerlink" href="#talktorials-t028-kinase-similarity-compare-perspectives--page-root" title="Permalink to this heading">¶</a></h1>
<p><strong>Note:</strong> This talktorial is a part of TeachOpenCADD, a platform that aims to teach domain-specific skills and to provide pipeline templates as starting points for research projects.</p>
<p>Authors:</p>
<ul class="simple">
<li><p>Talia B. Kimber, 2021, <a class="reference external" href="https://volkamerlab.org/">Volkamer lab, Charité</a></p></li>
<li><p>Dominique Sydow, 2021, <a class="reference external" href="https://volkamerlab.org/">Volkamer lab, Charité</a></p></li>
<li><p>Andrea Volkamer, 2021, <a class="reference external" href="https://volkamerlab.org/">Volkamer lab, Charité</a></p></li>
</ul>
<section id="Aim-of-this-talktorial">
<h2 id="Aim-of-this-talktorial">Aim of this talktorial<a class="headerlink" href="#Aim-of-this-talktorial" title="Permalink to this heading">¶</a></h2>
<p>We will compare different perspectives on kinase similarity, which were discussed in detail in previous notebooks:</p>
<ul class="simple">
<li><p><strong>Talktorial T024</strong>: Kinase pocket sequences (KLIFS pocket sequences)</p></li>
<li><p><strong>Talktorial T025</strong>: Kinase pocket structures (KiSSim fingerprint based on KLIFS pocket residues)</p></li>
<li><p><strong>Talktorial T026</strong>: Kinase-ligand interaction profiles (KLIFS IFPs based on KLIFS pocket residues)</p></li>
<li><p><strong>Talktorial T027</strong>: Ligand profiling data (using ChEMBL29 bioactivity data)</p></li>
</ul>
<p><em>Note</em>: We focus only on similarities between orthosteric kinase binding sites; similarities to allosteric binding sites are not covered (T027 is an exception since the profiling data does not distinguish between binding sites).</p>
<section id="Contents-in-Theory">
<h3 id="Contents-in-Theory">Contents in <em>Theory</em><a class="headerlink" href="#Contents-in-Theory" title="Permalink to this heading">¶</a></h3>
<ul class="simple">
<li><p>Kinase dataset</p></li>
<li><p>Kinase similarity descriptor (considering 4 different methods)</p></li>
<li><p>Distance matrix conditions</p></li>
</ul>
</section>
<section id="Contents-in-Practical">
<h3 id="Contents-in-Practical">Contents in <em>Practical</em><a class="headerlink" href="#Contents-in-Practical" title="Permalink to this heading">¶</a></h3>
<ul class="simple">
<li><p>Load kinase similarity and distance matrices</p></li>
<li><p>Distance matrix conditions</p></li>
<li><p>Visualize similarity for example perspective</p>
<ul>
<li><p>Visualize kinase similarity matrix as heatmap</p></li>
<li><p>Visualize similarity as dendrogram</p></li>
</ul>
</li>
<li><p>Visualize similarities from the four different perspectives</p>
<ul>
<li><p>Preprocess distance matrices</p>
<ul>
<li><p>Normalize matrices</p></li>
<li><p>Define kinase order</p></li>
</ul>
</li>
<li><p>Visualize kinase similarities</p></li>
<li><p>Analysis of results</p></li>
</ul>
</li>
</ul>
</section>
<section id="References">
<h3 id="References">References<a class="headerlink" href="#References" title="Permalink to this heading">¶</a></h3>
<ul class="simple">
<li><p>Kinase dataset: <a class="reference external" href="https://www.mdpi.com/1420-3049/26/3/629">Molecules (2021), 26(3), 629</a></p></li>
<li><p>Clustering and dendrograms with <code class="docutils literal notranslate"><span class="pre">scipy</span></code>: <a class="reference external" href="https://docs.scipy.org/doc/scipy/reference/cluster.hierarchy.html">https://docs.scipy.org/doc/scipy/reference/cluster.hierarchy.html</a></p></li>
<li><p>Distance matrix:</p>
<ul>
<li><p><a class="reference external" href="https://en.wikipedia.org/wiki/Distance_matrix">https://en.wikipedia.org/wiki/Distance_matrix</a></p></li>
<li><p>Gilbert, A. C. and Jain, L. “If it ain’t broke, don’t fix it: Sparse metric repair.” <em>2017 55th Annual Allerton Conference on Communication, Control, and Computing (Allerton)</em>. IEEE, 2017. <a class="reference external" href="https://arxiv.org/pdf/1710.10655.pdf">https://arxiv.org/pdf/1710.10655.pdf</a></p></li>
</ul>
</li>
</ul>
</section>
</section>
<section id="Theory">
<h2 id="Theory">Theory<a class="headerlink" href="#Theory" title="Permalink to this heading">¶</a></h2>
<section id="Kinase-dataset">
<h3 id="Kinase-dataset">Kinase dataset<a class="headerlink" href="#Kinase-dataset" title="Permalink to this heading">¶</a></h3>
<p>We use the kinase selection as defined in <strong>Talktorial T023</strong>.</p>
</section>
<section id="Kinase-similarity-descriptor-(considering-4-different-methods)">
<h3 id="Kinase-similarity-descriptor-(considering-4-different-methods)">Kinase similarity descriptor (considering 4 different methods)<a class="headerlink" href="#Kinase-similarity-descriptor-(considering-4-different-methods)" title="Permalink to this heading">¶</a></h3>
<ul class="simple">
<li><p><strong>Talktorial T024</strong> = KLIFS pocket sequence</p></li>
<li><p><strong>Talktorial T025</strong> = KiSSim fingerprint</p></li>
<li><p><strong>Talktorial T026</strong> = KLIFS interaction fingerprint</p></li>
<li><p><strong>Talktorial T027</strong> = Ligand profile: ChEMBL29, bioactivity</p></li>
</ul>
<p>Please refer to the talktorials for more details on each method.</p>
</section>
<section id="Distance-matrix-conditions">
<h3 id="Distance-matrix-conditions">Distance matrix conditions<a class="headerlink" href="#Distance-matrix-conditions" title="Permalink to this heading">¶</a></h3>
<p>A square, real matrix <span class="math notranslate nohighlight">\(M \in \mathbb{R}^{n \times n}\)</span> is a distance matrix if <span class="math notranslate nohighlight">\(M\)</span> satisfies the following conditions (following the definition of distance) :</p>
<ol class="arabic simple">
<li><ol class="loweralpha simple">
<li><p>Positivity: <span class="math notranslate nohighlight">\(M[x, y] \geq 0\)</span> for all <span class="math notranslate nohighlight">\(x, y\)</span> and b. Null condition: <span class="math notranslate nohighlight">\(M[x, y] = 0 \iff x = y.\)</span></p></li>
</ol>
</li>
<li><p>Symmetry: <span class="math notranslate nohighlight">\(M[x, y] = M[y, x]\)</span> for all <span class="math notranslate nohighlight">\(x, y.\)</span></p></li>
<li><p>Triangular inequality: <span class="math notranslate nohighlight">\(M[x, z] \leq M[x, y] + M[y, z]\)</span> for all <span class="math notranslate nohighlight">\(x, y, z.\)</span></p></li>
</ol>
<p><em>Technical note</em>: In some cases, the null condition can be relaxed and only the following has to be satisfied: <span class="math notranslate nohighlight">\(M[x, x] = 0, \forall x.\)</span> In this case, the term is known as semi-metric. We will use this condition in this notebook.</p>
</section>
</section>
<section id="Practical">
<h2 id="Practical">Practical<a class="headerlink" href="#Practical" title="Permalink to this heading">¶</a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">pathlib</span> <span class="kn">import</span> <span class="n">Path</span>

<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>
<span class="kn">from</span> <span class="nn">scipy.spatial</span> <span class="kn">import</span> <span class="n">distance_matrix</span><span class="p">,</span> <span class="n">distance</span>
<span class="kn">from</span> <span class="nn">scipy.cluster</span> <span class="kn">import</span> <span class="n">hierarchy</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">HERE</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">_dh</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
<span class="n">DATA</span> <span class="o">=</span> <span class="n">HERE</span> <span class="o">/</span> <span class="s2">"data"</span>
</pre></div>
</div>
</div>
<section id="Load-kinase-similarity-and-distance-matrices">
<h3 id="Load-kinase-similarity-and-distance-matrices">Load kinase similarity and distance matrices<a class="headerlink" href="#Load-kinase-similarity-and-distance-matrices" title="Permalink to this heading">¶</a></h3>
<p>We define the paths to the kinase distance matrices.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">kinase_matrix_paths</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">"sequence"</span><span class="p">:</span> <span class="s2">"T024_kinase_similarity_sequence"</span><span class="p">,</span>
    <span class="s2">"kissim"</span><span class="p">:</span> <span class="s2">"T025_kinase_similarity_kissim"</span><span class="p">,</span>
    <span class="s2">"ifp"</span><span class="p">:</span> <span class="s2">"T026_kinase_similarity_ifp"</span><span class="p">,</span>
    <span class="s2">"ligand-profile"</span><span class="p">:</span> <span class="s2">"T027_kinase_similarity_ligand_profile"</span><span class="p">,</span>
<span class="p">}</span>

<span class="c1"># Distance matrices (for dendrogram)</span>
<span class="n">kinase_distance_matrix_paths</span> <span class="o">=</span> <span class="p">{</span>
    <span class="n">perspective</span><span class="p">:</span> <span class="n">HERE</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">"../</span><span class="si">{</span><span class="n">folder_name</span><span class="si">}</span><span class="s2">/data/kinase_distance_matrix.csv"</span>
    <span class="k">for</span> <span class="n">perspective</span><span class="p">,</span> <span class="n">folder_name</span> <span class="ow">in</span> <span class="n">kinase_matrix_paths</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
<span class="p">}</span>
</pre></div>
</div>
</div>
<p>We load the distance matrices which were generated in the previous notebooks.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">kinase_distance_matrices</span> <span class="o">=</span> <span class="p">{}</span>
<span class="k">for</span> <span class="n">kinsim_perspective</span><span class="p">,</span> <span class="n">path</span> <span class="ow">in</span> <span class="n">kinase_distance_matrix_paths</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
    <span class="n">kinase_distance_matrices</span><span class="p">[</span><span class="n">kinsim_perspective</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">6</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="id1">
<h3 id="id1">Distance matrix conditions<a class="headerlink" href="#id1" title="Permalink to this heading">¶</a></h3>
<p>We now create functions to verify the distance conditions.</p>
<p>We check the dimension of the matrix: it should be a 2D, square array.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">check_dimensionality</span><span class="p">(</span><span class="n">matrix</span><span class="p">):</span>
<span class="w">    </span><span class="sd">"""</span>
<span class="sd">    Checks whether the input is a 2D array (matrix) and square.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    matrix : np.array</span>
<span class="sd">        The matrix for which the condition should be checked.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    bool :</span>
<span class="sd">        True if the condition is met, False otherwise.</span>
<span class="sd">    """</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">2</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">"The input is not a matrix, but an array of shape </span><span class="si">{</span><span class="n">matrix</span><span class="o">.</span><span class="n">shape</span><span class="si">}</span><span class="s2">."</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="n">matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">"The input is not a square matrix. Failing."</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">True</span>
</pre></div>
</div>
</div>
<p>We check that all values of the matrix are positive.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">check_positivity</span><span class="p">(</span><span class="n">matrix</span><span class="p">):</span>
<span class="w">    </span><span class="sd">"""</span>
<span class="sd">    Checks whether all values of a matrix are positive.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    matrix : np.array</span>
<span class="sd">        The matrix for which the condition should be checked.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    bool :</span>
<span class="sd">        True if the condition is met, False otherwise.</span>
<span class="sd">    """</span>
    <span class="k">return</span> <span class="p">(</span><span class="n">matrix</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">all</span><span class="p">()</span>
</pre></div>
</div>
</div>
<p>We check that the diagonal values of the matrix are zero.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">check_null_diagonal</span><span class="p">(</span><span class="n">matrix</span><span class="p">):</span>
<span class="w">    </span><span class="sd">"""</span>
<span class="sd">    Checks whether the diagonal entries of a matrix are all zero.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    matrix : np.array</span>
<span class="sd">        The matrix for which the condition should be checked.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    bool :</span>
<span class="sd">        True if the condition is met, False otherwise.</span>
<span class="sd">    """</span>
    <span class="k">return</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">diagonal</span><span class="p">(</span><span class="n">matrix</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">all</span><span class="p">()</span>
</pre></div>
</div>
</div>
<p>We check that the matrix is symmetric.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">check_symmetry</span><span class="p">(</span><span class="n">matrix</span><span class="p">):</span>
<span class="w">    </span><span class="sd">"""</span>
<span class="sd">    Checks whether a matrix M is symmetric, i.e. M = M^T.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    matrix : np.array</span>
<span class="sd">        The matrix for which the condition should be checked.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    bool :</span>
<span class="sd">        True if the condition is met, False otherwise.</span>
<span class="sd">    """</span>
    <span class="k">return</span> <span class="p">(</span><span class="n">matrix</span> <span class="o">==</span> <span class="n">matrix</span><span class="o">.</span><span class="n">transpose</span><span class="p">())</span><span class="o">.</span><span class="n">all</span><span class="p">()</span>
</pre></div>
</div>
</div>
<p>Finally, we check the triangular inequality condition.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">check_triangular_inequality</span><span class="p">(</span><span class="n">matrix</span><span class="p">,</span> <span class="n">tolerance</span><span class="o">=</span><span class="mf">0.11</span><span class="p">):</span>
<span class="w">    </span><span class="sd">"""</span>
<span class="sd">    Checks whether the triangular inequality is met for a matrix,</span>
<span class="sd">    i.e. the shortest distance between two points is the straight line.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    matrix : np.array</span>
<span class="sd">        The matrix for which the condition should be checked.</span>
<span class="sd">    tolerance : float</span>
<span class="sd">        The accepted tolerance for approximation.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    bool :</span>
<span class="sd">        True if the condition is met, False otherwise.</span>
<span class="sd">    """</span>
    <span class="n">triang_inequality</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]):</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]):</span>
            <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]):</span>
                <span class="k">if</span> <span class="n">matrix</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">matrix</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">k</span><span class="p">]</span> <span class="o">+</span> <span class="n">matrix</span><span class="p">[</span><span class="n">k</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">+</span> <span class="n">tolerance</span><span class="p">:</span>
                    <span class="n">triang_inequality</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">return</span> <span class="kc">False</span>
    <span class="k">return</span> <span class="n">triang_inequality</span>
</pre></div>
</div>
</div>
<p>We now create a function that combines all these conditions.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">distance_condition</span><span class="p">(</span><span class="n">matrix</span><span class="p">):</span>
<span class="w">    </span><span class="sd">"""</span>
<span class="sd">    Checks if all conditions for a matrix are met.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    matrix : np.array</span>
<span class="sd">        The matrix for which the conditions should be checked.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    None</span>
<span class="sd">    """</span>

    <span class="n">dimensionality</span> <span class="o">=</span> <span class="n">check_dimensionality</span><span class="p">(</span><span class="n">matrix</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">dimensionality</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">"</span><span class="si">{</span><span class="s1">'Dimensionality'</span><span class="w"> </span><span class="si">:</span><span class="s2"> &lt;30</span><span class="si">}</span><span class="s2">"</span> <span class="sa">f</span><span class="s2">"</span><span class="si">{</span><span class="s1">'Satisfied'</span><span class="si">}</span><span class="s2">"</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">"</span><span class="si">{</span><span class="s1">'Dimensionality'</span><span class="w"> </span><span class="si">:</span><span class="s2"> &lt;30</span><span class="si">}</span><span class="s2">"</span> <span class="sa">f</span><span class="s2">"</span><span class="si">{</span><span class="s1">'Not Satisfied'</span><span class="si">}</span><span class="s2">"</span><span class="p">)</span>

    <span class="n">positivity</span> <span class="o">=</span> <span class="n">check_positivity</span><span class="p">(</span><span class="n">matrix</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">positivity</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">"</span><span class="si">{</span><span class="s1">'Postivity'</span><span class="w"> </span><span class="si">:</span><span class="s2"> &lt;30</span><span class="si">}</span><span class="s2">"</span> <span class="sa">f</span><span class="s2">"</span><span class="si">{</span><span class="s1">'Satisfied'</span><span class="si">}</span><span class="s2">"</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">"</span><span class="si">{</span><span class="s1">'Postivity'</span><span class="w"> </span><span class="si">:</span><span class="s2"> &lt;30</span><span class="si">}</span><span class="s2">"</span> <span class="sa">f</span><span class="s2">"</span><span class="si">{</span><span class="s1">'Satisfied'</span><span class="si">}</span><span class="s2">"</span><span class="p">)</span>

    <span class="n">diag</span> <span class="o">=</span> <span class="n">check_null_diagonal</span><span class="p">(</span><span class="n">matrix</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">diag</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">"</span><span class="si">{</span><span class="s1">'Null diagonal'</span><span class="w"> </span><span class="si">:</span><span class="s2"> &lt;30</span><span class="si">}</span><span class="s2">"</span> <span class="sa">f</span><span class="s2">"</span><span class="si">{</span><span class="s1">'Satisfied'</span><span class="si">}</span><span class="s2">"</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">"</span><span class="si">{</span><span class="s1">'Null diagonal'</span><span class="w"> </span><span class="si">:</span><span class="s2"> &lt;30</span><span class="si">}</span><span class="s2">"</span> <span class="sa">f</span><span class="s2">"</span><span class="si">{</span><span class="s1">'Not Satisfied'</span><span class="si">}</span><span class="s2">"</span><span class="p">)</span>

    <span class="n">symmetry</span> <span class="o">=</span> <span class="n">check_symmetry</span><span class="p">(</span><span class="n">matrix</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">symmetry</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">"</span><span class="si">{</span><span class="s1">'Symmetry'</span><span class="w"> </span><span class="si">:</span><span class="s2"> &lt;30</span><span class="si">}</span><span class="s2">"</span> <span class="sa">f</span><span class="s2">"</span><span class="si">{</span><span class="s1">'Satisfied'</span><span class="si">}</span><span class="s2">"</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">"</span><span class="si">{</span><span class="s1">'Symmetry'</span><span class="w"> </span><span class="si">:</span><span class="s2"> &lt;30</span><span class="si">}</span><span class="s2">"</span> <span class="sa">f</span><span class="s2">"</span><span class="si">{</span><span class="s1">'Satisfied'</span><span class="si">}</span><span class="s2">"</span><span class="p">)</span>

    <span class="n">triang_inequ</span> <span class="o">=</span> <span class="n">check_triangular_inequality</span><span class="p">(</span><span class="n">matrix</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">triang_inequ</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">"</span><span class="si">{</span><span class="s1">'Triangular inequality'</span><span class="w"> </span><span class="si">:</span><span class="s2"> &lt;30</span><span class="si">}</span><span class="s2">"</span> <span class="sa">f</span><span class="s2">"</span><span class="si">{</span><span class="s1">'Satisfied'</span><span class="si">}</span><span class="s2">"</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">"</span><span class="si">{</span><span class="s1">'Triangular inequality'</span><span class="w"> </span><span class="si">:</span><span class="s2"> &lt;30</span><span class="si">}</span><span class="s2">"</span> <span class="sa">f</span><span class="s2">"</span><span class="si">{</span><span class="s1">'Not Satisfied'</span><span class="si">}</span><span class="s2">"</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">"</span><span class="se">\n</span><span class="s2">"</span><span class="p">)</span>
    <span class="k">return</span> <span class="kc">None</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">descriptor</span><span class="p">,</span> <span class="n">similarity_df</span> <span class="ow">in</span> <span class="n">kinase_distance_matrices</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">"</span><span class="si">{</span><span class="n">descriptor</span><span class="si">}</span><span class="s2">"</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">"=========="</span><span class="p">)</span>
    <span class="n">x</span> <span class="o">=</span> <span class="n">distance_condition</span><span class="p">(</span><span class="n">similarity_df</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
sequence
==========
Dimensionality                Satisfied
Postivity                     Satisfied
Null diagonal                 Satisfied
Symmetry                      Satisfied
Triangular inequality         Satisfied


kissim
==========
Dimensionality                Satisfied
Postivity                     Satisfied
Null diagonal                 Satisfied
Symmetry                      Satisfied
Triangular inequality         Satisfied


ifp
==========
Dimensionality                Satisfied
Postivity                     Satisfied
Null diagonal                 Satisfied
Symmetry                      Satisfied
Triangular inequality         Satisfied


ligand-profile
==========
Dimensionality                Satisfied
Postivity                     Satisfied
Null diagonal                 Satisfied
Symmetry                      Satisfied
Triangular inequality         Not Satisfied


</pre></div></div>
</div>
<p>If one of the following checks (dimensionality, positivity, null diagonal, or symmetry) fails, the rest of the notebook will not run properly and the matrix generation from the previous notebooks should be investigated.</p>
<p>The triangular inequality does not have to be satisfied in order for the notebook to run, as is the case for ligand-profile. The reason for its failing is due to the function that describes the (dis)similarity between two kinases, which is not a distance (see <strong>Talktorial T027</strong>).</p>
</section>
<section id="Visualize-similarity-for-example-perspective">
<h3 id="Visualize-similarity-for-example-perspective">Visualize similarity for example perspective<a class="headerlink" href="#Visualize-similarity-for-example-perspective" title="Permalink to this heading">¶</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">"Choices of precalculated descriptors: </span><span class="si">{</span><span class="n">kinase_distance_matrix_paths</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span><span class="si">}</span><span class="s2">"</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Choices of precalculated descriptors: dict_keys(['sequence', 'kissim', 'ifp', 'ligand-profile'])
</pre></div></div>
</div>
<p>We look at an example matrix:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">descriptor_selection</span> <span class="o">=</span> <span class="s2">"sequence"</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">kinase_distance_matrix</span> <span class="o">=</span> <span class="n">kinase_distance_matrices</span><span class="p">[</span><span class="n">descriptor_selection</span><span class="p">]</span>
<span class="n">kinase_distance_matrix</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
<thead>
<tr style="text-align: right;">
<th></th>
<th>EGFR</th>
<th>ErbB2</th>
<th>p110a</th>
<th>KDR</th>
<th>BRAF</th>
<th>CDK2</th>
<th>LCK</th>
<th>MET</th>
<th>p38a</th>
</tr>
</thead>
<tbody>
<tr>
<th>EGFR</th>
<td>0.000000</td>
<td>0.059037</td>
<td>0.572938</td>
<td>0.283953</td>
<td>0.344972</td>
<td>0.351553</td>
<td>0.288679</td>
<td>0.288742</td>
<td>0.355356</td>
</tr>
<tr>
<th>ErbB2</th>
<td>0.059037</td>
<td>0.000000</td>
<td>0.586362</td>
<td>0.297882</td>
<td>0.345400</td>
<td>0.369692</td>
<td>0.314925</td>
<td>0.302033</td>
<td>0.364827</td>
</tr>
<tr>
<th>p110a</th>
<td>0.572938</td>
<td>0.586362</td>
<td>0.000000</td>
<td>0.577295</td>
<td>0.563541</td>
<td>0.576006</td>
<td>0.548664</td>
<td>0.606301</td>
<td>0.568806</td>
</tr>
<tr>
<th>KDR</th>
<td>0.283953</td>
<td>0.297882</td>
<td>0.577295</td>
<td>0.000000</td>
<td>0.328732</td>
<td>0.346621</td>
<td>0.312352</td>
<td>0.286194</td>
<td>0.346556</td>
</tr>
<tr>
<th>BRAF</th>
<td>0.344972</td>
<td>0.345400</td>
<td>0.563541</td>
<td>0.328732</td>
<td>0.000000</td>
<td>0.353245</td>
<td>0.327067</td>
<td>0.361842</td>
<td>0.362088</td>
</tr>
<tr>
<th>CDK2</th>
<td>0.351553</td>
<td>0.369692</td>
<td>0.576006</td>
<td>0.346621</td>
<td>0.353245</td>
<td>0.000000</td>
<td>0.318747</td>
<td>0.343975</td>
<td>0.276907</td>
</tr>
<tr>
<th>LCK</th>
<td>0.288679</td>
<td>0.314925</td>
<td>0.548664</td>
<td>0.312352</td>
<td>0.327067</td>
<td>0.318747</td>
<td>0.000000</td>
<td>0.309121</td>
<td>0.337419</td>
</tr>
<tr>
<th>MET</th>
<td>0.288742</td>
<td>0.302033</td>
<td>0.606301</td>
<td>0.286194</td>
<td>0.361842</td>
<td>0.343975</td>
<td>0.309121</td>
<td>0.000000</td>
<td>0.370645</td>
</tr>
<tr>
<th>p38a</th>
<td>0.355356</td>
<td>0.364827</td>
<td>0.568806</td>
<td>0.346556</td>
<td>0.362088</td>
<td>0.276907</td>
<td>0.337419</td>
<td>0.370645</td>
<td>0.000000</td>
</tr>
</tbody>
</table>
</div></div>
</div>
<section id="Visualize-kinase-similarity-matrix-as-heatmap">
<h4 id="Visualize-kinase-similarity-matrix-as-heatmap">Visualize kinase similarity matrix as heatmap<a class="headerlink" href="#Visualize-kinase-similarity-matrix-as-heatmap" title="Permalink to this heading">¶</a></h4>
<p>We visualize the kinase similarity matrix in the form of a heatmap.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span>
    <span class="mi">1</span> <span class="o">-</span> <span class="n">kinase_distance_matrix</span><span class="p">,</span>
    <span class="n">linewidths</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
    <span class="n">annot</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">square</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">cbar_kws</span><span class="o">=</span><span class="p">{</span><span class="s2">"label"</span><span class="p">:</span> <span class="s2">"Kinase distance"</span><span class="p">},</span>
    <span class="n">cmap</span><span class="o">=</span><span class="s2">"viridis"</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/talktorials_T028_kinase_similarity_compare_perspectives_39_0.png" src="../_images/talktorials_T028_kinase_similarity_compare_perspectives_39_0.png"/>
</div>
</div>
</section>
<section id="Visualize-similarity-as-dendrogram">
<h4 id="Visualize-similarity-as-dendrogram">Visualize similarity as dendrogram<a class="headerlink" href="#Visualize-similarity-as-dendrogram" title="Permalink to this heading">¶</a></h4>
<p>Since distance matrices are symmetric, we use the <code class="docutils literal notranslate"><span class="pre">scipy</span></code> function <code class="docutils literal notranslate"><span class="pre">squareform</span></code> to create a condensed vector of the distance matrix of shape <span class="math notranslate nohighlight">\(n*(n-1)/2\)</span>, where <span class="math notranslate nohighlight">\(n\)</span> is the shape of the distance matrix. The values in this vector correspond to the values of the lower triangular matrix.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">D</span> <span class="o">=</span> <span class="n">kinase_distance_matrix</span><span class="o">.</span><span class="n">values</span>
<span class="n">D_condensed</span> <span class="o">=</span> <span class="n">distance</span><span class="o">.</span><span class="n">squareform</span><span class="p">(</span><span class="n">D</span><span class="p">)</span>
<span class="n">D_condensed</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
array([0.059037, 0.572938, 0.283953, 0.344972, 0.351553, 0.288679,
       0.288742, 0.355356, 0.586362, 0.297882, 0.3454  , 0.369692,
       0.314925, 0.302033, 0.364827, 0.577295, 0.563541, 0.576006,
       0.548664, 0.606301, 0.568806, 0.328732, 0.346621, 0.312352,
       0.286194, 0.346556, 0.353245, 0.327067, 0.361842, 0.362088,
       0.318747, 0.343975, 0.276907, 0.309121, 0.337419, 0.370645])
</pre></div></div>
</div>
<p>We can submit this condensed vector to a hierarchical clustering to extract the relationship between the different kinases. We use here <code class="docutils literal notranslate"><span class="pre">method="average"</span></code>, which stands for the linkage method UPGMA (unweighted pair group method with arithmetic mean). This means that the distance between two clusters A and B is defined as the average of all distances between pairs of elements in clusters A and B. At each clustering step, the two clusters with the lowest average distance are combined.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">hclust</span> <span class="o">=</span> <span class="n">hierarchy</span><span class="o">.</span><span class="n">linkage</span><span class="p">(</span><span class="n">D_condensed</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s2">"average"</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>We now generate a phylogenetic tree based on the clustering.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tree</span> <span class="o">=</span> <span class="n">hierarchy</span><span class="o">.</span><span class="n">to_tree</span><span class="p">(</span><span class="n">hclust</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>We visualize the tree as a dendrogram.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">()</span>
<span class="n">labels</span> <span class="o">=</span> <span class="n">kinase_distance_matrix</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">to_list</span><span class="p">()</span>
<span class="n">hierarchy</span><span class="o">.</span><span class="n">dendrogram</span><span class="p">(</span><span class="n">hclust</span><span class="p">,</span> <span class="n">labels</span><span class="o">=</span><span class="n">labels</span><span class="p">,</span> <span class="n">orientation</span><span class="o">=</span><span class="s2">"left"</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">"Distance"</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/talktorials_T028_kinase_similarity_compare_perspectives_48_0.png" src="../_images/talktorials_T028_kinase_similarity_compare_perspectives_48_0.png"/>
</div>
</div>
</section>
</section>
<section id="Visualize-similarities-from-the-four-different-perspectives">
<h3 id="Visualize-similarities-from-the-four-different-perspectives">Visualize similarities from the four different perspectives<a class="headerlink" href="#Visualize-similarities-from-the-four-different-perspectives" title="Permalink to this heading">¶</a></h3>
<section id="Preprocess-distance-matrices">
<h4 id="Preprocess-distance-matrices">Preprocess distance matrices<a class="headerlink" href="#Preprocess-distance-matrices" title="Permalink to this heading">¶</a></h4>
<section id="Normalize-matrices">
<h5 id="Normalize-matrices">Normalize matrices<a class="headerlink" href="#Normalize-matrices" title="Permalink to this heading">¶</a></h5>
<p>We normalize the different matrices to values in <span class="math notranslate nohighlight">\([0, 1]\)</span>.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">_min_max_normalize</span><span class="p">(</span><span class="n">distance_matrix_df</span><span class="p">):</span>
<span class="w">    </span><span class="sd">"""</span>
<span class="sd">    Apply min-max normalization to input DataFrame.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    kinase_distance_matrix_df : pd.DataFrame</span>
<span class="sd">        Kinase distance matrix.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    pd.DataFrame</span>
<span class="sd">        Normalized kinase distance matrix.</span>
<span class="sd">    """</span>
    <span class="n">min_</span> <span class="o">=</span> <span class="n">distance_matrix_df</span><span class="o">.</span><span class="n">min</span><span class="p">()</span><span class="o">.</span><span class="n">min</span><span class="p">()</span>
    <span class="n">max_</span> <span class="o">=</span> <span class="n">distance_matrix_df</span><span class="o">.</span><span class="n">max</span><span class="p">()</span><span class="o">.</span><span class="n">max</span><span class="p">()</span>
    <span class="n">distance_matrix_normalized_df</span> <span class="o">=</span> <span class="p">(</span><span class="n">distance_matrix_df</span> <span class="o">-</span> <span class="n">min_</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="n">max_</span> <span class="o">-</span> <span class="n">min_</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">distance_matrix_normalized_df</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">kinase_distance_matrices_normalized</span> <span class="o">=</span> <span class="p">{}</span>
<span class="k">for</span> <span class="n">descriptor</span><span class="p">,</span> <span class="n">score_df</span> <span class="ow">in</span> <span class="n">kinase_distance_matrices</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
    <span class="n">score_normalized_df</span> <span class="o">=</span> <span class="n">_min_max_normalize</span><span class="p">(</span><span class="n">score_df</span><span class="p">)</span>
    <span class="n">kinase_distance_matrices_normalized</span><span class="p">[</span><span class="n">descriptor</span><span class="p">]</span> <span class="o">=</span> <span class="n">score_normalized_df</span>
</pre></div>
</div>
</div>
</section>
<section id="Define-kinase-order">
<h5 id="Define-kinase-order">Define kinase order<a class="headerlink" href="#Define-kinase-order" title="Permalink to this heading">¶</a></h5>
<p>Define in which order to display the kinases.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">kinase_names</span> <span class="o">=</span> <span class="n">kinase_distance_matrices_normalized</span><span class="p">[</span><span class="s2">"sequence"</span><span class="p">]</span><span class="o">.</span><span class="n">columns</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">_define_kinase_order</span><span class="p">(</span><span class="n">kinase_distance_matrix_df</span><span class="p">,</span> <span class="n">kinase_names</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">""</span><span class="p">):</span>
<span class="w">    </span><span class="sd">"""</span>
<span class="sd">    Define the order in which kinases shall</span>
<span class="sd">    appear in the input DataFrame.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    kinase_distance_matrix_df : pd.DataFrame</span>
<span class="sd">        Kinase distance matrix.</span>
<span class="sd">    kinase_name : list of str</span>
<span class="sd">        List of kinase names to be used for sorting.</span>
<span class="sd">    label : str</span>
<span class="sd">        Add label for the input matrix.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    pd.DataFrame</span>
<span class="sd">        Kinase distance matrix with sorted columns/rows.</span>
<span class="sd">    """</span>
    <span class="c1"># Remove kinases from ordered kinase set that are not</span>
    <span class="c1"># present in input kinase matrix</span>
    <span class="n">kinase_names</span> <span class="o">=</span> <span class="p">[</span><span class="n">name</span> <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">kinase_names</span> <span class="k">if</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">kinase_distance_matrix_df</span><span class="o">.</span><span class="n">columns</span><span class="p">]</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">"Kinases present in </span><span class="si">{</span><span class="n">label</span><span class="si">}</span><span class="s2"> kinase matrix:</span><span class="se">\n</span><span class="s2"> </span><span class="si">{</span><span class="n">kinase_names</span><span class="si">}</span><span class="s2">"</span><span class="p">)</span>
    <span class="c1"># Reorder kinases in matrix</span>
    <span class="n">kinase_distance_matrix_df</span> <span class="o">=</span> <span class="n">kinase_distance_matrix_df</span><span class="o">.</span><span class="n">reindex</span><span class="p">(</span><span class="n">kinase_names</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">reindex</span><span class="p">(</span>
        <span class="n">kinase_names</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">kinase_distance_matrix_df</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">kinase_distance_matrices_normalized</span> <span class="o">=</span> <span class="p">{</span>
    <span class="n">descriptor</span><span class="p">:</span> <span class="n">_define_kinase_order</span><span class="p">(</span><span class="n">score_df</span><span class="p">,</span> <span class="n">kinase_names</span><span class="p">,</span> <span class="n">descriptor</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">descriptor</span><span class="p">,</span> <span class="n">score_df</span> <span class="ow">in</span> <span class="n">kinase_distance_matrices_normalized</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
<span class="p">}</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Kinases present in sequence kinase matrix:
 ['EGFR', 'ErbB2', 'p110a', 'KDR', 'BRAF', 'CDK2', 'LCK', 'MET', 'p38a']
Kinases present in kissim kinase matrix:
 ['EGFR', 'ErbB2', 'p110a', 'KDR', 'BRAF', 'CDK2', 'LCK', 'MET', 'p38a']
Kinases present in ifp kinase matrix:
 ['EGFR', 'ErbB2', 'p110a', 'KDR', 'BRAF', 'CDK2', 'LCK', 'MET', 'p38a']
Kinases present in ligand-profile kinase matrix:
 ['EGFR', 'ErbB2', 'p110a', 'KDR', 'BRAF', 'CDK2', 'LCK', 'MET', 'p38a']
</pre></div></div>
</div>
</section>
</section>
<section id="Visualize-kinase-similarities">
<h4 id="Visualize-kinase-similarities">Visualize kinase similarities<a class="headerlink" href="#Visualize-kinase-similarities" title="Permalink to this heading">¶</a></h4>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">heatmap</span><span class="p">(</span><span class="n">score_df</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s2">""</span><span class="p">):</span>
<span class="w">    </span><span class="sd">"""</span>
<span class="sd">    Generate a heatmap from a matrix.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    score_df : pd.DataFrame</span>
<span class="sd">        Distance or similarity score matrix.</span>
<span class="sd">    ax : matplotlib.axes</span>
<span class="sd">        Plot axis to use!</span>
<span class="sd">    title : str</span>
<span class="sd">        Plot title.</span>
<span class="sd">    """</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">score_df</span><span class="p">,</span> <span class="n">linewidths</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">annot</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">square</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s2">"viridis"</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">dendrogram</span><span class="p">(</span><span class="n">distance_matrix</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s2">""</span><span class="p">):</span>
<span class="w">    </span><span class="sd">"""</span>
<span class="sd">    Generate a dendrogram from a distance matrix.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    distance_matrix : pd.DataFrame</span>
<span class="sd">        Distance matrix.</span>
<span class="sd">    ax : matplotlib.axes</span>
<span class="sd">        Plot axis to use!</span>
<span class="sd">    title : str</span>
<span class="sd">        Plot title.</span>
<span class="sd">    """</span>
    <span class="n">D</span> <span class="o">=</span> <span class="n">distance_matrix</span><span class="o">.</span><span class="n">values</span>
    <span class="n">D_condensed</span> <span class="o">=</span> <span class="n">distance</span><span class="o">.</span><span class="n">squareform</span><span class="p">(</span><span class="n">D</span><span class="p">)</span>
    <span class="n">hclust</span> <span class="o">=</span> <span class="n">hierarchy</span><span class="o">.</span><span class="n">linkage</span><span class="p">(</span><span class="n">D_condensed</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s2">"average"</span><span class="p">)</span>
    <span class="n">tree</span> <span class="o">=</span> <span class="n">hierarchy</span><span class="o">.</span><span class="n">to_tree</span><span class="p">(</span><span class="n">hclust</span><span class="p">)</span>
    <span class="n">labels</span> <span class="o">=</span> <span class="n">distance_matrix</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">to_list</span><span class="p">()</span>
    <span class="n">hierarchy</span><span class="o">.</span><span class="n">dendrogram</span><span class="p">(</span><span class="n">hclust</span><span class="p">,</span> <span class="n">labels</span><span class="o">=</span><span class="n">labels</span><span class="p">,</span> <span class="n">orientation</span><span class="o">=</span><span class="s2">"left"</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">title</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">"Distance"</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Note for ease of interpretability, we show below:</p>
<ul class="simple">
<li><p>Heatmaps based on the <strong>similarity</strong> matrix (the higher the value, the higher the similarity).</p></li>
<li><p>Dendrograms calculated based on the <strong>distance</strong> matrix, where clusters describe the similarity between kinases.</p></li>
</ul>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">n_perspectives</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">kinase_distance_matrices_normalized</span><span class="p">)</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="n">n_perspectives</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="n">n_perspectives</span> <span class="o">*</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">10</span><span class="p">))</span>
<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">(</span><span class="n">perspective</span><span class="p">,</span> <span class="n">distance_matrix</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">kinase_distance_matrices_normalized</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>
    <span class="c1"># heatmap based on similarity matrix</span>
    <span class="n">similarity_matrix</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">-</span> <span class="n">distance_matrix</span>
    <span class="n">heatmap</span><span class="p">(</span><span class="n">similarity_matrix</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="n">i</span><span class="p">],</span> <span class="n">title</span><span class="o">=</span><span class="n">perspective</span><span class="p">)</span>
    <span class="c1"># dendrogram based on distance matrix</span>
    <span class="n">dendrogram</span><span class="p">(</span><span class="n">distance_matrix</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="n">i</span><span class="p">],</span> <span class="n">title</span><span class="o">=</span><span class="n">perspective</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/talktorials_T028_kinase_similarity_compare_perspectives_62_0.png" src="../_images/talktorials_T028_kinase_similarity_compare_perspectives_62_0.png"/>
</div>
</div>
<p>We load here again the input kinase information that we defined in <strong>Talktorial T023</strong>, to determine, among other, to which group each kinase belongs to.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">kinase_selection_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">HERE</span> <span class="o">/</span> <span class="s2">"../T023_what_is_a_kinase/data/kinase_selection.csv"</span><span class="p">)</span>
<span class="n">kinase_selection_df</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
<thead>
<tr style="text-align: right;">
<th></th>
<th>kinase</th>
<th>kinase_klifs</th>
<th>uniprot_id</th>
<th>group</th>
<th>full_kinase_name</th>
</tr>
</thead>
<tbody>
<tr>
<th>0</th>
<td>EGFR</td>
<td>EGFR</td>
<td>P00533</td>
<td>TK</td>
<td>Epidermal growth factor receptor</td>
</tr>
<tr>
<th>1</th>
<td>ErbB2</td>
<td>ErbB2</td>
<td>P04626</td>
<td>TK</td>
<td>Erythroblastic leukemia viral oncogene homolog 2</td>
</tr>
<tr>
<th>2</th>
<td>PI3K</td>
<td>p110a</td>
<td>P42336</td>
<td>Atypical</td>
<td>Phosphatidylinositol-3-kinase</td>
</tr>
<tr>
<th>3</th>
<td>VEGFR2</td>
<td>KDR</td>
<td>P35968</td>
<td>TK</td>
<td>Vascular endothelial growth factor receptor 2</td>
</tr>
<tr>
<th>4</th>
<td>BRAF</td>
<td>BRAF</td>
<td>P15056</td>
<td>TKL</td>
<td>Rapidly accelerated fibrosarcoma isoform B</td>
</tr>
<tr>
<th>5</th>
<td>CDK2</td>
<td>CDK2</td>
<td>P24941</td>
<td>CMGC</td>
<td>Cyclic-dependent kinase 2</td>
</tr>
<tr>
<th>6</th>
<td>LCK</td>
<td>LCK</td>
<td>P06239</td>
<td>TK</td>
<td>Lymphocyte-specific protein tyrosine kinase</td>
</tr>
<tr>
<th>7</th>
<td>MET</td>
<td>MET</td>
<td>P08581</td>
<td>TK</td>
<td>Mesenchymal-epithelial transition factor</td>
</tr>
<tr>
<th>8</th>
<td>p38a</td>
<td>p38a</td>
<td>Q16539</td>
<td>CMGC</td>
<td>p38 mitogen activated protein kinase alpha</td>
</tr>
</tbody>
</table>
</div></div>
</div>
</section>
<section id="Analysis-of-results">
<h4 id="Analysis-of-results">Analysis of results<a class="headerlink" href="#Analysis-of-results" title="Permalink to this heading">¶</a></h4>
<ul class="simple">
<li><p>The <em>sequence</em> approach reproduces nicely the Manning tree relationships.</p>
<ul>
<li><p>The TK kinases EGFR, ErbB2, MET, and KDR are grouped together with highest similarities for EGFR and ErbB2.</p></li>
<li><p>The TKL kinase LCK groups next to the TK kinases.</p></li>
<li><p>The CMGC kinases p38a and CDK2 are grouped together.</p></li>
<li><p>The atypical kinase p110a is a singleton, i.e., forms its own cluster.</p></li>
</ul>
</li>
<li><p>When taking into account the pocket structure (physicochemical and spatial properties) with the <em>kissim</em> approach, similar observations can be made:</p>
<ul>
<li><p>The typical kinases are grouped together, while the atypical kinase p110a is a singleton.</p></li>
<li><p>Within the typical kinase grouping, EGFR and ErbB2 are the most similar (but not as prominent as in the <em>sequence</em> approach).</p></li>
<li><p>Kinases from different groups, e.g. CDK2 (CMGC) and BRAF (TKL), cluster together.</p></li>
</ul>
</li>
<li><p>The <em>ifp</em> method:</p>
<ul>
<li><p>BRAF and the atypical kinase p110a form a singleton each.</p></li>
<li><p>Surprisingly, ErbB2 forms a singleton (=not grouped with EGFR), but we have to keep in mind that we have the least data points (<span class="math notranslate nohighlight">\(4\)</span>) for ErbB2.</p></li>
</ul>
</li>
<li><p>The <em>ligand-profile</em> method:</p>
<ul>
<li><p>Interestingly, CDK2 and p110a form their own cluster early on (i.e. low similarity to the rest).</p></li>
<li><p>BRAF (TKL), p38a (CMGC), KDR and LCK (both TK) as well as EGFR and ErbB2 (TK) form one cluster each.</p></li>
<li><p>Surprisingly, MET (TK) forms a singleton (not clustering with any of the other TK kinases).</p></li>
</ul>
</li>
<li><p>Overall observations:</p>
<ul>
<li><p>Members of the same family (here: EGFR and ErbB2) are grouped together closely in the <em>sequence</em>, <em>kissim</em>, and <em>ligand-profile</em> approaches.</p></li>
<li><p>The atypical kinase p110a shows high dissimilarity in the <em>sequence</em>, <em>kissim</em>, and <em>ifp</em> approaches.</p></li>
<li><p>Unexpectedly, KDR (TK) and p38a (CMGC) cluster together in all views (except for <em>sequence</em>). Thus, it would be interesting to explore this relationship further.</p></li>
</ul>
</li>
</ul>
<!---
- Comparison to multi-target profiles in [<i>Molecules</i> (2021), <b>26(3)</b>, 629](https://www.mdpi.com/1420-3049/26/3/629)
  - Profile 1:
    - EGFR/ErbB2 as on-targets and BRAF as anti-target
    - Should be doable from "sequence", "ligand-profile" view (less clear from "kisim" and "ifp" view)
  - Profile 2:
    - EGFR/p110a as on-targets and BRAF as anti-target
    - Should be hard to design from all views
  - Profile 3:
    - EGFR/KDR as on-targets and BRAF as anti-target
    - Should be hard to design except from a "sequence" view
---></section>
</section>
</section>
<section id="Discussion">
<h2 id="Discussion">Discussion<a class="headerlink" href="#Discussion" title="Permalink to this heading">¶</a></h2>
<p>We have assessed kinase similarity from many different angles: the pocket sequence, the pocket structure, ligand binding modes (interaction fingerprints), and ligand bioactivities. We see from our results that different conclusions can be drawn for each method, while some observations agree. Thus, it is beneficial to consider different strategies to study kinase similarity.</p>
<p>The different approaches have strengths and drawbacks:</p>
<ul class="simple">
<li><p>The <em>sequence</em> method: Sequence data is available for all kinases, however no information about structure or ligand binding is included.</p></li>
<li><p>The <em>kissim</em>, <em>ifp</em>, and <em>ligand-profile</em> methods: Information about structure, ligand binding, or ligand bioactivity is included, however some kinases may not have any data, some kinases may have only few data, and most kinases may not be fully explored (even if a lot of data is available).</p></li>
<li><p>All methods rely on simplifications!</p>
<ul>
<li><p>Dealing with multiple measurements</p>
<ul>
<li><p><em>ligand-profile</em> method: Only one measurement per kinase-ligand is used here.</p></li>
<li><p><em>kissim</em> and <em>ifp</em> method: Only one structure/IFP pair represents one kinase pair here.</p></li>
</ul>
</li>
<li><p>Bioactivity cutoffs to define active/inactive compounds in case of the <em>ligand-profile</em> method.</p></li>
</ul>
</li>
</ul>
</section>
<section id="Quiz">
<h2 id="Quiz">Quiz<a class="headerlink" href="#Quiz" title="Permalink to this heading">¶</a></h2>
<ol class="arabic simple">
<li><p>Can you think of another perspective to assess kinase similarity?</p></li>
<li><p>Are there other methods to assess and visualize the similarities?</p></li>
<li><p>Can you name at least one advantage and one disadvantage per perspective discussed here?</p></li>
<li><p>Could the different methods be applied to other proteins in general?</p></li>
<li><p>What is the difference between the distances that we show in the heatmap and the dendrogram?</p></li>
<li><p>Rerun the notebooks T023-T028 on another kinase dataset.</p>
<ol class="arabic simple">
<li><p>If you need inspiration, you can use the kinases used by <a class="reference external" href="https://doi.org/10.1371/journal.pcbi.1009302">Xiong et al.</a> to predict multi-targeting compounds for RET-driven cancers (see Table 1): RET, BRAF, SRC, RPS6KB1, MKNK1, TTK, PDK1, and PAK3 (notes: We are omitting ERK8 since this kinase has no structures; RPS6KB1 is listed as S6K in Table 1.). Update the <code class="docutils literal notranslate"><span class="pre">T023_what_is_a_kinase/data/kinase_selection.csv</span></code> file yourself with the mandatory columns <code class="docutils literal notranslate"><span class="pre">kinase_klifs</span></code> and <code class="docutils literal notranslate"><span class="pre">uniprot_id</span></code>
(or copy-paste the data from <code class="docutils literal notranslate"><span class="pre">T023_what_is_a_kinase/data/kinase_selection_xiong.csv</span></code>).</p></li>
<li><p>Update the configuration file <code class="docutils literal notranslate"><span class="pre">T023_what_is_a_kinase/data/pipeline_configs.csv</span></code> as follows: <code class="docutils literal notranslate"><span class="pre">DEMO=0</span></code>, <code class="docutils literal notranslate"><span class="pre">N_STRUCTURES_PER_KINASE=2</span></code> (the more structures used, the longer T025 will take), and <code class="docutils literal notranslate"><span class="pre">N_CORES</span></code> as your system allows.</p></li>
</ol>
</li>
</ol>
</section>
<section id="Appendix">
<h2 id="Appendix">Appendix<a class="headerlink" href="#Appendix" title="Permalink to this heading">¶</a></h2>
<p>In this appendix, we generate individual plots (both heatmap and dendrogram) for the sequence perspective. The plots are saved in svg format.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[29]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">()</span>
<span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span>
    <span class="mi">1</span> <span class="o">-</span> <span class="n">kinase_distance_matrices_normalized</span><span class="p">[</span><span class="s2">"sequence"</span><span class="p">],</span>
    <span class="n">linewidths</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
    <span class="n">annot</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">square</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">cbar_kws</span><span class="o">=</span><span class="p">{</span><span class="s2">"label"</span><span class="p">:</span> <span class="s2">"Kinase similarity"</span><span class="p">},</span>
    <span class="n">cmap</span><span class="o">=</span><span class="s2">"viridis"</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">"sequence"</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">DATA</span> <span class="o">/</span> <span class="s2">"heatmap_sequence.svg"</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/talktorials_T028_kinase_similarity_compare_perspectives_71_0.png" src="../_images/talktorials_T028_kinase_similarity_compare_perspectives_71_0.png"/>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">()</span>
<span class="n">D</span> <span class="o">=</span> <span class="n">kinase_distance_matrices_normalized</span><span class="p">[</span><span class="s2">"sequence"</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
<span class="n">D_condensed</span> <span class="o">=</span> <span class="n">distance</span><span class="o">.</span><span class="n">squareform</span><span class="p">(</span><span class="n">D</span><span class="p">)</span>
<span class="n">hclust</span> <span class="o">=</span> <span class="n">hierarchy</span><span class="o">.</span><span class="n">linkage</span><span class="p">(</span><span class="n">D_condensed</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s2">"average"</span><span class="p">)</span>
<span class="n">tree</span> <span class="o">=</span> <span class="n">hierarchy</span><span class="o">.</span><span class="n">to_tree</span><span class="p">(</span><span class="n">hclust</span><span class="p">)</span>
<span class="n">labels</span> <span class="o">=</span> <span class="n">distance_matrix</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">to_list</span><span class="p">()</span>
<span class="n">hierarchy</span><span class="o">.</span><span class="n">dendrogram</span><span class="p">(</span><span class="n">hclust</span><span class="p">,</span> <span class="n">labels</span><span class="o">=</span><span class="n">labels</span><span class="p">,</span> <span class="n">orientation</span><span class="o">=</span><span class="s2">"left"</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">"sequence"</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">"Distance"</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="mi">1</span> <span class="o">/</span> <span class="mi">100</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">DATA</span> <span class="o">/</span> <span class="s2">"dendrogram_sequence.svg"</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/talktorials_T028_kinase_similarity_compare_perspectives_73_0.png" src="../_images/talktorials_T028_kinase_similarity_compare_perspectives_73_0.png"/>
</div>
</div>
<section id="One-single-measure-of-comparison">
<h3 id="One-single-measure-of-comparison">One single measure of comparison<a class="headerlink" href="#One-single-measure-of-comparison" title="Permalink to this heading">¶</a></h3>
<p>Instead of having four different measures of similarity, we compute a equally weighted average to obtain one single similarity matrix (and associated heatmap), and one single distance matrix (and associated dendrogram).</p>
<p>Weighting differently the various perspectives could be further investigated.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[31]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">combined_distance_matrix</span> <span class="o">=</span> <span class="p">(</span>
    <span class="nb">sum</span><span class="p">([</span><span class="n">distance_matrix</span> <span class="k">for</span> <span class="n">distance_matrix</span> <span class="ow">in</span> <span class="n">kinase_distance_matrices_normalized</span><span class="o">.</span><span class="n">values</span><span class="p">()])</span>
    <span class="o">/</span> <span class="n">n_perspectives</span>
<span class="p">)</span>
<span class="n">combined_distance_matrix</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[31]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
<thead>
<tr style="text-align: right;">
<th></th>
<th>EGFR</th>
<th>ErbB2</th>
<th>p110a</th>
<th>KDR</th>
<th>BRAF</th>
<th>CDK2</th>
<th>LCK</th>
<th>MET</th>
<th>p38a</th>
</tr>
</thead>
<tbody>
<tr>
<th>EGFR</th>
<td>0.000000</td>
<td>0.345073</td>
<td>0.853070</td>
<td>0.488292</td>
<td>0.517561</td>
<td>0.522226</td>
<td>0.480177</td>
<td>0.517406</td>
<td>0.520092</td>
</tr>
<tr>
<th>ErbB2</th>
<td>0.345073</td>
<td>0.000000</td>
<td>0.991778</td>
<td>0.595015</td>
<td>0.631776</td>
<td>0.690813</td>
<td>0.670081</td>
<td>0.669546</td>
<td>0.631571</td>
</tr>
<tr>
<th>p110a</th>
<td>0.853070</td>
<td>0.991778</td>
<td>0.000000</td>
<td>0.924348</td>
<td>0.845578</td>
<td>0.779110</td>
<td>0.876636</td>
<td>0.926215</td>
<td>0.925134</td>
</tr>
<tr>
<th>KDR</th>
<td>0.488292</td>
<td>0.595015</td>
<td>0.924348</td>
<td>0.000000</td>
<td>0.482481</td>
<td>0.463151</td>
<td>0.412795</td>
<td>0.502286</td>
<td>0.405543</td>
</tr>
<tr>
<th>BRAF</th>
<td>0.517561</td>
<td>0.631776</td>
<td>0.845578</td>
<td>0.482481</td>
<td>0.000000</td>
<td>0.611584</td>
<td>0.518079</td>
<td>0.654472</td>
<td>0.499924</td>
</tr>
<tr>
<th>CDK2</th>
<td>0.522226</td>
<td>0.690813</td>
<td>0.779110</td>
<td>0.463151</td>
<td>0.611584</td>
<td>0.000000</td>
<td>0.545841</td>
<td>0.538301</td>
<td>0.575915</td>
</tr>
<tr>
<th>LCK</th>
<td>0.480177</td>
<td>0.670081</td>
<td>0.876636</td>
<td>0.412795</td>
<td>0.518079</td>
<td>0.545841</td>
<td>0.000000</td>
<td>0.535791</td>
<td>0.483474</td>
</tr>
<tr>
<th>MET</th>
<td>0.517406</td>
<td>0.669546</td>
<td>0.926215</td>
<td>0.502286</td>
<td>0.654472</td>
<td>0.538301</td>
<td>0.535791</td>
<td>0.000000</td>
<td>0.677882</td>
</tr>
<tr>
<th>p38a</th>
<td>0.520092</td>
<td>0.631571</td>
<td>0.925134</td>
<td>0.405543</td>
<td>0.499924</td>
<td>0.575915</td>
<td>0.483474</td>
<td>0.677882</td>
<td>0.000000</td>
</tr>
</tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[32]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">()</span>
<span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span>
    <span class="mi">1</span> <span class="o">-</span> <span class="n">combined_distance_matrix</span><span class="p">,</span>
    <span class="n">linewidths</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
    <span class="n">annot</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">square</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">cbar_kws</span><span class="o">=</span><span class="p">{</span><span class="s2">"label"</span><span class="p">:</span> <span class="s2">"Kinase similarity"</span><span class="p">},</span>
    <span class="n">cmap</span><span class="o">=</span><span class="s2">"viridis"</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">"combined measures for similarity"</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/talktorials_T028_kinase_similarity_compare_perspectives_76_0.png" src="../_images/talktorials_T028_kinase_similarity_compare_perspectives_76_0.png"/>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[33]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># In case one or more perspectives have missing kinases,</span>
<span class="c1"># the combined distance matrix will contain NaN values</span>
<span class="c1"># Remove the missing kinases (and NaN values) from the matrix</span>
<span class="c1"># before plotting the dendrogram</span>
<span class="n">combined_distance_matrix</span> <span class="o">=</span> <span class="n">combined_distance_matrix</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">"all"</span><span class="p">)</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span>
    <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">"all"</span>
<span class="p">)</span>
<span class="n">combined_distance_matrix</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[33]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
<thead>
<tr style="text-align: right;">
<th></th>
<th>EGFR</th>
<th>ErbB2</th>
<th>p110a</th>
<th>KDR</th>
<th>BRAF</th>
<th>CDK2</th>
<th>LCK</th>
<th>MET</th>
<th>p38a</th>
</tr>
</thead>
<tbody>
<tr>
<th>EGFR</th>
<td>0.000000</td>
<td>0.345073</td>
<td>0.853070</td>
<td>0.488292</td>
<td>0.517561</td>
<td>0.522226</td>
<td>0.480177</td>
<td>0.517406</td>
<td>0.520092</td>
</tr>
<tr>
<th>ErbB2</th>
<td>0.345073</td>
<td>0.000000</td>
<td>0.991778</td>
<td>0.595015</td>
<td>0.631776</td>
<td>0.690813</td>
<td>0.670081</td>
<td>0.669546</td>
<td>0.631571</td>
</tr>
<tr>
<th>p110a</th>
<td>0.853070</td>
<td>0.991778</td>
<td>0.000000</td>
<td>0.924348</td>
<td>0.845578</td>
<td>0.779110</td>
<td>0.876636</td>
<td>0.926215</td>
<td>0.925134</td>
</tr>
<tr>
<th>KDR</th>
<td>0.488292</td>
<td>0.595015</td>
<td>0.924348</td>
<td>0.000000</td>
<td>0.482481</td>
<td>0.463151</td>
<td>0.412795</td>
<td>0.502286</td>
<td>0.405543</td>
</tr>
<tr>
<th>BRAF</th>
<td>0.517561</td>
<td>0.631776</td>
<td>0.845578</td>
<td>0.482481</td>
<td>0.000000</td>
<td>0.611584</td>
<td>0.518079</td>
<td>0.654472</td>
<td>0.499924</td>
</tr>
<tr>
<th>CDK2</th>
<td>0.522226</td>
<td>0.690813</td>
<td>0.779110</td>
<td>0.463151</td>
<td>0.611584</td>
<td>0.000000</td>
<td>0.545841</td>
<td>0.538301</td>
<td>0.575915</td>
</tr>
<tr>
<th>LCK</th>
<td>0.480177</td>
<td>0.670081</td>
<td>0.876636</td>
<td>0.412795</td>
<td>0.518079</td>
<td>0.545841</td>
<td>0.000000</td>
<td>0.535791</td>
<td>0.483474</td>
</tr>
<tr>
<th>MET</th>
<td>0.517406</td>
<td>0.669546</td>
<td>0.926215</td>
<td>0.502286</td>
<td>0.654472</td>
<td>0.538301</td>
<td>0.535791</td>
<td>0.000000</td>
<td>0.677882</td>
</tr>
<tr>
<th>p38a</th>
<td>0.520092</td>
<td>0.631571</td>
<td>0.925134</td>
<td>0.405543</td>
<td>0.499924</td>
<td>0.575915</td>
<td>0.483474</td>
<td>0.677882</td>
<td>0.000000</td>
</tr>
</tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[34]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">()</span>
<span class="n">D</span> <span class="o">=</span> <span class="n">combined_distance_matrix</span>
<span class="n">D_condensed</span> <span class="o">=</span> <span class="n">distance</span><span class="o">.</span><span class="n">squareform</span><span class="p">(</span><span class="n">D</span><span class="p">)</span>
<span class="n">hclust</span> <span class="o">=</span> <span class="n">hierarchy</span><span class="o">.</span><span class="n">linkage</span><span class="p">(</span><span class="n">D_condensed</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s2">"average"</span><span class="p">)</span>
<span class="n">tree</span> <span class="o">=</span> <span class="n">hierarchy</span><span class="o">.</span><span class="n">to_tree</span><span class="p">(</span><span class="n">hclust</span><span class="p">)</span>
<span class="n">labels</span> <span class="o">=</span> <span class="n">combined_distance_matrix</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">to_list</span><span class="p">()</span>
<span class="n">hierarchy</span><span class="o">.</span><span class="n">dendrogram</span><span class="p">(</span><span class="n">hclust</span><span class="p">,</span> <span class="n">labels</span><span class="o">=</span><span class="n">labels</span><span class="p">,</span> <span class="n">orientation</span><span class="o">=</span><span class="s2">"left"</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">"combined measures"</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">"Distance"</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="mi">1</span> <span class="o">/</span> <span class="mi">100</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/talktorials_T028_kinase_similarity_compare_perspectives_78_0.png" src="../_images/talktorials_T028_kinase_similarity_compare_perspectives_78_0.png"/>
</div>
</div>
<script type="application/vnd.jupyter.widget-state+json">
{"state": {}, "version_major": 2, "version_minor": 0}
</script></section>
</section>
</section>


          </article>
        </div>
      </div>
    </main>
  </div>
  <footer class="md-footer">
    <div class="md-footer-nav">
      <nav class="md-footer-nav__inner md-grid">
          
            <a href="T027_kinase_similarity_ligand_profile.html" title="T027 · Kinase similarity: Ligand profile"
               class="md-flex md-footer-nav__link md-footer-nav__link--prev"
               rel="prev">
              <div class="md-flex__cell md-flex__cell--shrink">
                <i class="md-icon md-icon--arrow-back md-footer-nav__button"></i>
              </div>
              <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
                <span class="md-flex__ellipsis">
                  <span
                      class="md-footer-nav__direction"> Previous </span> T027 · Kinase similarity: Ligand profile </span>
              </div>
            </a>
          
          
            <a href="../talktorials.html" title="Talktorials by collection"
               class="md-flex md-footer-nav__link md-footer-nav__link--next"
               rel="next">
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title"><span
                class="md-flex__ellipsis"> <span
                class="md-footer-nav__direction"> Next </span> Talktorials by collection </span>
            </div>
            <div class="md-flex__cell md-flex__cell--shrink"><i
                class="md-icon md-icon--arrow-forward md-footer-nav__button"></i>
            </div>
          
        </a>
        
      </nav>
    </div>
    <div class="md-footer-meta md-typeset">
      <div class="md-footer-meta__inner md-grid">
        <div class="md-footer-copyright">
          <div class="md-footer-copyright__highlight">
              &#169; Copyright 2018-2023, Volkamer Lab. Project structure based on the Computational Molecular Science Python Cookiecutter version 1.1.
              
          </div>
            Created using
            <a href="http://www.sphinx-doc.org/">Sphinx</a> 6.1.3.
             and
            <a href="https://github.com/bashtage/sphinx-material/">Material for
              Sphinx</a>
        </div>
      </div>
    </div>
  </footer>
  <script src="../_static/javascripts/application.js"></script>
  <script>app.initialize({version: "1.0.4", url: {base: ".."}})</script>
  </body>
</html>